## PRJNA315294

**Identification of long noncoding RNAs regulated by p53**

Analyize the transcriptpme regulated by p53 in 3 pairs of isogenic p53WT and 
p53KO colorectal cancer cell lines untreated or treated with Doxorubicin. 
The hypothsis of this study is that p53 can regulate the expression of subset 
of lncRNAs to mediate its biological functions. Overall design: Total RNA was 
isolated from 3 pairs of isogenic colon cancer cell lines; HCT116-p53_WT and 
p53_KO; RKO-p53_WT and p53_KO: and SW48-p53_WT and p53_KO. Cells were untreated 
or treated with Doxorubicin (DOXO) at a final concentration of 300 nM for 16 hours

## PRJNA387040 

**Expression profiling of etoposide-induced senescent cells**

We report here the RNA-seq data of proliferating IMR90 cells and etoposide-induced 
senescent cells with control or cGAS knockdown Overall design: RNA-seq in 
primary IMR90 cells, two independent biological replicates, sequenced in Illumina 
NextSeq 500 platform. 

## PRJNA417314

**Transcriptome profiling of colorectal adenocarcinoma HT-29 cells after 5-FU treatment**

We performed RNA-Seq analysis of transcriptomic alterations in HT-29 colon 
cancer cells treated with 20 µM 5-FU for 72 h, the time period roughly
equivalent to three consecutive cell doublings. We expected that the high 
dose of the drug, which corresponds to 10 x IC50 and kills nearly 80% of 
cells, would selectively favor the survival of cells that acquire a transient 
resistance to 5-FU.

We identified lincRNAs LINC00973, LINC00941, CCAT1, and CASC19 as the most 
upregulated in 5-FU treated HT-29 cells. These results were further 
validated by RT-qPCR in HT-29 cells harvested after 72 h and 144 h of treatment 
with 2 µM 5-FU or 2 µM OXP (corresponding to the IC50 dose) or 5 µM I
RI (one third of its IC50).

We also added lincRNA BCAR4 to the analysis, which is undetectable by RNA-Seq,
but according to our RT-qPCR data is upregulated in 57% of colorectal tumors. 
Obtained results demonstrated that all five lincRNAs are upregulated by the 
treatment with all three drugs, indicating that monitoring by RT-qPCR of just 
two of the lincRNAs (LINC00973 and BCAR4) provides a robust instrument for 
distinguishing of untreated and treated cells.

## PRJNA335388  

**RNA sequencing of MDA-MB231 and U2OS cancer cell lines exposed to alkylating methylsufonate (MMS) and classical chemotherapeutics**

Understanding the mechanisms by which cells respond to chemotherapeutics is key
to identifying means to improve therapy effiicacy while reducing systemic 
toxicity of these widely used classes of drugs. While determining the role 
of NRF2-GSH and ER stress in cells exposed to alkylating compounds such as 
methyl-methanesulfonate (MMS), we asked if these pathways could also be a 
general cell damage response relevant to other clinically used chemotherapeutics 
or if it is an alkylation specific response. With this intent, we performed 
RNA sequencing of MDA-MB231 breast cancer and U2OS osteosarcoma cells lines 
treated for 8 hours with a topoisomerase II inhibitor etoposide (20 µM), the
antimitotic beta-tubulin-interacting drug paclitaxel (0.2 µM), doxorubicin
(1 µM) and compared to MMS (40 µg/mL) treated cells. Doses represent IC50 lev
el after 72 hours exposure. We observed that even though non-alkylating drugs,
especially etoposide, caused an increase in the mRNA expression of some NRF2 
and ER stress signaling markers, the number and magnitude of upregulation of 
genes markers in either pathway was more pronounced in alkylation treatments 
compared to other drugs. This indicates that alterations in NRF2 and ER stress 
pathways could be more likely associated with differential sensitivity to 
alkylating chemotherapies. Overall design: MDA-MB231 breast cancer and U2OS 
osteosarcoma cells lines were treated with the 72 h IC50  dose of etoposide
(20 µM), paclitaxel (0.2 µM),  doxorubicin (1 µM) or  MMS (40 µg/mL) for 8
 h, and RNA was extracted and analyzed.

## PRJNA392914

**Repression of stress-induced LINE-1 expression protects cancer cell populations from lethal drug-exposures**

DTP heterochromatin in genomic repeat regions protects the population from 
drug-induced death. Overall design: Investigation of RNA expression changes in 
heterogenous parental populations as compared to drug treated and drugtolerant 
populations. Also included are TSA treated parental and drug-tolerant populations

## PRJNA260377

**Specific molecular signatures underlie response to decitabine in CMML**

Myelodysplastic syndromes and chronic myelomonocytic leukemia (CMML) are 
characterized by mutations in epigenetic modifiers and aberrant DNA methylation. 
DNA methyltransferase inhibitors (DMTis) are used to treat these disorders, 
but response is highly variable with few means to predict which patients will 
benefit. To develop a molecular means of predicting response at diagnosis, we 
examined baseline differences in mutations, DNA methylation, and gene expression 
in 40 CMML patients responsive and resistant to decitabine (DAC). While 
somatic mutations did not differentiate responders and non-responders, we 
were able to identify for the first time 158 differentially methylated regions 
(DMRs) at baseline between responders and non-responders using next-generation 
sequencing. These DMRs were primarily localized to non-promoter regions and 
overlapped with distal regulatory enhancers. Using the methylation profiles, 
we developed an epigenetic classifier that accurately predicted DAC response 
at the time of diagnosis. We also found 53 differentially expressed genes 
between responders and non-responders. Genes up-regulated in responders were 
enriched in the cell cycle, potentially contributing to effective DAC 
incorporation. Two chemokines overexpressed in non-responders -- CXCL4 and 
CXCL7 -- were able to block the effect of DAC on normal CD34+ and primary 
CMML cells in vitro, suggesting their up-regulation contributes to primary 
DAC resistance. Overall design: mRNA profiling in bone marrow mononuclear 
cells (BM MNC) from 14 CMML patients (8 decitabine responders vs. 
6 non-responders).

## PRJNA312611

**RNA-seq of HL60 cells treated with epigenetic therapy**

RNA-seq was performed after HL60 cells were treated with S2101, UNC0638, 
GSK343, depsipeptide alone or in combination with decitabine Overall design: 
Biological triplicates were performed for a total of 30 samples. 
Fold change of each gene was calculated by comparing change in expression 
after inhibitor treatment to expression in the control samples

## PRJNA316529

**Potential of epigenetic therapy of acute myeloid leukemia using 5-aza-2- deoxycytidine (decitabine) and 3-deazaneplanocin-A**

Aberrant DNA methylation that silences tumor suppressor genes occurs frequently 
in patients with acute myeloid leukemia (AML). Treatment of AML patients with 
the inhibitor of DNA methylation, 5-aza-2-deoxycytidine (5-AZA-CdR) can induce 
complete remissions, but most patients will relapse. The clinical efficacy of 
5-AZA-CdR may be influenced by its limited capacity to activate tumor suppressor 
genes silenced by methylation of lysine 27 histone H3 (H3K27) by EZH2. In 
order to overcome this limitation, we investigated previously the antileukemic 
action of 5-AZA-CdR in combination with the EZH2 inhibitor, 3-deazaneplanocin 
A (DZNep) on HL-60 AML cells. We observed a remarkable synergistic interaction 
against these AML cells for this combination. In this study, we investigated 
in more depth the action of 5-AZA-CdR plus DZNep on gene expression in AML 
cells using RNA sequence analysis

## PRJNA497893

**Analysis of HPV16 E2 host gene expression using TERT immortalized keratinocytes (NOKs) cell lines and RNA-sequencing (human)**

HPV must reprogram host gene expression to promote infection, and HPV proteins 
E6 and E7 contribute to this via targeting of cellular transcription factors 
including p53 and pRb respectively. The HPV16 E2 protein regulated host gene 
expression in U2OS cells and in this study we extend these observations into 
NOKs that are capable of suporting late stages of the HPV16 life cycle. We 
observed repression of innate immune genes by E2 that are also repressed by 
the intact HPV16 genome in NOKs. There was a highly significant overlap of the 
E2 regulated genes with those regulated by the intact HPV16 genome in the same 
cell type. siRNA targeting of E2 reversed repression of E2 targed genes. The 
ability of E2 to repress innate immune genes was confirmed in an ano-genital 
immortalized keratinocyte cell line, N/Tert-1. Overall design: In total, 8 
samples are analyzed. Two biological replicates of NOK controls, two biological 
replicates of NOKs containing HPV16 full genome, and two different NOKs cell 
lines overexpressing the HPV16 E2 protein.

## PRJNA607909

**Epigenetic therapy strengthens immune synapses to potentiate susceptibility of lung cancer cells to adoptive γδ T cell thera**

Purpose: Epigenetic drug DNA methyltransferase inhibitors (DNMTis) are 
immunomodulating agents that may sensitize cancer cells to immune cell attack. 
However, it is still unclear how DNMTi affects cancer cells to the killing by 
immune cells. We therefore try to uncover the alternations of genomic DNA 
methylation, chromatin accessibility and transcriptomes after DNMTi treatment 
of cancer cells. Methods: Six human lung cancer cell lines (A549, H1299, 
CL1-0, CL1-5, H1792, PC9) treated without (Mock) or with 100 nM decitabine 
(DAC) for 72 hours followed by growing in drug-free medium for three days were 
subjected to mRNA-seq and Omni-ATAC-seq analyses. Results: mRNA-seq and 
ATAC-seq analyses reveal upregulation of actin and intermediate filament, and 
downregulation of microtubule modules by decitabine in human lung cancer cell 
lines. Conclusion: Decitabine primes lung cancer cells to γδ T cell-mediad 
killing through regulating cytoskeleton associated genes potentially involved 
in immune synapse formation. Overall design: mRNA-seq and ATAC-seq analyses 
of control (Mock) and decitabine (DAC) treated human lung cancer cell lines

## PRJNA649432

**Aging of preleukemic thymocytes drives CpG island hypermethylation in T-cell acute lymphoblastic leukemia**

Cancer cells display DNA hypermethylation at specific CpG islands in comparison 
to their normal healthy counterparts, but the mechanism that drives this 
so-called CpG island methylator phenotype (CIMP) remains poorly understood. 
Here, we show that CpG island methylation in human T-cell acute lymphoblastic 
leukemia (T-ALL) mainly occurs at promoters of PRC2 target genes that are not 
expressed in normal or malignant T-cells and which display a reciprocal 
association with H3K27me3 binding. In addition, we found that this aberrant 
methylation profile shows a strong correlation with the epigenetic age of the 
leukemic T cells and elucidate that a similar CpG island methylation signature 
is gradually established in aging pre-leukemic thymocytes from CD2-Lmo2 
transgenic mice. Finally, we unexpectedly uncover that this age-related CpG 
island hypermethylation signature is completely resistant to the FDA-approved 
hypomethylating agent Decitabine. Altogether, our work demonstrates that DNA 
methylation reflects the epigenetic history of leukemic T cells and suggests 
that methylation-based subtypes of human T-ALL have followed a different 
trajectory towards T-cell transformation, possibly mediated by differences in 
the self-renewing capacity of the putative T-ALL cell-of-origin. Given that 
the concept of preleukemic thymocytes has only been reported in T-ALL mouse 
models so far, we here provide, for the first time, conceptual evidence that a 
pre-leukemic phase might also be involved in the pathogenesis of the human 
disease. Overall design: RNAseq of T-ALL cell lines treated with Decitabine 
or vehicle control

## PRJNA265187

**An siRNA screen identifies CHD4 as a target for epigenetic therapy**

AIM: We performed RNA-sequencing experiments seeking genes whose expression 
changed after siCHD4 treatment. METHOD: We generated cDNA libraries from RNA 
purified from SW48 treated by siNegative control, siCHD4, siNegative control 
plus siDNMT1, siCHD4 plus siDNMT1, siCHD4 plus Decitabine and siNegative 
control plus Decitabine. We sequenced the cDNA libraries generating single 
end 50 bp reads on the illumina HiSeq 2500 platform. Sequencing reads were 
aligned to the human genome hg19 using TopHat RESULTS: We were able to map 
23600 genes. 284 genes in siCHD4, 289 genes in siDNMT1, 309 genes in DAC, 
1090 genes in siCHD4 plus siDNMT1 and 468 genes in siCHD4 plus siDNMT1 were 
significantly upregulated at least two fold compared to siNegative control. 
Overall design: We generated cDNA libraries from RNA purified from SW48 
treated by siNegative control, siCHD4, siNegative control plus siDNMT1, 
siCHD4 plus siDNMT1, siCHD4 plus Decitabine and siNegative control plus 
Decitabine

## PRJNA649431

**Aging of preleukemic thymocytes drives CpG island hypermethylation in T-cell acute lymphoblastic leukemia**

Cancer cells display DNA hypermethylation at specific CpG islands in comparison 
to their normal healthy counterparts, but the mechanism that drives this 
so-called CpG island methylator phenotype (CIMP) remains poorly understood. 
Here, we show that CpG island methylation in human T-cell acute lymphoblastic 
leukemia (T-ALL) mainly occurs at promoters of PRC2 target genes that are not 
expressed in normal or malignant T-cells and which display a reciprocal 
association with H3K27me3 binding. In addition, we found that this aberrant 
methylation profile shows a strong correlation with the epigenetic age of the 
leukemic T cells and elucidate that a similar CpG island methylation signature 
is gradually established in aging pre-leukemic thymocytes from CD2-Lmo2 
transgenic mice. Finally, we unexpectedly uncover that this age-related CpG 
island hypermethylation signature is completely resistant to the FDA-approved 
hypomethylating agent Decitabine. Altogether, our work demonstrates that DNA 
methylation reflects the epigenetic history of leukemic T cells and suggests 
that methylation-based subtypes of human T-ALL have followed a different 
trajectory towards T-cell transformation, possibly mediated by differences in 
the self-renewing capacity of the putative T-ALL cell-of-origin. Given that the 
concept of preleukemic thymocytes has only been reported in T-ALL mouse models 
so far, we here provide, for the first time, conceptual evidence that a 
pre-leukemic phase might also be involved in the pathogenesis of the human 
disease. Overall design: RNAseq of T-ALL patient derived xenografts treated with 
Decitabine or vehicle control

## PRJNA318019

**HLA peptides derived from tumor antigens induced by inhibition of DNA methylation for development of drug-facilitated immunotherapy**

Treatment of cancer cells with anti-cancer drugs often fails to achieve complete 
remission. Yet, such drug treatments may induce alteration in the tumors gene 
exprepression patterns, including those of Cancer/Testis Antigens (CTA).
The degradation products of such antigens can be presented as HLA peptides 
on the surface of the tumor cells and be developed into anti-cancer 
immunotherapeutics. For example, the DNA methyl transferase inhibitor, 
5-aza-2-deoxycytidine (Decitabine) has limited anti-tumor efficacy, yet it 
induces the expression of many genes, including CTAs that are normally silenced 
in the healthy adult tissues. In this study, the presentation of many new HLA 
peptides derived from CTAs and induced by Decitabine was demonstrated in three 
human Glioblastoma cell lines. Such presentation of CTA-derived HLA peptides 
can be exploited for development of new treatment modalities, combining drug 
treatment with anti-CTA targeted immunotherapy. The Decitabine-induced HLA 
peptidomes include many CTAs that are not normally detected in healthy tissues 
or in cancer cells, unless treated with the drug. In addition, the study 
included large-scale analyses of the simultaneous effects of Decitabine on the 
transcriptomes, proteomes and HLA peptidomes of the human Glioblastoma cells. 
It demonstrates the poor correlations between these three levels of gene 
expression, both in their total levels and in their response to the drug. 
Overall design: The transcriptomes, proteomes and HLA peptidomes of the U-87,
T98G and LNT-229 GBM human cell lines were analyzed before and after 
treatment with Decitabine. Overall, the RNA-Seq transcriptome analyses resulted 
in the identification of above 26000 transcripts, the proteome analyses 
identified about 7500 proteins and the HLA class I peptidome analyses resulted 
in above 25000 identified HLA peptides. Two biological repetitions of the 
transcriptome, three of the proteome and three of the HLA peptidome were 
performed with each of the cell lines and treatment, resulting in highly 
reproducible datasets

## PRJNA391270

**Stereotyped p53 binding tuned by chromatin accessibility**

We accessed the cell type specificity of p53 by directly measuring DNA binding 
in twelve cell lines in response to ionizing radiation. We find that that vast 
majority of binding sites are occupied across all cells lines uniformly, in 
contrast to p53 regulated gene expression which shows great diversity in the 
same context. We further identify a subset of p53 binding sites that are more 
restricted, appearing in one or a few cell lines. We find that chromatin 
accessibility explains much of these differential binding events. Overall 
design: p53 DNA binding was measured in twelve cell lines after IR by ChIPseq. 
The transcriptional response of each of the twelve lines was also measured by 
mRNAseq. To explore the influence of chromatin accessability p53 binding was 
compared to ATAC-seq data from two cell lines

## PRJNA596778

**DNA Demethylation promotes ERV expression and activation of immune signaling in renal cell cancer cells**

Recently, we reported that expression of endogenous retroviruses (ERVs), a 
class of transposable element, is associated with response to immune checkpoint 
blockade (ICB) in renal cell carcinoma (RCC). Aberrant expression of ERVs can 
activate host antiviral responses, as well as produce neoantigens. ERV expression 
is repressed by DNA methylation and can be activated by DNA hypomethylating agents. 
Here, we investigate whether Decitabine, a DNA hypomethylating agent, 
can activate ERV expression and host antiviral defenses in RCC to potentially 
enhance response to ICB. Decitabine induced expression of ERV3-2 and ERV4700 
in RCC cells lines, which was accompanied by activation of host antiviral 
defense genes and increased secretion of inflammatory cytokines. We 
validated this effect in primary human RCC cell lines. Knockout of RIG-I and 
MDA5 dsRNA sensors attenuated activation of antiviral responses associated 
with Decitabine treatment, and RIG-I and MDA5 immunoprecipitations showed 
increased ERV binding in RCC cells treated with Decitabine. Bioinformatic 
analysis of RNAseq data showed the Decitabine-induced gene signature could be 
associated with increased CD8 infiltration and response to ICB. Conditioned 
media from Decitabine treated RCC cells was capable of inducing host anti-viral 
defense in naïve RCC cells and could modestly improve activation of T-cells
from healthy donors. Further, conditioned media from RCC cells treated with 
Decitabine significantly enhanced T-cell migration. In a small retrospective 
cohort of metastatic RCC patients treated with single-agent PD-1/PD-L1 blockade, 
activation of some host antiviral defense genes was significantly higher in 
responders compared with nonresponders. Thus, modulation of ERV expression by 
Decitabine to activate host antiviral defenses could represent a novel 
therapeutic strategy to enhance RCC patient response to immune checkpoint blockade. 
Overall design: RNAseq profiles in HKC and 786-0 kidney cells treated with DMSO 
or Decitabine (100nM or 300nM)

## PRJNA657045

**Low-dose decitabine priming endows CAR T cells with enhanced and persistent anti-tumor potential by epigenetic reprogramming in vivo and in vitro**

Chimeric antigen receptor T (CAR T) cells targeting CD19 have achieved 
breakthroughs in the treatment of haematological malignancies, but many clinical 
studies have also shown that a proportion of patients relapse after remission. 
In this study, we designed CART treatment by DNMTi (dCART) inhibitor and found 
that dCAR T cells retained relatively potent antitumour activity compared with 
CAR T cell upon target antigen recognition. It may be associated with memory 
and anti-exhausion. Our transcriptional analysis underscores the potential of 
dCAR T cells. Overall design: In order to assess the different phenotypic and 
functional patterns of CARs between dCAR T and CAR T, we compared the 
genome-wide transcriptional profiles of dCAR and CAR T cell before and after 
cocultured with Raji cells for 24 h.

