## PRJNA315294

**Identification of long noncoding RNAs regulated by p53**

Analyize the transcriptpme regulated by p53 in 3 pairs of isogenic p53WT and 
p53KO colorectal cancer cell lines untreated or treated with Doxorubicin. 
The hypothsis of this study is that p53 can regulate the expression of subset 
of lncRNAs to mediate its biological functions. Overall design: Total RNA was 
isolated from 3 pairs of isogenic colon cancer cell lines; HCT116-p53_WT and 
p53_KO; RKO-p53_WT and p53_KO: and SW48-p53_WT and p53_KO. Cells were untreated 
or treated with Doxorubicin (DOXO) at a final concentration of 300 nM for 16 hours

## PRJNA387040 

**Expression profiling of etoposide-induced senescent cells**

We report here the RNA-seq data of proliferating IMR90 cells and etoposide-induced 
senescent cells with control or cGAS knockdown Overall design: RNA-seq in 
primary IMR90 cells, two independent biological replicates, sequenced in Illumina 
NextSeq 500 platform. 

## PRJNA417314

**Transcriptome profiling of colorectal adenocarcinoma HT-29 cells after 5-FU treatment**

We performed RNA-Seq analysis of transcriptomic alterations in HT-29 colon 
cancer cells treated with 20 µM 5-FU for 72 h, the time period roughly
equivalent to three consecutive cell doublings. We expected that the high 
dose of the drug, which corresponds to 10 x IC50 and kills nearly 80% of 
cells, would selectively favor the survival of cells that acquire a transient 
resistance to 5-FU.

We identified lincRNAs LINC00973, LINC00941, CCAT1, and CASC19 as the most 
upregulated in 5-FU treated HT-29 cells. These results were further 
validated by RT-qPCR in HT-29 cells harvested after 72 h and 144 h of treatment 
with 2 µM 5-FU or 2 µM OXP (corresponding to the IC50 dose) or 5 µM I
RI (one third of its IC50).

We also added lincRNA BCAR4 to the analysis, which is undetectable by RNA-Seq,
but according to our RT-qPCR data is upregulated in 57% of colorectal tumors. 
Obtained results demonstrated that all five lincRNAs are upregulated by the 
treatment with all three drugs, indicating that monitoring by RT-qPCR of just 
two of the lincRNAs (LINC00973 and BCAR4) provides a robust instrument for 
distinguishing of untreated and treated cells.

## PRJNA335388  

**RNA sequencing of MDA-MB231 and U2OS cancer cell lines exposed to alkylating methylsufonate (MMS) and classical chemotherapeutics**

Understanding the mechanisms by which cells respond to chemotherapeutics is key
to identifying means to improve therapy effiicacy while reducing systemic 
toxicity of these widely used classes of drugs. While determining the role 
of NRF2-GSH and ER stress in cells exposed to alkylating compounds such as 
methyl-methanesulfonate (MMS), we asked if these pathways could also be a 
general cell damage response relevant to other clinically used chemotherapeutics 
or if it is an alkylation specific response. With this intent, we performed 
RNA sequencing of MDA-MB231 breast cancer and U2OS osteosarcoma cells lines 
treated for 8 hours with a topoisomerase II inhibitor etoposide (20 µM), the
antimitotic beta-tubulin-interacting drug paclitaxel (0.2 µM), doxorubicin
(1 µM) and compared to MMS (40 µg/mL) treated cells. Doses represent IC50 lev
el after 72 hours exposure. We observed that even though non-alkylating drugs,
especially etoposide, caused an increase in the mRNA expression of some NRF2 
and ER stress signaling markers, the number and magnitude of upregulation of 
genes markers in either pathway was more pronounced in alkylation treatments 
compared to other drugs. This indicates that alterations in NRF2 and ER stress 
pathways could be more likely associated with differential sensitivity to 
alkylating chemotherapies. Overall design: MDA-MB231 breast cancer and U2OS 
osteosarcoma cells lines were treated with the 72 h IC50  dose of etoposide
(20 µM), paclitaxel (0.2 µM),  doxorubicin (1 µM) or  MMS (40 µg/mL) for 8
 h, and RNA was extracted and analyzed.

## PRJNA392914

**Repression of stress-induced LINE-1 expression protects cancer cell populations from lethal drug-exposures**

DTP heterochromatin in genomic repeat regions protects the population from 
drug-induced death. Overall design: Investigation of RNA expression changes in 
heterogenous parental populations as compared to drug treated and drugtolerant 
populations. Also included are TSA treated parental and drug-tolerant populations

## PRJNA260377

## PRJNA312611

## PRJNA316529

## PRJNA497893

## PRJNA607909

## PRJNA649432

## PRJNA265187
