---
title: "cdk12-degraders"
output: html_document
date: '2022-06-24'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load Libraries

```{r}
suppressPackageStartupMessages(library(SummarizedExperiment))
suppressPackageStartupMessages(library(here))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(PCAtools))
suppressPackageStartupMessages(library(patchwork))
suppressPackageStartupMessages(library(coriell))
suppressPackageStartupMessages(library(ComplexHeatmap))
suppressPackageStartupMessages(library(fgsea))

theme_set(
  theme_light() +
  theme(legend.position = "bottom",
        axis.text.x = element_text(angle = 45, hjust = 1),
        plot.title = element_text(size = 16, face = "bold"),
        strip.text = element_text(size = 12, face = "bold"))
)
```

---

## PRJNA679028 Analysis

```{r}
# Read in results SE object
PRJNA679028 <- readRDS(here("results", "PRJNA679028", "rds-files", "se.rds"))

# Extract the differential expression data
de <- metadata(PRJNA679028)$de

# Extract DE for each feature type
de_genes <- de[!feature_id %like% ".*\\..*\\..*"]
de_repeats <- de[feature_id %like% ".*\\..*\\..*"]
```

Create plots for each contrast

```{r}
v_genes <- lapply(split(de_genes, by="contrast"), function(x) plot_volcano(x, fdr = 0.05, lab_size = 6))
m_genes <- lapply(split(de_genes, by="contrast"), function(x) plot_md(x, fdr = 0.05, lab_size = 6))
v_repeats <- lapply(split(de_repeats, by="contrast"), function(x) plot_volcano(x, fdr = 0.05, lab_size = 6))
m_repeats <- lapply(split(de_repeats, by="contrast"), function(x) plot_md(x, fdr = 0.05, lab_size = 6))

# Give the plots better names and format
plot_titles <- c("BSJ_4_23 vs BSJ_4_23NC", "THZ531 vs DMSO", "BSJ_4_23NC vs DMSO")
fmt_plot <- function(x, y) x + ggtitle(y) + theme(plot.title = element_text(size = 18, face="bold"))
v_genes <- mapply(fmt_plot, x=v_genes, y=plot_titles, SIMPLIFY = FALSE)
m_genes <- mapply(fmt_plot, x=m_genes, y=plot_titles, SIMPLIFY = FALSE)
v_repeats <- mapply(fmt_plot, x=v_repeats, y=plot_titles, SIMPLIFY = FALSE)
m_repeats <- mapply(fmt_plot, x=m_repeats, y=plot_titles, SIMPLIFY = FALSE)
```

Create patchwork and save

```{r}
# Create patchwork of results
(v_genes[[1]] | v_genes[[2]]) / (m_genes[[1]] | m_genes[[2]]) +
  plot_annotation(title = "Protein Coding Genes") +
  plot_layout(guides = "collect") &
  theme(plot.title = element_text(size=18, face="bold"))
ggsave(here("results", "PRJNA679028", "figures", "de-genes.png"), width = 12, height = 8)

(v_repeats[[1]] | v_repeats[[2]]) / (m_repeats[[1]] | m_repeats[[2]]) +
  plot_annotation(title = "Repetitive Elements") +
  plot_layout(guides = "collect") &
  theme(plot.title = element_text(size=18, face="bold"))
ggsave(here("results", "PRJNA679028", "figures", "de-repeats.png"), width = 12, height = 8)
```

Do they share the same features?

```{r}
# Cast data wider
de_genes.w <- dcast(de_genes, feature_id ~ contrast, value.var = "logFC")
de_repeats.w <-  dcast(de_repeats, feature_id ~ contrast, value.var = "logFC")
```

Plot the DE for Genes and Repeats

```{r}
# Genes
p1 <- ggplot(de_genes.w, aes(x = T_ALL.Jurkat_cells.BSJ_4_23_vs_T_ALL.Jurkat_cells.BSJ_4_23NC, 
                 y = T_ALL.Jurkat_cells.THZ531_vs_T_ALL.Jurkat_cells.DMSO)) +
  geom_hline(yintercept = 0, linetype = 3) +
  geom_vline(xintercept = 0, linetype = 3) +
  geom_point(alpha = 0.1) +
  geom_point(
    data = de_genes.w[feature_id == "CDK12"], 
    aes(x = T_ALL.Jurkat_cells.BSJ_4_23_vs_T_ALL.Jurkat_cells.BSJ_4_23NC,
        y = T_ALL.Jurkat_cells.THZ531_vs_T_ALL.Jurkat_cells.DMSO), size = 2, color = "red") +
  geom_abline(slope = 1, intercept = 0, linetype = 2, color = "blue") +
  scale_x_continuous(limits = c(-10, 10)) +
  scale_y_continuous(limits = c(-10, 10)) +
  labs(title = "Protein Coding Genes",
       subtitle = "CDK12 shown in Red",
       x = "logFC BSJ-4-23",
       y = "logFC THZ531")

# Repeats
p2 <- ggplot(de_repeats.w, aes(x = T_ALL.Jurkat_cells.BSJ_4_23_vs_T_ALL.Jurkat_cells.BSJ_4_23NC, 
                 y = T_ALL.Jurkat_cells.THZ531_vs_T_ALL.Jurkat_cells.DMSO)) +
  geom_hline(yintercept = 0, linetype = 3) +
  geom_vline(xintercept = 0, linetype = 3) +
  geom_point(alpha = 0.5) +
  geom_abline(slope = 1, intercept = 0, linetype = 2, color = "blue") +
  scale_x_continuous(limits = c(-10, 10)) +
  scale_y_continuous(limits = c(-10, 10)) +
  labs(title = "Repetitive Elements",
       x = "logFC BSJ-4-23",
       y = "logFC THZ531")

# Patch together
(p1 | p2) +
  plot_layout(guides = "collect") +
  plot_annotation(title = "THZ531 and BSJ-4-23 Correlation of logFC Values") &
  theme(plot.title = element_text(size = 18, face = "bold"))
ggsave(here("results", "PRJNA679028", "figures", "THZ531_vs_BSJ_4_23.png"), width = 12, height = 6)
```

How much do the DE features overlap between drugs?

```{r}
thz_up.genes <- de_genes[contrast == "T_ALL.Jurkat_cells.THZ531_vs_T_ALL.Jurkat_cells.DMSO" & 
                           FDR < 0.05 & logFC > 0, feature_id] 
thz_down.genes <- de_genes[contrast == "T_ALL.Jurkat_cells.THZ531_vs_T_ALL.Jurkat_cells.DMSO" & 
                           FDR < 0.05 & logFC < 0, feature_id] 
bsj_up.genes <- de_genes[contrast == "T_ALL.Jurkat_cells.BSJ_4_23_vs_T_ALL.Jurkat_cells.BSJ_4_23NC" & 
                           FDR < 0.05 & logFC > 0, feature_id]
bsj_down.genes <- de_genes[contrast == "T_ALL.Jurkat_cells.BSJ_4_23_vs_T_ALL.Jurkat_cells.BSJ_4_23NC" & 
                           FDR < 0.05 & logFC < 0, feature_id]
l_genes <- list_to_matrix(list(THZ531.up = thz_up.genes, THZ531.down = thz_down.genes, 
                               BSJ.up = bsj_up.genes, BSJ.down = bsj_down.genes))
m_genes <- make_comb_mat(l_genes, mode = "intersect")
m_genes <- m_genes[comb_degree(m_genes) == 2]
```

Upset plot of shared genes

```{r}
png(here("results", "PRJNA679028", "figures", "upset-genes.png"), width = 8, height = 6, res = 300, units = "in")
UpSet(m_genes, 
      top_annotation = upset_top_annotation(m_genes, add_numbers = TRUE, numbers_gp = gpar(fontsize = 14)), 
      right_annotation = upset_right_annotation(m_genes, add_numbers = TRUE, numbers_gp = gpar(fontsize = 8)),
      comb_order = order(comb_size(m_genes)),
      comb_col = c("firebrick", "purple4", "steelblue")
      )
dev.off()
```

```{r}
thz_up.repeats <- de_repeats[contrast == "T_ALL.Jurkat_cells.THZ531_vs_T_ALL.Jurkat_cells.DMSO" & 
                               FDR < 0.05 & logFC > 0, feature_id] 
thz_down.repeats <- de_repeats[contrast == "T_ALL.Jurkat_cells.THZ531_vs_T_ALL.Jurkat_cells.DMSO" & 
                                 FDR < 0.05 & logFC < 0, feature_id] 
bsj_up.repeats <- de_repeats[contrast == "T_ALL.Jurkat_cells.BSJ_4_23_vs_T_ALL.Jurkat_cells.BSJ_4_23NC" & 
                               FDR < 0.05 & logFC > 0, feature_id]
bsj_down.repeats <- de_repeats[contrast == "T_ALL.Jurkat_cells.BSJ_4_23_vs_T_ALL.Jurkat_cells.BSJ_4_23NC" & 
                                 FDR < 0.05 & logFC < 0, feature_id]
l_repeats <- list_to_matrix(list(THZ531.up = thz_up.repeats, THZ531.down = thz_down.repeats, 
                                 BSJ.up = bsj_up.repeats, BSJ.down = bsj_down.repeats))
m_repeats <- make_comb_mat(l_repeats, mode = "intersect")
m_repeats <- m_repeats[comb_degree(m_repeats) == 2]
```

Upset plot of shared repeats

```{r}
png(here("results", "PRJNA679028", "figures", "upset-repeats.png"), width = 8, height = 6, res = 300, units = "in")
UpSet(m_repeats, 
      top_annotation = upset_top_annotation(m_repeats, add_numbers = TRUE, numbers_gp = gpar(fontsize = 14)), 
      right_annotation = upset_right_annotation(m_repeats, add_numbers = TRUE, numbers_gp = gpar(fontsize = 8)),
      comb_order = order(comb_size(m_repeats)),
      comb_col = c("firebrick", "purple4", "steelblue")
      )
dev.off()
```

What are the overlapping REs?

```{r}
shared_res <- sort(intersect(thz_up.repeats, bsj_up.repeats))
shared_dt <- data.table(elem = shared_res)
shared_dt[, c("class", "family", "subfamily") := tstrsplit(elem, ".", fixed = TRUE)]

shared_dt[, .N, by = .(class, family)][order(N, decreasing = TRUE)]
```

## PCA

```{r}
logcounts <- assay(PRJNA679028, "logcounts")
metadata <- colData(PRJNA679028)
metadata$treatment <- gsub("T_ALL.Jurkat_cells.", "", metadata$group, fixed = TRUE)
metadata$treatment <- factor(metadata$treatment, levels = c("DMSO", "BSJ_4_23NC", "THZ531", "BSJ_4_23"))
logcounts.genes <- logcounts[!grepl(".*\\..*\\..*", rownames(logcounts)), ]
logcounts.repeats <- logcounts[grepl(".*\\..*\\..*", rownames(logcounts)), ]
pca.genes <- pca(logcounts.genes, metadata, removeVar = 0.5)
pca.repeats <- pca(logcounts.repeats, metadata, removeVar = 0.25)
```

PCA biplots

```{r}
b1 <- biplot(
  pca.genes,
  colby = "treatment",
  lab = NULL,
  vline = 0,
  vlineType = 2,
  hline = 0,
  hlineType = 2,
  pointSize = 5,
  legendPosition = "bottom",
  title = "Protein Coding Genes"
  ) + scale_color_brewer(palette = "Paired")

b2 <- biplot(
  pca.repeats,
  colby = "treatment",
  lab = NULL,
  vline = 0,
  vlineType = 2,
  hline = 0,
  hlineType = 2,
  pointSize = 5,
  legendPosition = "bottom",
  title = "Repetitive Elements"
  ) + scale_color_brewer(palette = "Paired")

(b1 | b2) +
  plot_layout(guides = "collect") &
  theme(legend.position = "bottom")
ggsave(here("results", "PRJNA679028", "figures", "pca.png"), width = 11, height = 7)
```

## GSEA analysis

They claim down-regulation of DNA repair genes

```{r}
set.seed(123)

# Use same doible strand break repair as in the paper
pathways <- list("GOBP_DOUBLE_STRAND_BREAK_REPAIR" = c("PARP2", "PARP3", "ACTR2", "RAD50", "PSMD14", "MAD2L2", "CIB1", "KAT5", "RAD51AP1", "POLQ", "RUVBL2", "BRD8", "MORF4L1", "TENT4A", "HSF2BP", "CHEK1", "DMC1", "CHEK2", "SHLD3", "PNKP", "CDCA5", "HELQ", "CGAS", "RMI2", "ERCC8", "SFR1", "ZSWIM7", "SWSAP1", "ANKLE1", "HUS1B", "PARP1", "EME1", "KASH5", "SHLD1", "DTX3L", "PPP4R2", "KHDC3L", "ESCO2", "MCMDC2", "USP51", "DDX1", "RNF168", "DDX11", "DNA2", "DNTT", "ARID2", "EME2", "NSMCE1", "GGN", "APLF", "ERCC1", "ERCC4", "ERCC5", "ERCC6", "EYA1", "EYA3", "FANCD2", "FANCB", "FEN1", "ZNF365", "FAN1", "KDM2A", "KDM1A", "FOXM1", "SETX", "POGZ", "SMC5", "SMCHD1", "ZFYVE26", "SPIDR", "SPO11", "TFIP11", "FUS", "MMS22L", "RNF169", "MCM9", "MEIOB", "XRCC6", "RAD54B", "NIPBL", "IFFO1", "POT1", "SAMHD1", "EPC2", "PRPF19", "POLL", "POLM", "MEIOC", "NSMCE2", "PAXX", "SETD2", "BABAM1", "BRD7", "RPA4", "H2AX", "HMGB1", "HMGB2", "C14orf39", "HSF1", "HUS1", "GEN1", "POLN", "SWI5", "LIG3", "LIG4", "MIR221", "MCM2", "MCM3", "MCM4", "MCM5", "MCM6", "MCM7", "MGMT", "MLH1", "MRE11", "OOEP", "MSH2", "NBN", "ATM", "TONSL", "EID3", "OGG1", "SMARCAL1", "KMT5B", "MRNIP", "ZBTB7A", "UBR5", "ACTL6B", "RNF138", "SIRT7", "SIRT6", "TDP2", "PIAS4", "GINS2", "UIMC1", "RTEL1", "PML", "POLA1", "POLB", "TERF2IP", "ATR", "SHLD2", "ING3", "INO80", "NSMCE4A", "MBTD1", "APTX", "SLX1A", "HPF1", "PARPBP", "ZCWPW1", "RADX", "WDR70", "WRAP53", "RFWD3", "RIF1", "PBRM1", "EXD2", "MRGBP", "PHF10", "PPP4C", "AP5S1", "PPP5C", "OTUB1", "KDM4D", "SLF2", "TDP1", "PRKDC", "DMAP1", "TEX15", "NSMCE3", "FMN2", "UBQLN4", "SPIRE1", "SMARCAD1", "HMCES", "PRDM9", "PPP4R3B", "ARID1B", "WDR48", "EP400", "FANCM", "INIP", "RAD21", "RAD51", "RAD51C", "RAD51B", "RAD51D", "RAD52", "RBBP8", "RECQL", "DPF2", "REV3L", "ACTB", "BCL7A", "RPA1", "RPA2", "RPA3", "FIGNL1", "BLM", "MAGEF1", "SETMAR", "SFPQ", "MMS19", "RAD21L1", "DCLRE1C", "NUCKS1", "MEAF6", "DCLRE1B", "NABP1", "INTS3", "SMARCA2", "SMARCA4", "SMARCB1", "SMARCC1", "SMARCC2", "SMARCD1", "SMARCD2", "SMARCD3", "SMARCE1", "BRCA1", "BRCA2", "VPS72", "TOP3A", "TP53", "TP53BP1", "TWIST1", "RAD51AP2", "UBE2N", "UBE2V2", "UVRAG", "VCP", "WAS", "NSD2", "WRN", "XRCC1", "XRCC2", "XRCC3", "XRCC4", "XRCC5", "YY1", "OTUB2", "CYREN", "AUNIP", "SLX1B", "NABP2", "DEK", "BRME1", "BRCC3", "SMC6", "PALB2", "NHEJ1", "ACTR5", "RMI1", "MUS81", "KLHL15", "EPC1", "YEATS4", "HMGA2", "DPF3", "DPF1", "ARID1A", "TRRAP", "CDC7", "CDC45", "PARP9", "BRIP1", "ATRIP", "ABRAXAS1", "SLF1", "GINS4", "NUDT16L1", "RAD54L", "SLX4", "SPIRE2", "MCM8", "HDGFL2", "KMT5C", "FBH1", "TNKS1BP1", "ACTL6A", "RUVBL1", "TIMELESS", "RNF8", "MTA1", "ATP23", "BCL7C", "BCL7B", "HELB", "TRIP13", "TRIP12", "ACTR8", "RECQL5", "RECQL4", "BABAM2", "MORF4L2", "MARF1", "AP5Z1", "DCLRE1A", "REC8"))

# Rank by Pvalue and logFC
de_genes[, stat := logFC * -log10(PValue)]

# Extract stats
thz531_stats <- de_genes[contrast == "T_ALL.Jurkat_cells.THZ531_vs_T_ALL.Jurkat_cells.DMSO", stat]
bsj23_stats <- de_genes[contrast == "T_ALL.Jurkat_cells.BSJ_4_23_vs_T_ALL.Jurkat_cells.BSJ_4_23NC", stat]
names(thz531_stats) <- de_genes[contrast == "T_ALL.Jurkat_cells.THZ531_vs_T_ALL.Jurkat_cells.DMSO", feature_id]
names(bsj23_stats) <- de_genes[contrast == "T_ALL.Jurkat_cells.BSJ_4_23_vs_T_ALL.Jurkat_cells.BSJ_4_23NC", feature_id]
```

Perform GSEA

```{r}
res.thz531 <- fgsea(
    pathways = pathways, 
    stats = thz531_stats,
    nPermSimple = 1e5,
    eps = 0.0
  )

res.bsj23 <- fgsea(
    pathways = pathways, 
    stats = bsj23_stats,
    nPermSimple = 1e5,
    eps = 0.0
  )
```

Plot the enrichment

```{r}
e1 <- plotEnrichment(pathway = pathways[["GOBP_DOUBLE_STRAND_BREAK_REPAIR"]], stats = thz531_stats) +
  labs(title = "THZ531 - DOUBLE STRAND BREAK REPAIR", subtitle = "PValue = 0.087; ES = 0.46")

e2 <- plotEnrichment(pathway = pathways[["GOBP_DOUBLE_STRAND_BREAK_REPAIR"]], stats = bsj23_stats) +
  labs(title = "BSJ-23 - DOUBLE STRAND BREAK REPAIR", subtitle = "PValue = 0.46; ES = -0.418")

(e1 | e2)
ggsave(here("results", "PRJNA679028", "figures", "dna-damage-enrichment.png"), width = 12, height = 5)
```

Heatmap of the DNA Damage response genes

```{r}
m <- logcounts[rownames(logcounts) %chin% pathways[["GOBP_DOUBLE_STRAND_BREAK_REPAIR"]], ]
coldata <- data.frame(metadata[, "treatment", drop=FALSE])
trt_pal <- RColorBrewer::brewer.pal(4, "Paired")
names(trt_pal) <- levels(metadata$treatment)

quickmap(m, annotation_col = coldata, main = "Double Strand Break Repair", 
         show_rownames = FALSE, show_colnames = FALSE, fontsize_row = 6, 
         annotation_colors = list(treatment = trt_pal),
         filename = here("results", "PRJNA679028", "figures", "dna-damage-heatmap.png"))
```

## Comparison to TCGA PRAD REs

```{r eval=FALSE, include=FALSE}
metaVote <- fread(here("doc", "TCGA-PRAD", "differential-expression_metaVote.tsv"))
metaCombine <- fread(here("doc", "TCGA-PRAD", "differential-expression_metaCombine.tsv"))

# Extract the significant REs from meta combine method
sig_tcga <- metaCombine[!feature_id %like% "^ENSG" & meta_p < 0.05 & meta_lfc > 0, unique(feature_id)]

# Extract the significant REs from BSJ-4
sig_re <- de_repeats[
  contrast == "T_ALL.Jurkat_cells.BSJ_4_23_vs_T_ALL.Jurkat_cells.BSJ_4_23NC" & FDR < 0.05 & logFC > 0, 
  unique(feature_id)]

m <- make_comb_mat(list_to_matrix(list(CDK12_mut = sig_tcga, BSJ_4 = sig_re)), mode = "intersect")
UpSet(m, 
      top_annotation = upset_top_annotation(m, add_numbers = TRUE, numbers_gp = gpar(fontsize = 14)), 
      right_annotation = upset_right_annotation(m, add_numbers = TRUE, numbers_gp = gpar(fontsize = 8)),
      comb_order = order(comb_size(m)),
      )

library(GeneOverlap)
ov <- newGeneOverlap(sig_tcga, sig_re, genome.size = 1123)
ov <- testGeneOverlap(ov)
```


---

## PRJNA814284 analysis

- Fibroblasts are MPAF
- Seminoma are TCam2
- Embryonal carcinoma are 2102EP

```{r}
# Read in results SE object
PRJNA814284 <- readRDS(here("results", "PRJNA814284", "rds-files", "se.rds"))

# Extract differential expression results
de <- metadata(PRJNA814284)$de

# Extract DE for each feature type
de_genes <- de[!feature_id %like% ".*\\..*\\..*"]
de_repeats <- de[feature_id %like% ".*\\..*\\..*"]
```

Function to create patchworks for all contrasts 

```{r}
# Create a function to patch all plots together given the input lists
patchAll <- function(l, patch_title) {
  plot_names <- tstrsplit(names(l), "\\.")[[2]]
  plot_names <- gsub("_vs.*", "", plot_names)
  plot_names <- gsub("_", " ", plot_names)
  
  # Create a descriptive title for each plot
  l2 <- mapply(function(x, y) x + ggtitle(y), x=l, y=plot_names, SIMPLIFY = FALSE)

  (l2[[1]] | l2[[2]]) / (l2[[3]] | l2[[4]]) / (l2[[5]] | l2[[6]]) / (l2[[7]] | l2[[8]]) +
    plot_annotation(title = patch_title) &
    theme(plot.title = element_text(size = 18, face = "bold"))
}

# Saving patches
savePatch <- function(x, y, z) {
  ggsave(filename = here("results", "PRJNA814284", "figures", paste0(z, "-", y, ".png")), 
         plot = x, 
         width = 10, 
         height = 18)
}

# Monolithic function for creating all plots
plotAllDeResults <- function(dt, feature_type) {
  # Create base plots for each contrast
  dts <- split(dt, by = "contrast")
  message("Creating Plots...")
  v <- lapply(dts, plot_volcano, fdr = 0.05, lab_size = 6)
  m <- lapply(dts, plot_md, fdr = 0.05, x = "AveExpr", lab_size = 6)
  v <- lapply(names(v), function(x) v[[x]] + ggtitle(x))
  m <- lapply(names(m), function(x) m[[x]] + ggtitle(x))
  names(v) <- names(dts)
  names(m) <- names(dts)
  
  # Split by DMSO control since they don't specify which they actually used so create them all
  message("Splitting by concentration...")
  v_low <- v[grepl("0001p", names(v))]
  v_high <- v[grepl("\\.001p", names(v))]
  m_low <- m[grepl("0001p", names(m))]
  m_high <- m[grepl("\\.001p", names(m))]
  
  # Now split by cell type
  message("Splitting by cell type...")
  e_v_low <- v_low[grepl("Embryonal", names(v_low))]
  e_v_high <-  v_high[grepl("Embryonal", names(v_high))]
  e_m_low <- m_low[grepl("Embryonal", names(m_low))]
  e_m_high <-  m_high[grepl("Embryonal", names(m_high))]
  f_v_low <- v_low[grepl("Fibroblasts", names(v_low))]
  f_v_high <-  v_high[grepl("Fibroblasts", names(v_high))]
  f_m_low <- m_low[grepl("Fibroblasts", names(m_low))]
  f_m_high <-  m_high[grepl("Fibroblasts", names(m_high))]
  s_v_low <- v_low[grepl("Seminoma", names(v_low))]
  s_v_high <-  v_high[grepl("Seminoma", names(v_high))]
  s_m_low <- m_low[grepl("Seminoma", names(m_low))]
  s_m_high <-  m_high[grepl("Seminoma", names(m_high))]
  
  # Create patchworks for all
  message("Creating patchworks...")
  p_e_v_low <- patchAll(e_v_low, "2102EP Cell Line; Treatment vs DMSO 0.0001%")
  p_e_v_high <-  patchAll(e_v_high, "2102EP Cell Line; Treatment vs DMSO 0.001%")
  p_e_m_low <- patchAll(e_m_low, "2102EP Cell Line; Treatment vs DMSO 0.0001%")
  p_e_m_high <- patchAll(e_m_high, "2102EP Cell Line; Treatment vs DMSO 0.001%")
  p_f_v_low <- patchAll(f_v_low, "MPAF Cell Line; Treatment vs DMSO 0.0001%")
  p_f_v_high <-  patchAll(f_v_high, "MPAF Cell Line; Treatment vs DMSO 0.001%")
  p_f_m_low <- patchAll(f_m_low, "MPAF Cell Line; Treatment vs DMSO 0.0001%")
  p_f_m_high <- patchAll(f_m_high, "MPAF Cell Line; Treatment vs DMSO 0.001%")
  p_s_v_low <- patchAll(s_v_low, "TCam2 Cell Line; Treatment vs DMSO 0.0001%")
  p_s_v_high <-  patchAll(s_v_high, "TCam2 Cell Line; Treatment vs DMSO 0.001%")
  p_s_m_low <- patchAll(s_m_low, "TCam2 Cell Line; Treatment vs DMSO 0.0001%")
  p_s_m_high <- patchAll(s_m_high, "TCam2 Cell Line; Treatment vs DMSO 0.001%")
  
  # Save all to files
  p_names <- c("2102EP_vs_DMSO_0.0001p", "2102EP_vs_DMSO_0.001p", "MPAF_vs_DMSO_0.0001p", 
               "MPAF_vs_DMSO_0.001p", "TCam2_vs_DMSO_0.0001p", "TCam2_vs_DMSO_0.001p")
  v_list <- list(p_e_v_low, p_e_v_high, p_f_v_low, p_f_v_high, p_s_v_low, p_s_v_high)
  m_list <- list(p_e_m_low, p_e_m_high, p_f_m_low, p_f_m_high, p_s_m_low, p_s_m_high)
  message("Saving patchworks...")
  invisible(mapply(savePatch, x = v_list, y = p_names, z = paste0(feature_type, "_volcano")))
  invisible(mapply(savePatch, x = m_list, y = p_names, z = paste0(feature_type, "_ma")))
}
```

```{r}
# All features
plotAllDeResults(de, feature_type = "all")
plotAllDeResults(de_genes, feature_type = "genes")
plotAllDeResults(de_repeats, feature_type = "repeats")
```

Compare results to published results

Published results are in `GSE198248_DE_files` in doc directory. Since they messed up their files, 
read in without colnames and then set names manually. 

```{r}
files <- list.files(here("doc", "GSE198248_DE_files"), pattern = "*.txtSeq2_DE", full.names = TRUE)
names(files) <- gsub("_DE.txtSeq2_DE", "", basename(files), fixed = TRUE)
published <- rbindlist(lapply(files, fread, skip = 1), idcol = "contrast")

# Fix duplicated names in files
setnames(
  published, 
  old = paste0("V", 1:13), 
  new = c("feature_id", "logCPM", "logFC", "lfcSE", "pvalue", "FDR", "chromosome", "gene_id", 
          "gene_name", "chromosome2", "gene_id2", "gene_name2", "gene_id3")
  )
```

```{r}
# what are the number of de genes per contrast?
published[, summarize_dge(.SD, fdr = 0.05), by = contrast][
  Direction != "Unperturbed"][
  order(N, decreasing = TRUE)]
```

Plots from published results

```{r}
p_dts <- split(published, by = "contrast")
p_v <- lapply(p_dts, plot_volcano, fdr = 0.05, lab_size = 6)
p_m <- lapply(p_dts, plot_md, fdr = 0.05, lab_size = 6)
p_v <- lapply(names(p_v), function(x) p_v[[x]] + ggtitle(x))
p_m <- lapply(names(p_m), function(x) p_m[[x]] + scale_x_log10() + ggtitle(x))
names(p_v) <- names(p_dts)
names(p_m) <- names(p_dts)
```

Create a patchwork for each cell type

```{r}
# Embryonal
e_v <- (p_v[[1]] | p_v[[2]]) / (p_v[[3]] | p_v[[4]]) / (p_v[[5]] | p_v[[6]]) / (p_v[[7]] | p_v[[8]])
e_m <- (p_m[[1]] | p_m[[2]]) / (p_m[[3]] | p_m[[4]]) / (p_m[[5]] | p_m[[6]]) / (p_m[[7]] | p_m[[8]])

# Fibroblasts
f_v <- (p_v[[9]] | p_v[[10]]) / (p_v[[11]] | p_v[[12]]) / (p_v[[13]] | p_v[[14]]) / (p_v[[15]] | p_v[[16]])
f_m <- (p_m[[9]] | p_m[[10]]) / (p_m[[11]] | p_m[[12]]) / (p_m[[13]] | p_m[[14]]) / (p_m[[15]] | p_m[[16]])

# Seminoma
s_v <- (p_v[[17]] | p_v[[18]]) / (p_v[[19]] | p_v[[20]]) / (p_v[[21]] | p_v[[22]]) / (p_v[[23]] | p_v[[24]])
s_m <- (p_m[[17]] | p_m[[18]]) / (p_m[[19]] | p_m[[20]]) / (p_m[[21]] | p_m[[22]]) / (p_m[[23]] | p_m[[24]])

# Add plot titles
e_v <- e_v + plot_annotation(title = "2102EP Cell Line") & theme(plot.title = element_text(size = 18, face = "bold"))
e_m <- e_m + plot_annotation(title = "2102EP Cell Line") & theme(plot.title = element_text(size = 18, face = "bold"))
f_v <- f_v + plot_annotation(title = "MPAF Cell Line") & theme(plot.title = element_text(size = 18, face = "bold"))
f_m <- f_m + plot_annotation(title = "MPAF Cell Line") & theme(plot.title = element_text(size = 18, face = "bold"))
s_v <- s_v + plot_annotation(title = "TCam2 Cell Line") & theme(plot.title = element_text(size = 18, face = "bold"))
s_m <- s_m + plot_annotation(title = "TCam2 Cell Line") & theme(plot.title = element_text(size = 18, face = "bold"))
```

Save all

```{r}
savePatch(e_v, "2102EP", "published_volcano")
savePatch(e_m, "2102EP", "published_ma")
savePatch(f_v, "MPAF", "published_volcano")
savePatch(f_m, "MPAF", "published_ma")
savePatch(s_v, "TCam2", "published_volcano")
savePatch(s_m, "TCam2", "published_ma")
```

## PCA

```{r}
logcounts <- assay(PRJNA814284, "logcounts")
metadata <- setDT(data.frame(colData(PRJNA814284)))
metadata[, `:=`(
  treatment = fcase(
    group %like% "DMSO_0.0001p", "DMSO 0.0001%",
    group %like% "DMSO_0.001p", "DMSO 0.001%",
    group %like% "SY0351", "SY0351 10nM",
    group %like% "THZ531", "THZ531 100nM",
    group %like% "YKL5124", "YKL5124 100nM",
    group %like% "NVP", "NVP 10nM"),
  time = ifelse(group %like% "1h", "1hr", "24hr"))][, 
  group2 := paste(treatment, time)][, 
  cell_line := fcase(group %like% "Fibroblast", "MPAF",
                     group %like% "Seminoma", "Tcam2",
                     group %like% "Embryonal", "2102EP")]
```

On all cell lines

```{r}
setDF(metadata, rownames = metadata$BioSample)
res <- pca(logcounts, metadata, removeVar = 0.5)


biplot(res, colby = "cell_line", hline = 0, hlineType = 2, vline = 0, vlineType = 2, 
       title = "PCA on All Groups", legendPosition = "bottom")
```


```{r}
# Get sample names for each specific cell type
f_samples <- metadata[group %like% "Fibroblast", BioSample]
e_samples <- metadata[group %like% "Embryonal", BioSample]
s_samples <- metadata[group %like% "Seminoma", BioSample]

# Convert to a data.frame for PCAtools
setDF(metadata, rownames = metadata$BioSample)
```

All Features

```{r}
res_f <- pca(logcounts[, f_samples], metadata = metadata[f_samples, ], removeVar = 0.5)
res_e <- pca(logcounts[, e_samples], metadata = metadata[e_samples, ], removeVar = 0.5)
res_s <- pca(logcounts[, s_samples], metadata = metadata[s_samples, ], removeVar = 0.5)

b1 <- biplot(
  res_f, 
  colby = "treatment", 
  shape = "time",
  lab = NULL, 
  legendPosition = "bottom",
  hline = 0, 
  hlineType = 2,
  vline = 0,
  vlineType = 2,
  pointSize = 5,
  title = "MPAF Cell Line") +
  scale_color_brewer(palette = "Dark2")

b2 <- biplot(
  res_e, 
  colby = "treatment", 
  shape = "time",
  lab = NULL, 
  legendPosition = "bottom",
  hline = 0, 
  hlineType = 2,
  vline = 0,
  vlineType = 2,
  pointSize = 5,
  title = "2102EP Cell Line") +
  scale_color_brewer(palette = "Dark2")

b3 <- biplot(
  res_s, 
  colby = "treatment", 
  shape = "time",
  lab = NULL, 
  legendPosition = "bottom",
  hline = 0, 
  hlineType = 2,
  vline = 0,
  vlineType = 2,
  pointSize = 5,
  title = "TCam2 Cell Line") +
  scale_color_brewer(palette = "Dark2")

(b1 | b2 | b3) +
  plot_annotation(title = "All Features") +
  plot_layout(guides = "collect") & theme(legend.position = "bottom")
ggsave(here("results", "PRJNA814284", "figures", "all-pca.png"), width = 20, height = 8)
```

genes only

```{r}
genes <- rownames(logcounts)[!rownames(logcounts) %like% ".*\\..*\\..*"]
res_fg <- pca(logcounts[genes, f_samples], metadata = metadata[f_samples, ], removeVar = 0.5)
res_eg <- pca(logcounts[genes, e_samples], metadata = metadata[e_samples, ], removeVar = 0.5)
res_sg <- pca(logcounts[genes, s_samples], metadata = metadata[s_samples, ], removeVar = 0.5)

b1 <- biplot(
  res_fg, 
  colby = "treatment", 
  shape = "time",
  lab = NULL, 
  legendPosition = "bottom",
  hline = 0, 
  hlineType = 2,
  vline = 0,
  vlineType = 2,
  pointSize = 5,
  title = "MPAF Cell Line") +
  scale_color_brewer(palette = "Dark2")

b2 <- biplot(
  res_eg, 
  colby = "treatment", 
  shape = "time",
  lab = NULL, 
  legendPosition = "bottom",
  hline = 0, 
  hlineType = 2,
  vline = 0,
  vlineType = 2,
  pointSize = 5,
  title = "2102EP Cell Line") +
  scale_color_brewer(palette = "Dark2")

b3 <- biplot(
  res_sg, 
  colby = "treatment", 
  shape = "time",
  lab = NULL, 
  legendPosition = "bottom",
  hline = 0, 
  hlineType = 2,
  vline = 0,
  vlineType = 2,
  pointSize = 5,
  title = "TCam2 Cell Line") +
  scale_color_brewer(palette = "Dark2")

(b1 | b2 | b3) +
  plot_annotation(title = "Protein Coding Genes") +
  plot_layout(guides = "collect") & theme(legend.position = "bottom")
ggsave(here("results", "PRJNA814284", "figures", "genes-pca.png"), width = 20, height = 8)
```

repeats only 

```{r}
repeats <- rownames(logcounts)[rownames(logcounts) %like% ".*\\..*\\..*"]
res_fr <- pca(logcounts[repeats, f_samples], metadata = metadata[f_samples, ], removeVar = 0.5)
res_er <- pca(logcounts[repeats, e_samples], metadata = metadata[e_samples, ], removeVar = 0.5)
res_sr <- pca(logcounts[repeats, s_samples], metadata = metadata[s_samples, ], removeVar = 0.5)

b1 <- biplot(
  res_fr, 
  colby = "treatment", 
  shape = "time",
  lab = NULL, 
  legendPosition = "bottom",
  hline = 0, 
  hlineType = 2,
  vline = 0,
  vlineType = 2,
  pointSize = 5,
  title = "MPAF Cell Line") +
  scale_color_brewer(palette = "Dark2")

b2 <- biplot(
  res_er, 
  colby = "treatment", 
  shape = "time",
  lab = NULL, 
  legendPosition = "bottom",
  hline = 0, 
  hlineType = 2,
  vline = 0,
  vlineType = 2,
  pointSize = 5,
  title = "2102EP Cell Line") +
  scale_color_brewer(palette = "Dark2")

b3 <- biplot(
  res_sr, 
  colby = "treatment", 
  shape = "time",
  lab = NULL, 
  legendPosition = "bottom",
  hline = 0, 
  hlineType = 2,
  vline = 0,
  vlineType = 2,
  pointSize = 5,
  title = "TCam2 Cell Line") +
  scale_color_brewer(palette = "Dark2")

(b1 | b2 | b3) +
  plot_annotation(title = "Repetitive Elements") +
  plot_layout(guides = "collect") & theme(legend.position = "bottom")
ggsave(here("results", "PRJNA814284", "figures", "repeats-pca.png"), width = 20, height = 8)
```
