---
title: "Find Stable Repeats 2"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## List the `SummarizedExperiment` results

```{r}
library(here)


se_files <- list.files(
  path = here("results"),
  pattern = "se(.*)?.rds",
  recursive = TRUE,
  full.names = TRUE
)

# Remove combined datasets
se_files <- se_files[!grepl("combined", se_files)]

# Do some work to add batch information to SE objects
projects <- regmatches(se_files, regexpr("PRJNA[0-9]+", se_files))
fnames <- basename(se_files)
batches <- gsub("-", "", gsub("\\.rds", "", gsub("se", "", fnames)))
se_names <- paste(projects, batches, sep = ".")
se_names <- gsub("\\.$", "", se_names)
names(se_files) <- se_names
```

## Get dispersion estimates and avgLogCPM for each experiment

```{r}
suppressPackageStartupMessages(library(SummarizedExperiment))
library(edgeR)
library(data.table)


getDispersionData <- function(fpath, assay_name = "qs_norm", disp_type = "tagwise.dispersion") {
  se <- readRDS(fpath)
  counts <- assay(se, assay_name)
  avgLogCPM <- aveLogCPM(counts) 
  dispersions <- metadata(se)[["dispersions"]][[disp_type]]
  feature_ids <- rownames(counts)
  dt <- data.table(feature_id = feature_ids, avgLogCPM = avgLogCPM, dispersion = dispersions)
  
  return(dt)
}

dt <- rbindlist(lapply(se_files, getDispersionData), idcol = "BioProject")
```

EDA on dispersions

```{r}
library(ggplot2)
library(ggrepel)
theme_set(theme_light())


# Read in gene annotations
cds <- readRDS(here("doc", "gencode.v26.annotation.cds.rds"))

# Calculate the mean dispersion parameters across all
by_feature <- dt[, .(mean_dispersion = mean(dispersion, na.rm = TRUE),
                     median_dispersion = median(dispersion, na.rm = TRUE),
                     sd_dispersion = sd(dispersion, na.rm = TRUE),
                     var_dispersion = var(dispersion, na.rm = TRUE),
                     mad_dispersion = mad(dispersion, na.rm = TRUE),
                     mean_avgLogCPM = mean(avgLogCPM, na.rm = TRUE),
                     sd_avgLogCPM = sd(avgLogCPM, na.rm = TRUE),
                     var_avgLogCPM = var(avgLogCPM, na.rm = TRUE),
                     mad_avgLogCPM = mad(avgLogCPM, na.rm = TRUE),
                     N = .N),
                 by = "feature_id"]

repeat_by_feature <- by_feature[!feature_id %chin% cds$gene_name]

# write this table out 
fwrite(by_feature, here("results", "normalization", "data-files", "feature-data.tsv"), sep = "\t")
```

Create a BCV plot for the averages across all features

```{r}
ggplot(by_feature, aes(x = mean_avgLogCPM, y = sqrt(mean_dispersion), label = feature_id)) +
  geom_point(color = "grey", alpha = 0.5) +
  geom_point(data = repeat_by_feature[N < 106], color = "blue") +
  geom_point(data = repeat_by_feature[N >= 106], color = "red") +
  geom_point(data = by_feature[feature_id %chin% c("MYC", "MYCN", "GAPDH")], color = "green") +
  geom_hline(yintercept = 0.4, color = "blue", linetype = 3) +
  labs(title = "Expression vs BCV Across All Studies",
       subtitle = "Red points are 187 repeat elements in all experiments",
       x = "Mean logCPM",
       y = "Mean BCV",
       caption = "Dotted blue line shows typical value for a well controlled experiment") +
  theme(plot.title = element_text(face = "bold", size = 14))
ggsave(here("results", "normalization", "figures", "bcv-plot.png"), width = 11, height = 7)
```

How much do these elements vary around their means?

typical values for the common BCV (square-root-dispersion) for datasets arising 
from well-controlled experiments are 0.4 for human data, 0.1 for data on genetically 
identical model organisms or 0.01 for technical replicates

```{r}
library(forcats)


repeat_by_feature[, c("class", "family", "subfamily") := tstrsplit(feature_id, "\\.")]


top <- repeat_by_feature[N >= 106][order(mean_dispersion)][1:25]
bottom <- repeat_by_feature[N>= 106][order(mean_dispersion, decreasing = TRUE)][1:25]
bcv_dt <- rbindlist(list(top, bottom))

ggplot(bcv_dt, aes(y = fct_reorder(feature_id, -sqrt(mean_dispersion)), x = sqrt(mean_dispersion), color = class)) +
  geom_point(size = 3) +
  labs(title = "BCV of top and bottom 25 REs",
       x = "Mean BCV",
       y = NULL) +
  scale_color_brewer(palette = "Set1") +
  theme(plot.title = element_text(face = "bold", size = 14),
        legend.position = "bottom")
ggsave(here("results", "normalization", "figures", "stable-repeat-bcv.png"), width = 11, height = 7)
```

## Can we normalize on these elements?

```{r}
stable_repeats <- repeat_by_feature[N == 106, feature_id]
stable_repeats
```

---

## DE pipeline for HH1 experiment (PRJNA413957)

Calculate the normalization factors on the filtered counts:

- on all features by TMM
- on only repeats by TMM
- by QSmooth
- by RUVg using stable repeats as control genes
- only library size

```{r}
library(RUVSeq)


# Extract data from SE object
se <- readRDS(se_files[["PRJNA413957"]])
counts <- assay(se, "counts")
design <- metadata(se)$design
contrast.matrix <- metadata(se)$contrasts
metadata <- setDT(as.data.frame(colData(se)))
y <- DGEList(counts, samples = metadata)

# Calculate norm factors for edgeR
y <- calcNormFactors(y, method = "TMM")
norm_factors <- calcNormFactors(counts[stable_repeats, ], method = "TMM")

# Calculate logCPMs
lcpm.tmm <- cpm(y, log = TRUE)
lcpm.tmm_re <- cpm(counts, lib.size = norm_factors * colSums(counts), log = TRUE)
lcpm.qs <- assay(se, "lcpm")
lcpm.raw <- cpm(counts, log = TRUE)

ruv.out <- RUVg(counts, stable_repeats, k = 1)
lcpm.ruv <- cpm(ruv.out$normalizedCounts, log = TRUE)
```

plot

```{r}
library(patchwork)


# get data into ggplot format
raw_dt <- as.data.table(lcpm.raw, keep.rownames = "feature_id")
qs_dt <- as.data.table(lcpm.qs, keep.rownames = "feature_id")
tmm_dt <- as.data.table(lcpm.tmm, keep.rownames = "feature_id")
tmmRE_dt <- as.data.table(lcpm.tmm_re, keep.rownames = "feature_id")
ruv_dt <- as.data.table(lcpm.ruv, keep.rownames = "feature_id")

# Pivot counts longer
raw_dt.m <- melt(raw_dt, id.vars = "feature_id", variable.name = "BioSample", value.name = "logCPM")
qs_dt.m <- melt(qs_dt, id.vars = "feature_id", variable.name = "BioSample", value.name = "logCPM")
tmm_dt.m <- melt(tmm_dt, id.vars = "feature_id", variable.name = "BioSample", value.name = "logCPM")
tmmRE_dt.m <- melt(tmmRE_dt, id.vars = "feature_id", variable.name = "BioSample", value.name = "logCPM")
ruv_dt.m <- melt(ruv_dt, id.vars = "feature_id", variable.name = "BioSample", value.name = "logCPM")

# Join on metadata information
raw_dt.m <- merge(raw_dt.m, metadata, by = "BioSample", all.x = TRUE, all.y = FALSE)
qs_dt.m <- merge(qs_dt.m, metadata, by = "BioSample", all.x = TRUE, all.y = FALSE)
tmm_dt.m <- merge(tmm_dt.m, metadata, by = "BioSample", all.x = TRUE, all.y = FALSE)
tmmRE_dt.m <- merge(tmmRE_dt.m, metadata, by = "BioSample", all.x = TRUE, all.y = FALSE)
ruv_dt.m <- merge(ruv_dt.m, metadata, by = "BioSample", all.x = TRUE, all.y = FALSE)

# create labels per treatment/time
tmm_dt.m[, c("treatment", "dose", "time") := tstrsplit(group, "\\.")]
tmmRE_dt.m[, c("treatment", "dose", "time") := tstrsplit(group, "\\.")]
raw_dt.m[, c("treatment", "dose", "time") := tstrsplit(group, "\\.")]
qs_dt.m[, c("treatment", "dose", "time") := tstrsplit(group, "\\.")]
ruv_dt.m[, c("treatment", "dose", "time") := tstrsplit(group, "\\.")]

# Define Group factor levels
group_lvls <- c("dmso..2", "dmso..4", "dmso..8", "dmso..24", "dmso..96",
                "hh1.10uM.2", "hh1.10uM.4", "hh1.10uM.8", "hh1.10uM.24", "hh1.10uM.96",
                "hh1.25uM.96", "dac_dmso..96", "dac_hh1.10uM.96", "dac_hh1.25uM.96",
                "tetOn..72", "tetOff..72")

raw_dt.m[, group := factor(group, levels = group_lvls)]
tmm_dt.m[, group := factor(group, levels = group_lvls)]
tmmRE_dt.m[, group := factor(group, levels = group_lvls)]
qs_dt.m[, group := factor(group, levels = group_lvls)]
ruv_dt.m[, group := factor(group, levels = group_lvls)]

# boxplots ---
plotBoxplot <- function(df, plot_title, fill_col = treatment) {
  ggplot(df, aes(x = group, y = logCPM, fill = {{ fill_col }})) +
    geom_boxplot(outlier.alpha = 0.25) +
    labs(title = plot_title,
         x = NULL,
         y = "logCPM") +
    scale_fill_brewer(palette = "Dark2") +
    theme_light() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          legend.position = "bottom")
}

p.raw <- plotBoxplot(raw_dt.m, "Library Size Normalized")
p.tmm <- plotBoxplot(tmm_dt.m, "TMM Normalized")
p.tmmRE <- plotBoxplot(tmmRE_dt.m, "Normalized on Stable Repeats")
p.qs <- plotBoxplot(qs_dt.m, "QS Normalized")
p.ruv <- plotBoxplot(ruv_dt.m, "RUVg Normalized")

# Density plots
plotDensity <- function(df, plot_title, color_col = treatment) {
  ggplot(df, aes(x = logCPM, group = sample_name, color = {{ color_col }})) +
    geom_density(size = 0.5) +
    labs(title = plot_title,
         x = "logCPM",
         y = "Density") +
    theme_light() +
    scale_color_brewer(palette = "Dark2") +
    theme(legend.position = "bottom")
}

p2.raw <- plotDensity(raw_dt.m, "Library Size Normalized")
p2.tmm <- plotDensity(tmm_dt.m, "TMM Normalized")
p2.tmmRE <- plotDensity(tmmRE_dt.m, "Normalized on Stable Repeats")
p2.qs <- plotDensity(qs_dt.m, "QSmooth Normalized")
p2.ruv <- plotDensity(ruv_dt.m, "RUVg Normalized")

# Combine into patchwork
(p.raw | p.tmm | p.tmmRE | p.qs | p.ruv) / (p2.raw | p2.tmm | p2.tmmRE | p2.qs | p2.ruv) +
  plot_layout(guides = "collect") & 
  theme(legend.position = "bottom")
ggsave(here("results", "normalization", "figures", "hh1-normalization.png"), width = 20, height = 9)
```

## PCA biplots

```{r}
library(PCAtools)


setDT(metadata)
metadata[, c("treatment", "dose", "time") := tstrsplit(group, "\\.")]
setDF(metadata, rownames = metadata$BioSample)

pca.raw <- pca(lcpm.raw, metadata, scale = FALSE, center = TRUE)
pca.tmm <- pca(lcpm.tmm, metadata, scale = FALSE, center = TRUE)
pca.tmmRE <- pca(lcpm.tmm_re, metadata, scale = FALSE, center = TRUE)
pca.qs <- pca(lcpm.qs, metadata, scale = FALSE, center = TRUE)
pca.ruv <- pca(lcpm.ruv, metadata, scale = FALSE, center = TRUE)

b.raw <- biplot(
  pca.raw, 
  colby = "treatment", 
  lab = metadata$time,
  vline = 0,
  vlineType = 2,
  hline = 0,
  hlineType = 2,
  title = "Library Size Normalized") + 
  scale_color_brewer(palette = "Dark2")

b.tmm <- biplot(
  pca.tmm, 
  colby = "treatment", 
  lab = metadata$time,
  vline = 0,
  vlineType = 2,
  hline = 0,
  hlineType = 2,
  title = "TMM Normalized") +
  scale_color_brewer(palette = "Dark2")

b.tmmRE <- biplot(
  pca.tmmRE, 
  colby = "treatment", 
  lab = metadata$time,
  vline = 0,
  vlineType = 2,
  hline = 0,
  hlineType = 2,
  title = "Normalized by Stable REs") +
  scale_color_brewer(palette = "Dark2")

b.qs <- biplot(
  pca.qs, 
  colby = "treatment", 
  lab = metadata$time,
  vline = 0,
  vlineType = 2,
  hline = 0,
  hlineType = 2,
  title = "QSmooth Normalized") +
  scale_color_brewer(palette = "Dark2")

b.ruv <- biplot(
  pca.ruv, 
  colby = "treatment", 
  lab = metadata$time,
  vline = 0,
  vlineType = 2,
  hline = 0,
  hlineType = 2,
  title = "RUVg Normalized") +
  scale_color_brewer(palette = "Dark2")

(b.raw | b.tmm | b.tmmRE | b.qs | b.ruv) +
  plot_layout(guides = "collect") &
  theme(legend.position = "bottom")
ggsave(here("results", "normalization", "figures", "hh1-biplots.png"), width = 22, height = 6)
```

## Do the strategies result in differential expression differences?

Since trended dispersion is used in QL pipeline, fit to trended dispersions

### TMM

```{r}
y <- estimateDisp(y, design = design, robust = TRUE)
fit.tmm <- glmQLFit(y, design = design, robust = TRUE)
plotQLDisp(fit.tmm)
```

### TMM RE norm factors

**Use effective library sizes when fitting dispersion**

```{r}
disp.tmmRE <- estimateDisp(y$counts, design = design, lib.size = y$samples$lib.size * norm_factors)
fit.tmmRE <- glmQLFit(y$counts, design = design, dispersion = disp.tmmRE$trended.dispersion, robust = TRUE)
plotQLDisp(fit.tmmRE)
```

### QSmooth

```{r}
library(qsmooth)


qs <- qsmooth(y$counts, group_factor = y$samples$group)
disp.qs <- estimateDisp(qsmoothData(qs), design = design, robust = TRUE)
fit.qs <- glmQLFit(qsmoothData(qs), design = design, dispersion = disp.qs$trended.dispersion, robust = TRUE)
plotQLDisp(fit.qs)
```

### RUVg

```{r}
W <- as.numeric(ruv.out$W)
design.ruv <- model.matrix(~0 + metadata$group + W)
colnames(design.ruv) <- gsub("metadata$group", "", colnames(design.ruv), fixed = TRUE)

y.ruv <- DGEList(counts, samples = metadata)
y.ruv <- calcNormFactors(y.ruv, method = "upperquartile")
y.ruv <- estimateDisp(y.ruv, design = design.ruv, robust = TRUE)
fit.ruv <- glmQLFit(y.ruv, design = design.ruv, robust = TRUE)
plotQLDisp(fit.ruv)

# need separate contrast mat for RUV
contrast.matrix2 <- makeContrasts(
  YB5.HH1.10uM_vs_DMSO_2hr = hh1.10uM.2 - dmso..2,
  YB5.HH1.10uM_vs_DMSO_4hr = (hh1.10uM.4 - hh1.10uM.2) - (dmso..4 - dmso..2),
  YB5.HH1.10uM_vs_DMSO_8hr = (hh1.10uM.8 - hh1.10uM.2) - (dmso..8 - dmso..2),
  YB5.HH1.10uM_vs_DMSO_24hr = (hh1.10uM.24 - hh1.10uM.2) - (dmso..24 - dmso..2),
  YB5.HH1.10uM_vs_DMSO_96hr = (hh1.10uM.96 - hh1.10uM.2) - (dmso..96 - dmso..2),
  
  YB5.Decitabine_vs_DMSO_96hr = dac_dmso..96 - dmso..96,
  YB5.HH1.25uM_vs_DMSO_96hr = hh1.25uM.96 - dmso..96,
  YB5.Decitabine_HH1.10uM_vs_DMSO_96hr = dac_hh1.10uM.96 - dmso..96,
  YB5.Decitabine_HH1.25uM_vs_DMSO_96hr = dac_hh1.25uM.96 - dmso..96,
  YB5.TetOff_vs_TetOn = tetOff..72 - tetOn..72,
  
  levels = design.ruv
)
```

### Library Size (Total Count)

```{r}
disp.tc <- estimateDisp(y$counts, design = design, robust = TRUE)
fit.tc <- glmQLFit(y$counts, design = design, dispersion = disp.tc$trended.dispersion, robust = TRUE)
```

## Perform DE testing on all contrasts for each fit type

```{r}
library(coriell)


getDeResults <- function(fit, contrast_matrix, contrast, fc = 1.5) {
  res <- glmTreat(fit, contrast = contrast_matrix[, contrast], lfc = log2(fc))
  return(edger_to_df(res))
}

res.tmm <- lapply(colnames(contrast.matrix), getDeResults, fit = fit.tmm, contrast_matrix = contrast.matrix)
res.tmmRE <- lapply(colnames(contrast.matrix), getDeResults, fit = fit.tmmRE, contrast_matrix = contrast.matrix)
res.qs <- lapply(colnames(contrast.matrix), getDeResults, fit = fit.qs, contrast_matrix = contrast.matrix)
res.ruv <- lapply(colnames(contrast.matrix2), getDeResults, fit = fit.ruv, contrast_matrix = contrast.matrix2)
res.tc <- lapply(colnames(contrast.matrix), getDeResults, fit = fit.tc, contrast_matrix = contrast.matrix)
names(res.tmm) <- names(res.tmmRE) <- names(res.qs) <- names(res.ruv) <- names(res.tc) <- colnames(contrast.matrix)
```

## Plot the results

```{r}
plotAllDeResults <- function(contrast, type = c("volcano", "ma"), fdr = 0.1, fc = 1.5) {
  if (type == "volcano") {
    p1 <- plot_volcano(res.tmm[[contrast]], fdr = fdr, lfc = log2(fc)) + 
      geom_point(data = subset(res.tmm[[contrast]], feature_id %in% stable_repeats), color = "green") +
      ggtitle("TMM")
    p2 <- plot_volcano(res.tmmRE[[contrast]], fdr = fdr, lfc = log2(fc)) + 
      geom_point(data = subset(res.tmmRE[[contrast]], feature_id %in% stable_repeats), color = "green") +
      ggtitle("Stable REs")
    p3 <- plot_volcano(res.qs[[contrast]], fdr = fdr, lfc = log2(fc)) + 
      geom_point(data = subset(res.qs[[contrast]], feature_id %in% stable_repeats), color = "green") +
      ggtitle("QSmooth")
    p4 <- plot_volcano(res.ruv[[contrast]], fdr = fdr, lfc = log2(fc)) + 
      geom_point(data = subset(res.ruv[[contrast]], feature_id %in% stable_repeats), color = "green") +
      ggtitle("RUVg")
    p5 <- plot_volcano(res.tc[[contrast]], fdr = fdr, lfc = log2(fc)) + 
      geom_point(data = subset(res.tc[[contrast]], feature_id %in% stable_repeats), color = "green") +
      ggtitle("Library Size")
  } else {
    p1 <- plot_md(res.tmm[[contrast]], fdr = fdr, lfc = log2(fc)) + 
      geom_point(data = subset(res.tmm[[contrast]], feature_id %in% stable_repeats), color = "green") +
      ggtitle("TMM")
    p2 <- plot_md(res.tmmRE[[contrast]], fdr = fdr, lfc = log2(fc)) +
      geom_point(data = subset(res.tmmRE[[contrast]], feature_id %in% stable_repeats), color = "green") +
      ggtitle("Stable REs")
    p3 <- plot_md(res.qs[[contrast]], fdr = fdr, lfc = log2(fc)) +
      geom_point(data = subset(res.qs[[contrast]], feature_id %in% stable_repeats), color = "green") +
      ggtitle("QSmooth")
    p4 <- plot_md(res.ruv[[contrast]], fdr = fdr, lfc = log2(fc)) + 
      geom_point(data = subset(res.ruv[[contrast]], feature_id %in% stable_repeats), color = "green") +
      ggtitle("RUVg")
    p5 <- plot_md(res.tc[[contrast]], fdr = fdr, lfc = log2(fc)) +
      geom_point(data = subset(res.tc[[contrast]], feature_id %chin% stable_repeats), color = "green") +
      ggtitle("Library Size")
  }
  
  patch <- (p1 | p2 | p3) / (p4 | p5 | plot_spacer()) +
    plot_layout(guides = "collect") +
    plot_annotation(title = contrast, subtitle = "Control REs colored green") &
    theme(legend.position = "bottom")

  return(patch)
}

# Volcano plots
for (con in colnames(contrast.matrix)) {
  plotAllDeResults(con, type = "volcano")
  ggsave(here("results", "normalization", "figures", paste0(con, "-volcano.png")), width = 18, height = 12)
}

# MA plots
for (con in colnames(contrast.matrix)) {
  plotAllDeResults(con, type = "ma")
  ggsave(here("results", "normalization", "figures", paste0(con, "-ma.png")), width = 18, height = 12)
}
```

---

## Process toyocamycin reads

```{r}
toyo_files <- list.files(
  path = "/mnt/data/data_gc/projects/yb5-combined-analysis/data/toyocamycin/quants",
  pattern = "quant.sf",
  recursive = TRUE,
  full.names = TRUE
)

names(toyo_files) <- regmatches(toyo_files, regexpr("(dmso|toyocamycin)_[0-9]+hr_[1-3]", toyo_files))

# Read in annotations for processed steps
gencode_dt <- readRDS("/mnt/data/gdata/human/REdiscoverTE_hg38/gencode_dt.rds")
exon_dt <- readRDS("/mnt/data/gdata/human/REdiscoverTE_hg38/exon_dt.rds")
intron_dt <- readRDS("/mnt/data/gdata/human/REdiscoverTE_hg38/intron_dt.rds")
intergenic_dt <- readRDS("/mnt/data/gdata/human/REdiscoverTE_hg38/intergenic_dt.rds")

# Process all quant files
processed <- parallel::mclapply(
  toyo_files, 
  process_quant_file,
  gene_annot = gencode_dt,
  exon_annot = exon_dt,
  intron_annot = intron_dt,
  intergenic_annot = intergenic_dt, 
  mc.cores = length(toyo_files)
  )

gene_dt <- rbindlist(lapply(processed, function(x) x[["gene"]]), idcol = "sample_name")
re_dt <- rbindlist(lapply(processed, function(x) x[["re"]]), idcol = "sample_name")

gene_mat <- as.matrix(
  dcast(gene_dt, symbol ~ sample_name, value.var = "count", fill = 0L),
  rownames = "symbol"
  )

re_mat <- as.matrix(
  dcast(re_dt, repElem ~ sample_name, value.var = "count", fill = 0L),
  rownames = "repElem"
  )

cds <- readRDS(here("doc", "gencode.v26.annotation.cds.rds"))
re_mat <- subset(re_mat, grepl(rownames(re_mat), pattern = "^LINE\\.|^SINE\\.|^LTR\\.|.*\\.SVA\\..*|^DNA\\."))
gene_mat <- subset(gene_mat, rownames(gene_mat) %in% cds$gene_name)

all(colnames(gene_mat) == colnames(re_mat))
counts <- rbind(gene_mat, re_mat)

metadata <- data.frame(sample_name = colnames(counts))
metadata$treatment <- factor(regmatches(metadata$sample_name, regexpr("(dmso|toyocamycin)", metadata$sample_name)))
metadata$time <- regmatches(metadata$sample_name, regexpr("[0-9]+hr", metadata$sample_name))
metadata$group <- factor(paste(metadata$treatment, metadata$time, sep = "."))
rownames(metadata) <- colnames(counts)
```

## Calculate Normfactors and logCPMs

```{r}
y <- DGEList(counts, samples = metadata)

# Calculate norm factors for edgeR
y <- calcNormFactors(y, method = "TMM")
norm_factors <- calcNormFactors(counts[stable_repeats, ], method = "TMM")

# Calculate logCPMs
lcpm.tmm <- cpm(y, log = TRUE)
lcpm.tmm_re <- cpm(counts, lib.size = norm_factors * colSums(counts), log = TRUE)
lcpm.raw <- cpm(counts, log = TRUE)

ruv.out <- RUVg(counts, stable_repeats, k = 1)
lcpm.ruv <- cpm(ruv.out$normalizedCounts, log = TRUE)

qs <- qsmooth(counts, group_factor = metadata$group)
lcpm.qs <- cpm(qsmoothData(qs), log = TRUE)
```

```{r}
# get data into ggplot format
raw_dt <- as.data.table(lcpm.raw, keep.rownames = "feature_id")
qs_dt <- as.data.table(lcpm.qs, keep.rownames = "feature_id")
tmm_dt <- as.data.table(lcpm.tmm, keep.rownames = "feature_id")
tmmRE_dt <- as.data.table(lcpm.tmm_re, keep.rownames = "feature_id")
ruv_dt <- as.data.table(lcpm.ruv, keep.rownames = "feature_id")

# Pivot counts longer
raw_dt.m <- melt(raw_dt, id.vars = "feature_id", variable.name = "sample_name", value.name = "logCPM")
qs_dt.m <- melt(qs_dt, id.vars = "feature_id", variable.name = "sample_name", value.name = "logCPM")
tmm_dt.m <- melt(tmm_dt, id.vars = "feature_id", variable.name = "sample_name", value.name = "logCPM")
tmmRE_dt.m <- melt(tmmRE_dt, id.vars = "feature_id", variable.name = "sample_name", value.name = "logCPM")
ruv_dt.m <- melt(ruv_dt, id.vars = "feature_id", variable.name = "sample_name", value.name = "logCPM")

# Join on metadata information
raw_dt.m <- merge(raw_dt.m, metadata, by = "sample_name", all.x = TRUE, all.y = FALSE)
qs_dt.m <- merge(qs_dt.m, metadata, by = "sample_name", all.x = TRUE, all.y = FALSE)
tmm_dt.m <- merge(tmm_dt.m, metadata, by = "sample_name", all.x = TRUE, all.y = FALSE)
tmmRE_dt.m <- merge(tmmRE_dt.m, metadata, by = "sample_name", all.x = TRUE, all.y = FALSE)
ruv_dt.m <- merge(ruv_dt.m, metadata, by = "sample_name", all.x = TRUE, all.y = FALSE)

# Define Group factor levels
group_lvls <- c("dmso.2hr", "toyocamycin.2hr", 
                "dmso.10hr", "toyocamycin.10hr", 
                "dmso.24hr", "toyocamycin.24hr",
                "dmso.48hr", "toyocamycin.48hr",
                "dmso.72hr", "toyocamycin.72hr")

raw_dt.m[, group := factor(group, levels = group_lvls)]
tmm_dt.m[, group := factor(group, levels = group_lvls)]
tmmRE_dt.m[, group := factor(group, levels = group_lvls)]
qs_dt.m[, group := factor(group, levels = group_lvls)]
ruv_dt.m[, group := factor(group, levels = group_lvls)]

# Create plots
p.raw <- plotBoxplot(raw_dt.m, "Library Size Normalized")
p.tmm <- plotBoxplot(tmm_dt.m, "TMM Normalized")
p.tmmRE <- plotBoxplot(tmmRE_dt.m, "Normalized on Stable Repeats")
p.qs <- plotBoxplot(qs_dt.m, "QS Normalized")
p.ruv <- plotBoxplot(ruv_dt.m, "RUVg Normalized")

p2.raw <- plotDensity(raw_dt.m, "Library Size Normalized")
p2.tmm <- plotDensity(tmm_dt.m, "TMM Normalized")
p2.tmmRE <- plotDensity(tmmRE_dt.m, "Normalized on Stable Repeats")
p2.qs <- plotDensity(qs_dt.m, "QSmooth Normalized")
p2.ruv <- plotDensity(ruv_dt.m, "RUVg Normalized")

# Combine into patchwork
(p.raw | p.tmm | p.tmmRE | p.qs | p.ruv) / (p2.raw | p2.tmm | p2.tmmRE | p2.qs | p2.ruv) +
  plot_layout(guides = "collect") & 
  theme(legend.position = "bottom")
ggsave(here("results", "normalization", "figures", "toyo-normalization.png"), width = 20, height = 9)
```

## PCA

```{r}
pca.raw <- pca(lcpm.raw, metadata, scale = FALSE, center = TRUE)
pca.tmm <- pca(lcpm.tmm, metadata, scale = FALSE, center = TRUE)
pca.tmmRE <- pca(lcpm.tmm_re, metadata, scale = FALSE, center = TRUE)
pca.qs <- pca(lcpm.qs, metadata, scale = FALSE, center = TRUE)
pca.ruv <- pca(lcpm.ruv, metadata, scale = FALSE, center = TRUE)

b.raw <- biplot(
  pca.raw, 
  colby = "treatment", 
  lab = metadata$time,
  vline = 0,
  vlineType = 2,
  hline = 0,
  hlineType = 2,
  title = "Library Size Normalized") + 
  scale_color_brewer(palette = "Dark2")

b.tmm <- biplot(
  pca.tmm, 
  colby = "treatment", 
  lab = metadata$time,
  vline = 0,
  vlineType = 2,
  hline = 0,
  hlineType = 2,
  title = "TMM Normalized") +
  scale_color_brewer(palette = "Dark2")

b.tmmRE <- biplot(
  pca.tmmRE, 
  colby = "treatment", 
  lab = metadata$time,
  vline = 0,
  vlineType = 2,
  hline = 0,
  hlineType = 2,
  title = "Normalized by Stable REs") +
  scale_color_brewer(palette = "Dark2")

b.qs <- biplot(
  pca.qs, 
  colby = "treatment", 
  lab = metadata$time,
  vline = 0,
  vlineType = 2,
  hline = 0,
  hlineType = 2,
  title = "QSmooth Normalized") +
  scale_color_brewer(palette = "Dark2")

b.ruv <- biplot(
  pca.ruv, 
  colby = "treatment", 
  lab = metadata$time,
  vline = 0,
  vlineType = 2,
  hline = 0,
  hlineType = 2,
  title = "RUVg Normalized") +
  scale_color_brewer(palette = "Dark2")

(b.raw | b.tmm | b.tmmRE | b.qs | b.ruv) +
  plot_layout(guides = "collect") &
  theme(legend.position = "bottom")
ggsave(here("results", "normalization", "figures", "toyo-biplots.png"), width = 22, height = 6)
```

## DE pipeline

```{r}
design <- model.matrix(~0 + group, data = metadata)

contrast.matrix <- makeContrasts(
  levels = design
)

contrast.matrix2 <- model.matrix(
  levels = design.ruv
)

# TMM
y <- estimateDisp(y, design = design, robust = TRUE)
fit.tmm <- glmQLFit(y, design = design, robust = TRUE)

# TMM RE norm
disp.tmmRE <- estimateDisp(y$counts, design = design, lib.size = y$samples$lib.size * norm_factors)
fit.tmmRE <- glmQLFit(y$counts, design = design, dispersion = disp.tmmRE$trended.dispersion, robust = TRUE)

# QSmooth
disp.qs <- estimateDisp(qsmoothData(qs), design = design, robust = TRUE)
fit.qs <- glmQLFit(qsmoothData(qs), design = design, dispersion = disp.qs$trended.dispersion, robust = TRUE)

# RUVg
W <- as.numeric(ruv.out$W)
design.ruv <- model.matrix(~0 + metadata$group + W)
colnames(design.ruv) <- gsub("metadata$group", "", colnames(design.ruv), fixed = TRUE)

y.ruv <- DGEList(counts, samples = metadata)
y.ruv <- calcNormFactors(y.ruv, method = "upperquartile")
y.ruv <- estimateDisp(y.ruv, design = design.ruv, robust = TRUE)
fit.ruv <- glmQLFit(y.ruv, design = design.ruv, robust = TRUE)
```



