---
title: "LINCS1000 Analysis"
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    theme: flatly
    highlight: tango
    df_print: paged
fig_width: 8
fig_height: 6
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
```

## Analysis Plan

1. Split the data into disease or organ groups (i.e. Leukemia or Colon)
2. Within disease or organ groups, split experiments for each drug class
3. Perform Meta-Analysis two ways (vote counting and p-value combination) on these subsets
4. Submit results of meta-analysis to LINCS L1000 database to see closest matching signatures

```{r}
suppressPackageStartupMessages(library(here))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(coriell))
suppressPackageStartupMessages(library(reticulate))
suppressPackageStartupMessages(library(SummarizedExperiment))
suppressPackageStartupMessages(library(glue))
suppressPackageStartupMessages(library(ggplot2))


# Specify Python interpreter
Sys.setenv(RETICULATE_PYTHON = "/usr/local/programs/anaconda3/bin/python")

# Read in meta-analysis logFC values (for attached metadata) 
se <- readRDS(here("results", "combined", "rds-files", "dge_se.rds"))

# Read in the differential expression results
de <- fread(here("results", "combined", "data-files", "dge-all-experiments.csv.gz"))

# extract metadata from SE as data.table
meta <- as.data.table(as.data.frame(colData(se)), keep.rownames = "id")

# get the ids for leukemia samples of each class
l_dnmt_ids <- meta[disease %ilike% "leukemia" & epigenetic_class == "DNMTi", id]
l_combo_ids <- meta[disease %ilike% "leukemia" & epigenetic_class == "COMBINATION", id]
l_hdac_ids <- meta[disease %ilike% "leukemia" & epigenetic_class == "HDACi", id]
l_other_ids <- meta[disease %ilike% "leukemia" & epigenetic_class == "OTHER", id]
l_non_ids <- meta[disease %ilike% "leukemia" & epigenetic_class == "NOT-EPIGENETIC", id]
l_hmt_ids <- meta[disease %ilike% "leukemia" & epigenetic_class == "HMTi", id]
l_pah_ids <- meta[disease %ilike% "leukemia" & epigenetic_class == "PAHi", id]
l_hdm_ids <- meta[disease %ilike% "leukemia" & epigenetic_class == "HDMi", id]
l_ids <- list("DNMTi" = l_dnmt_ids, "COMBINATION" = l_combo_ids, "HDACi" = l_hdac_ids,
              "OTHER" = l_other_ids, "NOT-EPIGENETIC" = l_non_ids, "HMTi" = l_hmt_ids,
              "PAHi" = l_pah_ids, "HDMi" = l_hdm_ids)

# extract the leukemia ids for each drug class
l_by_class <- lapply(l_ids, function(x) { de[id %chin% x]})

# create DGE list for each class by splitting on ID
l_dnmti <- split(l_by_class[["DNMTi"]], by = "id")
l_combo <- split(l_by_class[["COMBINATION"]], by = "id")
l_hdaci <- split(l_by_class[["HDACi"]], by = "id")
l_non <- split(l_by_class[["NOT-EPIGENETIC"]], by = "id")
l_hmti <- split(l_by_class[["HMTi"]], by = "id")
l_pahi <- split(l_by_class[["PAHi"]], by = "id")
l_hdmi <- split(l_by_class[["HDMi"]], by = "id")

# get the ids for colon samples of each class
c_dnmt_ids <- meta[tissue %ilike% "colon" & epigenetic_class == "DNMTi", id]
c_combo_ids <- meta[tissue %ilike% "colon" & epigenetic_class == "COMBINATION", id]
c_hdac_ids <- meta[tissue %ilike% "colon" & epigenetic_class == "HDACi", id]
c_other_ids <- meta[tissue %ilike% "colon" & epigenetic_class == "OTHER", id]
c_non_ids <- meta[tissue %ilike% "colon" & epigenetic_class == "NOT-EPIGENETIC", id]
c_hmt_ids <- meta[tissue %ilike% "colon" & epigenetic_class == "HMTi", id]
c_pah_ids <- meta[tissue %ilike% "colon" & epigenetic_class == "PAHi", id]
c_hdm_ids <- meta[tissue %ilike% "colon" & epigenetic_class == "HDMi", id]
c_ids <- list("DNMTi" = c_dnmt_ids, "COMBINATION" = c_combo_ids, "HDACi" = c_hdac_ids,
              "OTHER" = c_other_ids, "NOT-EPIGENETIC" = c_non_ids, "HMTi" = c_hmt_ids,
              "PAHi" = c_pah_ids, "HDMi" = c_hdm_ids)

# extract the colon ids for each drug class
c_by_class <- lapply(c_ids, function(x) { de[id %chin% x]})

# create DGE list for each class by splitting on ID
c_dnmti <- split(c_by_class[["DNMTi"]], by = "id")
c_combo <- split(c_by_class[["COMBINATION"]], by = "id")
c_hdaci <- split(c_by_class[["HDACi"]], by = "id")
c_non <- split(c_by_class[["NOT-EPIGENETIC"]], by = "id")
c_hmti <- split(c_by_class[["HMTi"]], by = "id")
c_pahi <- split(c_by_class[["PAHi"]], by = "id")
c_hdmi <- split(c_by_class[["HDMi"]], by = "id")

# Set Default Cutoffs for Meta-Analysis Functions
PVAL <- 0.01
FC <- 2
PROP <- 0.5

# Wrapper for plotting meta-volcano
plotMetavolcano <- function(dt, plot_title) {
  plot_metavolcano(dt) +
    ggtitle(plot_title)
}

# Wrapper for plotting volcano of p-combined values
plotPVolcano <- function(dt, plot_title) {
    plot_volcano(
      dt, 
      x = meta_lfc, 
      y = meta_p, 
      fdr = PVAL, 
      lfc = log2(FC), 
      label_sig = TRUE) +
    labs(title = plot_title,
         x = "Mean LogFC",
         y = "Combined P-Value [fisher]")
}

# Function to pad input list with fake gene names if too short
pad_list <- function(genes) {
  if (length(genes) < 3) {
    return(c(genes, rep("XXX", 3 - length(genes))))
  }
  return(genes)
}

# Function to extract the topMeta and combinations data from L1000 db search
extract_L1000 <- function(result) {
  if (result == "Input genes have no overlap with the LINCS L1000 genome.") {
    return("Input genes have no overlap with the LINCS L1000 genome.")
  }
  
  res_dt <- as.data.table(result)[, shareId := NULL]
  res_dt[, resId := 1:NROW(res_dt)]
  comb_dt <- res_dt[, .(comb_name = names(unlist(combinations)), comb_val = unlist(combinations)), by = "resId"]
  comb_dt.w <- dcast(comb_dt, resId ~ comb_name, value.var = "comb_val")
  setnames(comb_dt.w, old = c("X1", "X2"), new = c("Combination1", "Combination2"))
  tm_dt <- res_dt[, .(meta_name = names(unlist(topMeta)),
                      meta_val = unlist(topMeta)), by = "resId"][!meta_name %like% "overlap"]
  tm_dt.w <- dcast(tm_dt, resId ~ meta_name, value.var = "meta_val")
  dt <- merge(tm_dt.w, comb_dt.w, by = "resId", all.x = TRUE, all.y = TRUE)
  return(dt)
}
```

```{python}
# Function to Query L1000 for meta-up/down genes
import requests
import json
url = 'https://maayanlab.cloud/L1000CDS2/query'


def submit_request(up_genes, down_genes):
    """Submit a search to L1000 using the web API"""
    data = {"upGenes" : up_genes, "dnGenes" : down_genes}
    config = {"aggravate" : True, 
              "searchMethod" : "geneSet", 
              "share" : True, 
              "combination" : True, 
              "db-version" : "latest"}
    metadata = [{"key" : "Tag", "value" : "gene-set meta-search"}, 
                {"key" : "Set", "value" : "Meta-Search"}]
    payload = {"data" : data, "config" : config, "meta" : metadata}
    headers = {'content-type' : 'application/json'}
    req = requests.post(url, data=json.dumps(payload), headers=headers)
    resGeneSet = req.json()
    
    return(resGeneSet)
```

## Leukemia DNMTi

```{r}
# number of experiments
glue("Total Number of experiments: {n}", n = length(l_dnmti))

# vote-counting meta-analysis
l_dnmti_metav <- meta_vote(l_dnmti, meta_prop = PROP, pval = 0.1)

# extract up/down genes
l_dnmti_vote_up <- pad_list(l_dnmti_metav[vote == "up", feature_id])
l_dnmti_vote_down <- pad_list(l_dnmti_metav[vote == "down", feature_id])
glue("Vote-Counting Genes Up-Regulated: {vote_up}", vote_up = length(l_dnmti_vote_up))
glue("Vote-Counting Genes Down-Regulated: {vote_down}", vote_down = length(l_dnmti_vote_down))

# p-value combining meta-analysis
l_dnmti_metap <- meta_pcombine(
  l_dnmti, 
  lfc_fun = mean,
  na.rm = TRUE,
  plot_lfc = log2(FC), 
  plot_pval = PVAL)

# extract up/down genes
l_dnmti_pcomb_up <- pad_list(l_dnmti_metap[meta_p < PVAL & meta_lfc > log2(FC), feature_id])
l_dnmti_pcomb_down <- pad_list(l_dnmti_metap[meta_p < PVAL & meta_lfc < -log2(FC), feature_id])
glue("P-Combination Genes Up-Regulated: {p_up}", p_up = length(l_dnmti_pcomb_up))
glue("P-Combination Genes Down-Regulated: {p_down}", p_down = length(l_dnmti_pcomb_down))

# plot ---
plotMetavolcano(l_dnmti_metav, "Leukemia DNMTi - Vote Counting")
plotPVolcano(l_dnmti_metap, "Leukemia DNMTi - PValue Combination")
```

```{python}
# Query L1000
l_dnmti_vote_res = submit_request(r.l_dnmti_vote_up, r.l_dnmti_vote_down)
l_dnmti_pcom_res = submit_request(r.l_dnmti_pcomb_up, r.l_dnmti_pcomb_down)
```

### Vote-Counting Hits

```{r}
# Vote Counting
extract_L1000(py$l_dnmti_vote_res)[]
```

### P-Value Combination Hits

```{r}
# P-Combination
extract_L1000(py$l_dnmti_pcom_res)[]
```

## Leukemia HDACi

```{r}
# number of experiments
glue("Total Number of experiments: {n}", n = length(l_hdaci))

# vote-counting meta-analysis
l_hdaci_metav <- meta_vote(l_hdaci, meta_prop = PROP, pval = 0.05)

# extract up/down genes
l_hdaci_vote_up <- pad_list(l_hdaci_metav[vote == "up", feature_id])
l_hdaci_vote_down <- pad_list(l_hdaci_metav[vote == "down", feature_id])
glue("Vote-Counting Genes Up-Regulated: {vote_up}", vote_up = length(l_hdaci_vote_up))
glue("Vote-Counting Genes Down-Regulated: {vote_down}", vote_down = length(l_hdaci_vote_down))

# p-value combining meta-analysis
l_hdaci_metap <- meta_pcombine(
  l_hdaci, 
  lfc_fun = mean,
  na.rm = TRUE,
  plot_lfc = log2(FC), 
  plot_pval = PVAL)

# extract up/down genes
l_hdaci_pcomb_up <- pad_list(l_hdaci_metap[meta_p < PVAL & meta_lfc > log2(FC), feature_id])
l_hdaci_pcomb_down <- pad_list(l_hdaci_metap[meta_p < PVAL & meta_lfc < -log2(FC), feature_id])
glue("P-Combination Genes Up-Regulated: {p_up}", p_up = length(l_hdaci_pcomb_up))
glue("P-Combination Genes Down-Regulated: {p_down}", p_down = length(l_hdaci_pcomb_down))

# plot ---
plotMetavolcano(l_hdaci_metav, "Leukemia HDACi - Vote Counting")
plotPVolcano(l_hdaci_metap, "Leukemia HDACi - PValue Combination")
```

```{python}
# Query L1000
l_hdaci_vote_res = submit_request(r.l_hdaci_vote_up, r.l_hdaci_vote_down)
l_hdaci_pcom_res = submit_request(r.l_hdaci_pcomb_up, r.l_hdaci_pcomb_down)
```

### Vote-Counting Hits

```{r}
# Vote Counting
extract_L1000(py$l_hdaci_vote_res)[]
```

### P-Value Combination Hits

```{r}
# P-Combination
extract_L1000(py$l_hdaci_pcom_res)[]
```

## Leukemia HMTi

```{r}
# number of experiments
glue("Total Number of experiments: {n}", n = length(l_hmti))

# vote-counting meta-analysis
l_hmti_metav <- meta_vote(l_hmti, meta_prop = PROP, pval = 0.1)

# extract up/down genes
l_hmti_vote_up <- pad_list(l_hmti_metav[vote == "up", feature_id])
l_hmti_vote_down <- pad_list(l_hmti_metav[vote == "down", feature_id])
glue("Vote-Counting Genes Up-Regulated: {vote_up}", vote_up = length(l_hmti_vote_up))
glue("Vote-Counting Genes Down-Regulated: {vote_down}", vote_down = length(l_hmti_vote_down))

# p-value combining meta-analysis
l_hmti_metap <- meta_pcombine(
  l_hmti, 
  lfc_fun = mean,
  na.rm = TRUE,
  plot_lfc = log2(FC), 
  plot_pval = PVAL)

# extract up/down genes
l_hmti_pcomb_up <- pad_list(l_hmti_metap[meta_p < PVAL & meta_lfc > log2(FC), feature_id])
l_hmti_pcomb_down <- pad_list(l_hmti_metap[meta_p < PVAL & meta_lfc < -log2(FC), feature_id])
glue("P-Combination Genes Up-Regulated: {p_up}", p_up = length(l_hmti_pcomb_up))
glue("P-Combination Genes Down-Regulated: {p_down}", p_down = length(l_hmti_pcomb_down))

# plot ---
plotMetavolcano(l_hmti_metav, "Leukemia HMTi - Vote Counting")
plotPVolcano(l_hmti_metap, "Leukemia HMTi - PValue Combination")
```

```{python}
# Query L1000
l_hmti_vote_res = submit_request(r.l_hmti_vote_up, r.l_hmti_vote_down)
l_hmti_pcom_res = submit_request(r.l_hmti_pcomb_up, r.l_hmti_pcomb_down)
```

### Vote-Counting Hits

```{r}
# Vote Counting
extract_L1000(py$l_hmti_vote_res)[]
```

### P-Value Combination Hits

```{r}
# P-Combination
extract_L1000(py$l_hmti_pcom_res)[]
```

## Leukemia PAHi

```{r}
# number of experiments
glue("Total Number of experiments: {n}", n = length(l_pahi))

# vote-counting meta-analysis
l_pahi_metav <- meta_vote(l_pahi, meta_prop = PROP, pval = 0.1)

# extract up/down genes
l_pahi_vote_up <- pad_list(l_pahi_metav[vote == "up", feature_id])
l_pahi_vote_down <- pad_list(l_pahi_metav[vote == "down", feature_id])
glue("Vote-Counting Genes Up-Regulated: {vote_up}", vote_up = length(l_pahi_vote_up))
glue("Vote-Counting Genes Down-Regulated: {vote_down}", vote_down = length(l_pahi_vote_down))

# p-value combining meta-analysis
l_pahi_metap <- meta_pcombine(
  l_pahi, 
  lfc_fun = mean,
  na.rm = TRUE,
  plot_lfc = log2(FC), 
  plot_pval = PVAL)

# extract up/down genes
l_pahi_pcomb_up <- pad_list(l_pahi_metap[meta_p < PVAL & meta_lfc > log2(FC), feature_id])
l_pahi_pcomb_down <- pad_list(l_pahi_metap[meta_p < PVAL & meta_lfc < -log2(FC), feature_id])
glue("P-Combination Genes Up-Regulated: {p_up}", p_up = length(l_pahi_pcomb_up))
glue("P-Combination Genes Down-Regulated: {p_down}", p_down = length(l_pahi_pcomb_down))

# plot ---
plotMetavolcano(l_pahi_metav, "Leukemia PAHi - Vote Counting")
plotPVolcano(l_pahi_metap, "Leukemia PAHi - PValue Combination")
```

```{python}
# Query L1000
l_pahi_vote_res = submit_request(r.l_pahi_vote_up, r.l_pahi_vote_down)
l_pahi_pcom_res = submit_request(r.l_pahi_pcomb_up, r.l_pahi_pcomb_down)
```

### Vote-Counting Hits

```{r}
# Vote Counting
extract_L1000(py$l_pahi_vote_res)[]
```

### P-Value Combination Hits

```{r}
# P-Combination
extract_L1000(py$l_pahi_pcom_res)[]
```

## Leukemia HDMi

```{r}
# number of experiments
glue("Total Number of experiments: {n}", n = length(l_hdmi))

# vote-counting meta-analysis
l_hdmi_metav <- meta_vote(l_hdmi, meta_prop = PROP, pval = 0.1)

# extract up/down genes
l_hdmi_vote_up <- pad_list(l_hdmi_metav[vote == "up", feature_id])
l_hdmi_vote_down <- pad_list(l_hdmi_metav[vote == "down", feature_id])
glue("Vote-Counting Genes Up-Regulated: {vote_up}", vote_up = length(l_hdmi_vote_up))
glue("Vote-Counting Genes Down-Regulated: {vote_down}", vote_down = length(l_hdmi_vote_down))

# p-value combining meta-analysis
l_hdmi_metap <- meta_pcombine(
  l_hdmi, 
  lfc_fun = mean,
  na.rm = TRUE,
  plot_lfc = log2(FC), 
  plot_pval = PVAL)

# extract up/down genes
l_hdmi_pcomb_up <- pad_list(l_hdmi_metap[meta_p < PVAL & meta_lfc > log2(FC), feature_id])
l_hdmi_pcomb_down <- pad_list(l_hdmi_metap[meta_p < PVAL & meta_lfc < -log2(FC), feature_id])
glue("P-Combination Genes Up-Regulated: {p_up}", p_up = length(l_hdmi_pcomb_up))
glue("P-Combination Genes Down-Regulated: {p_down}", p_down = length(l_hdmi_pcomb_down))

# plot ---
plotMetavolcano(l_hdmi_metav, "Leukemia HDMi - Vote Counting")
plotPVolcano(l_hdmi_metap, "Leukemia HDMi - PValue Combination")
```

```{python}
# Query L1000
l_hdmi_vote_res = submit_request(r.l_hdmi_vote_up, r.l_hdmi_vote_down)
l_hdmi_pcom_res = submit_request(r.l_hdmi_pcomb_up, r.l_hdmi_pcomb_down)
```

### Vote-Counting Hits

```{r}
# Vote Counting
extract_L1000(py$l_hdmi_vote_res)[]
```

### P-Value Combination Hits

```{r}
# P-Combination
extract_L1000(py$l_hdmi_pcom_res)[]
```

## Leukemia Combination Epigenetic Drugs

```{r}
# number of experiments
glue("Total Number of experiments: {n}", n = length(l_combo))

# vote-counting meta-analysis
l_combo_metav <- meta_vote(l_combo, meta_prop = PROP, pval = 0.1)

# extract up/down genes
l_combo_vote_up <- pad_list(l_combo_metav[vote == "up", feature_id])
l_combo_vote_down <- pad_list(l_combo_metav[vote == "down", feature_id])
glue("Vote-Counting Genes Up-Regulated: {vote_up}", vote_up = length(l_combo_vote_up))
glue("Vote-Counting Genes Down-Regulated: {vote_down}", vote_down = length(l_combo_vote_down))

# p-value combining meta-analysis
l_combo_metap <- meta_pcombine(
  l_combo, 
  lfc_fun = mean,
  na.rm = TRUE,
  plot_lfc = log2(FC), 
  plot_pval = PVAL)

# extract up/down genes
l_combo_pcomb_up <- pad_list(l_combo_metap[meta_p < PVAL & meta_lfc > log2(FC), feature_id])
l_combo_pcomb_down <- pad_list(l_combo_metap[meta_p < PVAL & meta_lfc < -log2(FC), feature_id])
glue("P-Combination Genes Up-Regulated: {p_up}", p_up = length(l_combo_pcomb_up))
glue("P-Combination Genes Down-Regulated: {p_down}", p_down = length(l_combo_pcomb_down))

# plot ---
plotMetavolcano(l_combo_metav, "Leukemia Combination Epigenetic - Vote Counting")
plotPVolcano(l_combo_metap, "Leukemia Combination Epigenetic - PValue Combination")
```

```{python}
# Query L1000
l_combo_vote_res = submit_request(r.l_combo_vote_up, r.l_combo_vote_down)
l_combo_pcom_res = submit_request(r.l_combo_pcomb_up, r.l_combo_pcomb_down)
```

### Vote-Counting Hits

```{r}
# Vote Counting
extract_L1000(py$l_combo_vote_res)[]
```

### P-Value Combination Hits

```{r}
# P-Combination
extract_L1000(py$l_combo_pcom_res)[]
```

## Leukemia Non-Epigenetic Drugs

```{r}
# number of experiments
glue("Total Number of experiments: {n}", n = length(l_non))

# vote-counting meta-analysis
l_non_metav <- meta_vote(l_non, meta_prop = PROP, pval = 0.1)

# extract up/down genes
l_non_vote_up <- pad_list(l_non_metav[vote == "up", feature_id])
l_non_vote_down <- pad_list(l_non_metav[vote == "down", feature_id])
glue("Vote-Counting Genes Up-Regulated: {vote_up}", vote_up = length(l_non_vote_up))
glue("Vote-Counting Genes Down-Regulated: {vote_down}", vote_down = length(l_non_vote_down))

# p-value combining meta-analysis
l_non_metap <- meta_pcombine(
  l_non, 
  lfc_fun = mean,
  na.rm = TRUE,
  plot_lfc = log2(FC), 
  plot_pval = PVAL)

# extract up/down genes
l_non_pcomb_up <- pad_list(l_non_metap[meta_p < PVAL & meta_lfc > log2(FC), feature_id])
l_non_pcomb_down <- pad_list(l_non_metap[meta_p < PVAL & meta_lfc < -log2(FC), feature_id])
glue("P-Combination Genes Up-Regulated: {p_up}", p_up = length(l_non_pcomb_up))
glue("P-Combination Genes Down-Regulated: {p_down}", p_down = length(l_non_pcomb_down))

# plot ---
plotMetavolcano(l_non_metav, "Leukemia Non-Epigenetic - Vote Counting")
plotPVolcano(l_non_metap, "Leukemia Non-Epigenetic - PValue Combination")
```

```{python}
# Query L1000
l_non_vote_res = submit_request(r.l_non_vote_up, r.l_non_vote_down)
l_non_pcom_res = submit_request(r.l_non_pcomb_up, r.l_non_pcomb_down)
```

### Vote-Counting Hits

```{r}
# Vote Counting
extract_L1000(py$l_non_vote_res)[]
```

### P-Value Combination Hits

```{r}
# P-Combination
extract_L1000(py$l_non_pcom_res)[]
```

---

## Colon DNMTi

```{r}
# number of experiments
glue("Total Number of experiments: {n}", n = length(c_dnmti))

# vote-counting meta-analysis
c_dnmti_metav <- meta_vote(c_dnmti, meta_prop = PROP, pval = 0.1)

# extract up/down genes
c_dnmti_vote_up <- pad_list(c_dnmti_metav[vote == "up", feature_id])
c_dnmti_vote_down <- pad_list(c_dnmti_metav[vote == "down", feature_id])
glue("Vote-Counting Genes Up-Regulated: {vote_up}", vote_up = length(c_dnmti_vote_up))
glue("Vote-Counting Genes Down-Regulated: {vote_down}", vote_down = length(c_dnmti_vote_down))

# p-value combining meta-analysis
c_dnmti_metap <- meta_pcombine(
  c_dnmti, 
  lfc_fun = mean,
  na.rm = TRUE,
  plot_lfc = log2(FC), 
  plot_pval = PVAL)

# extract up/down genes
c_dnmti_pcomb_up <- pad_list(c_dnmti_metap[meta_p < PVAL & meta_lfc > log2(FC), feature_id])
c_dnmti_pcomb_down <- pad_list(c_dnmti_metap[meta_p < PVAL & meta_lfc < -log2(FC), feature_id])
glue("P-Combination Genes Up-Regulated: {p_up}", p_up = length(c_dnmti_pcomb_up))
glue("P-Combination Genes Down-Regulated: {p_down}", p_down = length(c_dnmti_pcomb_down))

# plot ---
plotMetavolcano(c_dnmti_metav, "Colon DNMTi - Vote Counting")
plotPVolcano(c_dnmti_metap, "Colon DNMTi - PValue Combination")
```

```{python}
# Query L1000
c_dnmti_vote_res = submit_request(r.c_dnmti_vote_up, r.c_dnmti_vote_down)
c_dnmti_pcom_res = submit_request(r.c_dnmti_pcomb_up, r.c_dnmti_pcomb_down)
```

### Vote-Counting Hits

```{r}
# Vote Counting
extract_L1000(py$c_dnmti_vote_res)[]
```

### P-Value Combination Hits

```{r}
# P-Combination
extract_L1000(py$c_dnmti_pcom_res)[]
```

## Colon HDACi

```{r}
# number of experiments
glue("Total Number of experiments: {n}", n = length(c_hdaci))

# vote-counting meta-analysis
c_hdaci_metav <- meta_vote(c_hdaci, meta_prop = PROP, pval = 0.05)

# extract up/down genes
c_hdaci_vote_up <- pad_list(c_hdaci_metav[vote == "up", feature_id])
c_hdaci_vote_down <- pad_list(c_hdaci_metav[vote == "down", feature_id])
glue("Vote-Counting Genes Up-Regulated: {vote_up}", vote_up = length(c_hdaci_vote_up))
glue("Vote-Counting Genes Down-Regulated: {vote_down}", vote_down = length(c_hdaci_vote_down))

# p-value combining meta-analysis
c_hdaci_metap <- meta_pcombine(
  c_hdaci, 
  lfc_fun = mean,
  na.rm = TRUE,
  plot_lfc = log2(FC), 
  plot_pval = PVAL)

# extract up/down genes
c_hdaci_pcomb_up <- pad_list(c_hdaci_metap[meta_p < PVAL & meta_lfc > log2(FC), feature_id])
c_hdaci_pcomb_down <- pad_list(c_hdaci_metap[meta_p < PVAL & meta_lfc < -log2(FC), feature_id])
glue("P-Combination Genes Up-Regulated: {p_up}", p_up = length(c_hdaci_pcomb_up))
glue("P-Combination Genes Down-Regulated: {p_down}", p_down = length(c_hdaci_pcomb_down))

# plot ---
plotMetavolcano(c_hdaci_metav, "Colon HDACi - Vote Counting")
plotPVolcano(c_hdaci_metap, "Colon HDACi - PValue Combination")
```

```{python}
# Query L1000
c_hdaci_vote_res = submit_request(r.c_hdaci_vote_up, r.c_hdaci_vote_down)
c_hdaci_pcom_res = submit_request(r.c_hdaci_pcomb_up, r.c_hdaci_pcomb_down)
```

### Vote-Counting Hits

```{r}
# Vote Counting
extract_L1000(py$c_hdaci_vote_res)[]
```

### P-Value Combination Hits

```{r}
# P-Combination
extract_L1000(py$c_hdaci_pcom_res)[]
```

## Colon HMTi

```{r}
# number of experiments
glue("Total Number of experiments: {n}", n = length(c_hmti))

# vote-counting meta-analysis
c_hmti_metav <- meta_vote(c_hmti, meta_prop = PROP, pval = 0.1)

# extract up/down genes
c_hmti_vote_up <- pad_list(c_hmti_metav[vote == "up", feature_id])
c_hmti_vote_down <- pad_list(c_hmti_metav[vote == "down", feature_id])
glue("Vote-Counting Genes Up-Regulated: {vote_up}", vote_up = length(c_hmti_vote_up))
glue("Vote-Counting Genes Down-Regulated: {vote_down}", vote_down = length(c_hmti_vote_down))

# p-value combining meta-analysis
c_hmti_metap <- meta_pcombine(
  c_hmti, 
  lfc_fun = mean,
  na.rm = TRUE,
  plot_lfc = log2(FC), 
  plot_pval = PVAL)

# extract up/down genes
c_hmti_pcomb_up <- pad_list(c_hmti_metap[meta_p < PVAL & meta_lfc > log2(FC), feature_id])
c_hmti_pcomb_down <- pad_list(c_hmti_metap[meta_p < PVAL & meta_lfc < -log2(FC), feature_id])
glue("P-Combination Genes Up-Regulated: {p_up}", p_up = length(c_hmti_pcomb_up))
glue("P-Combination Genes Down-Regulated: {p_down}", p_down = length(c_hmti_pcomb_down))

# plot ---
plotMetavolcano(c_hmti_metav, "Colon HMTi - Vote Counting")
plotPVolcano(c_hmti_metap, "Colon HMTi - PValue Combination")
```

```{python}
# Query L1000
c_hmti_vote_res = submit_request(r.c_hmti_vote_up, r.c_hmti_vote_down)
c_hmti_pcom_res = submit_request(r.c_hmti_pcomb_up, r.c_hmti_pcomb_down)
```

### Vote-Counting Hits

```{r}
# Vote Counting
extract_L1000(py$c_hmti_vote_res)[]
```

### P-Value Combination Hits

```{r}
# P-Combination
extract_L1000(py$c_hmti_pcom_res)[]
```

## Colon PAHi

```{r}
# number of experiments
glue("Total Number of experiments: {n}", n = length(c_pahi))

# vote-counting meta-analysis
c_pahi_metav <- meta_vote(c_pahi, meta_prop = PROP, pval = 0.1)

# extract up/down genes
c_pahi_vote_up <- pad_list(c_pahi_metav[vote == "up", feature_id])
c_pahi_vote_down <- pad_list(c_pahi_metav[vote == "down", feature_id])
glue("Vote-Counting Genes Up-Regulated: {vote_up}", vote_up = length(c_pahi_vote_up))
glue("Vote-Counting Genes Down-Regulated: {vote_down}", vote_down = length(c_pahi_vote_down))

# p-value combining meta-analysis
c_pahi_metap <- meta_pcombine(
  c_pahi, 
  lfc_fun = mean,
  na.rm = TRUE,
  plot_lfc = log2(FC), 
  plot_pval = PVAL)

# extract up/down genes
c_pahi_pcomb_up <- pad_list(c_pahi_metap[meta_p < PVAL & meta_lfc > log2(FC), feature_id])
c_pahi_pcomb_down <- pad_list(c_pahi_metap[meta_p < PVAL & meta_lfc < -log2(FC), feature_id])
glue("P-Combination Genes Up-Regulated: {p_up}", p_up = length(c_pahi_pcomb_up))
glue("P-Combination Genes Down-Regulated: {p_down}", p_down = length(c_pahi_pcomb_down))

# plot ---
plotMetavolcano(c_pahi_metav, "Colon PAHi - Vote Counting")
plotPVolcano(c_pahi_metap, "Colon PAHi - PValue Combination")
```

```{python}
# Query L1000
c_pahi_vote_res = submit_request(r.c_pahi_vote_up, r.c_pahi_vote_down)
c_pahi_pcom_res = submit_request(r.c_pahi_pcomb_up, r.c_pahi_pcomb_down)
```

### Vote-Counting Hits

```{r}
# Vote Counting
extract_L1000(py$c_pahi_vote_res)[]
```

### P-Value Combination Hits

```{r}
# P-Combination
extract_L1000(py$c_pahi_pcom_res)[]
```

## Colon Combination Epigenetic Drugs

```{r}
# number of experiments
glue("Total Number of experiments: {n}", n = length(c_combo))

# vote-counting meta-analysis
c_combo_metav <- meta_vote(c_combo, meta_prop = PROP, pval = 0.1)

# extract up/down genes
c_combo_vote_up <- pad_list(c_combo_metav[vote == "up", feature_id])
c_combo_vote_down <- pad_list(c_combo_metav[vote == "down", feature_id])
glue("Vote-Counting Genes Up-Regulated: {vote_up}", vote_up = length(c_combo_vote_up))
glue("Vote-Counting Genes Down-Regulated: {vote_down}", vote_down = length(c_combo_vote_down))

# p-value combining meta-analysis
c_combo_metap <- meta_pcombine(
  c_combo, 
  lfc_fun = mean,
  na.rm = TRUE,
  plot_lfc = log2(FC), 
  plot_pval = PVAL)

# extract up/down genes
c_combo_pcomb_up <- pad_list(c_combo_metap[meta_p < PVAL & meta_lfc > log2(FC), feature_id])
c_combo_pcomb_down <- pad_list(c_combo_metap[meta_p < PVAL & meta_lfc < -log2(FC), feature_id])
glue("P-Combination Genes Up-Regulated: {p_up}", p_up = length(c_combo_pcomb_up))
glue("P-Combination Genes Down-Regulated: {p_down}", p_down = length(c_combo_pcomb_down))

# plot ---
plotMetavolcano(c_combo_metav, "Colon Combination Epigenetic - Vote Counting")
plotPVolcano(c_combo_metap, "Colon Combination Epigenetic - PValue Combination")
```

```{python}
# Query L1000
c_combo_vote_res = submit_request(r.c_combo_vote_up, r.c_combo_vote_down)
c_combo_pcom_res = submit_request(r.c_combo_pcomb_up, r.c_combo_pcomb_down)
```

### Vote-Counting Hits

```{r}
# Vote Counting
extract_L1000(py$c_combo_vote_res)[]
```

### P-Value Combination Hits

```{r}
# P-Combination
extract_L1000(py$c_combo_pcom_res)[]
```

## Colon Non-Epigenetic Drugs

```{r}
# number of experiments
glue("Total Number of experiments: {n}", n = length(c_non))

# vote-counting meta-analysis
c_non_metav <- meta_vote(c_non, meta_prop = PROP, pval = 0.1)

# extract up/down genes
c_non_vote_up <- pad_list(c_non_metav[vote == "up", feature_id])
c_non_vote_down <- pad_list(c_non_metav[vote == "down", feature_id])
glue("Vote-Counting Genes Up-Regulated: {vote_up}", vote_up = length(c_non_vote_up))
glue("Vote-Counting Genes Down-Regulated: {vote_down}", vote_down = length(c_non_vote_down))

# p-value combining meta-analysis
c_non_metap <- meta_pcombine(
  c_non, 
  lfc_fun = mean,
  na.rm = TRUE,
  plot_lfc = log2(FC), 
  plot_pval = PVAL)

# extract up/down genes
c_non_pcomb_up <- pad_list(c_non_metap[meta_p < PVAL & meta_lfc > log2(FC), feature_id])
c_non_pcomb_down <- pad_list(c_non_metap[meta_p < PVAL & meta_lfc < -log2(FC), feature_id])
glue("P-Combination Genes Up-Regulated: {p_up}", p_up = length(c_non_pcomb_up))
glue("P-Combination Genes Down-Regulated: {p_down}", p_down = length(c_non_pcomb_down))

# plot ---
plotMetavolcano(c_non_metav, "Colon Non-Epigenetic - Vote Counting")
plotPVolcano(c_non_metap, "Colon Non-Epigenetic - PValue Combination")
```

```{python}
# Query L1000
c_non_vote_res = submit_request(r.c_non_vote_up, r.c_non_vote_down)
c_non_pcom_res = submit_request(r.c_non_pcomb_up, r.c_non_pcomb_down)
```

### Vote-Counting Hits

```{r}
# Vote Counting
extract_L1000(py$c_non_vote_res)[]
```

### P-Value Combination Hits

```{r}
# P-Combination
extract_L1000(py$c_non_pcom_res)[]
```
