---
title: "Combine Datasets"
output:
  html_document:
    df_print: paged
    highlight: tango
    theme: united
    toc: yes
    toc_depth: 4
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(here)
library(magrittr)
library(data.table)
```

## Read in the log expression matrices

Read in the expression matrices from each study and combine into a single 
dataframe. This might get huge so maybe a SQLite db is in order. IDK yet...

```{r}
gene_mat_files <- list.files(
  path = here("results"),
  pattern = "gene-lcpms.rds",
  recursive = TRUE,
  full.names = TRUE
)

re_mat_files <- list.files(
  path = here("results"),
  pattern = "re-lcpms.rds",
  recursive = TRUE,
  full.names = TRUE
)

# remove files from the 'combined' directory
gene_mat_files <- gene_mat_files[!grepl("combined", gene_mat_files)]
re_mat_files <- re_mat_files[!grepl("combined", re_mat_files)]

# define function for cleaning file paths 
clean_fpath = function(s) { gsub("/data-files/", "", 
                            gsub("/results/", "", 
                            regmatches(s, 
                            regexpr("\\/results\\/.*\\/", s)))) 
}

names(gene_mat_files) <- clean_fpath(gene_mat_files) %>% gsub(pattern = "/rds-files/", replacement = "")
names(re_mat_files) <- clean_fpath(re_mat_files) %>% gsub(pattern = "/rds-files/", replacement = "")

# create a function for reading and making count matrices long
process_count_matrix <- function(rds_path, exp_id) {
  X <- as.data.table(readRDS(rds_path), keep.rownames = "feature")
  M <- melt(X, id.vars = "feature", variable.name = "sample_name", value.name = "logCPM")
  M[, sample_name := paste0(exp_id, "__", sample_name)]
}

# process all of the rds files into list of DTs
gene_dts <- mapply(process_count_matrix, gene_mat_files, names(gene_mat_files), SIMPLIFY = FALSE)
re_dts <- mapply(process_count_matrix, re_mat_files, names(re_mat_files), SIMPLIFY = FALSE)

# bind all together
gene_lcpm_dt <- rbindlist(gene_dts)
re_lcpm_dt <- rbindlist(re_dts)

saveRDS(gene_lcpm_dt, here("results", "combined", "rds-files", "gene-lcpm-dt.rds"))
saveRDS(re_lcpm_dt, here("results", "combined", "rds-files", "re-lcpm-dt.rds"))

# cast wide and save
wide_gene_lcpm_dt <- dcast(gene_lcpm_dt, feature ~ sample_name, value.var = "logCPM", fill = 0)
wide_re_lcpm_dt <- dcast(re_lcpm_dt, feature ~ sample_name, value.var = "logCPM", fill = 0)
```

## Read in the parsed metadata files

I collected metadata from SRA for each study. Use this for sample annotation

They are all formatted differently except for the first column. We can maybe 
read in all of the files and then just unite all other columns into a single
description column.

I will still have to export this file and annotate by hand...

```{r eval=FALSE, include=FALSE}
library(tidyverse)


meta_files <- list.files(
  path = here("data"),
  pattern = "metadata.txt",
  recursive = TRUE,
  full.names = TRUE
)

names(meta_files) <- str_extract(meta_files, "PRJNA[0-9]+")

# function for reading in and uniting metadata
clean_meta_txt <- function(fpath, project_id) {
  df <- read_tsv(fpath, col_names = FALSE) %>% 
    rename(run_accession = X1) %>% 
    unite(-run_accession, sep = " | ", col = "description") %>% 
    mutate(bio_project = project_id) %>% 
    select(run_accession, bio_project, description)
  
  df
}

meta_df <- mapply(clean_meta_txt, 
                  meta_files, 
                  names(meta_files), 
                  SIMPLIFY = FALSE) %>% 
  bind_rows()

# write out
write_tsv(meta_df, here("results", "combined", "data-files", "run-metadata.tsv"))
```

## Read in the manual sample-level annotation

```{r}
sample_metadata <- fread(here("doc", "2021-04-20_run-metadata.tsv"), sep = "\t")

# update sample name to match matrix sample names
sample_metadata[, sample_name := paste0(bio_project, "__", sample_name)]

# include only those that are in the count matrix
sample_metadata <- sample_metadata[sample_name %in% colnames(wide_gene_lcpm_dt)]

saveRDS(sample_metadata, here("results", "combined", "rds-files", "sample-metadata.rds"))
```

## Model the experiment as a batch effect and attempt to correct 

```{r}
library(limma)


gene_mat <- as.matrix(wide_gene_lcpm_dt, rownames = "feature")
re_mat <- as.matrix(wide_re_lcpm_dt, rownames = "feature")

saveRDS(gene_mat, here("results", "combined", "rds-files", "gene-lcpms.rds"))
saveRDS(re_mat, here("results", "combined", "rds-files", "re-lcpms.rds"))

# get indeces of project with multiple batches
idxs <- match(sample_metadata[bio_project == "PRJNA379584"]$sample_name, colnames(gene_mat))

# add the per-project batch as a new vector (batch info stored in comment column)
replacements <- sample_metadata[bio_project == "PRJNA379584", 
                                .(bio_project2 = paste0(bio_project, ".", comment, "__", run_accession))]

# create a vector with project id + project batch added
new_ids <- replace(colnames(gene_mat), idxs, replacements$bio_project2)

# extract Project Id + project batch for batch correction
batches <- factor(sub("__.*", "", new_ids))

# regress away with limma
gene_mat.c <- removeBatchEffect(gene_mat, batch = batches)
re_mat.c <- removeBatchEffect(re_mat, batch = batches)

saveRDS(gene_mat.c, here("results", "combined", "rds-files", "batchCorrected-gene-lcpms.rds"))
saveRDS(re_mat.c, here("results", "combined", "rds-files", "batchCorrected-re-lcpms.rds"))
```

## PCA

```{r}
library(PCAtools)


# convert to data.frame for PCAtools
sample_metadata <- data.frame(as.matrix(sample_metadata, rownames = "sample_name"))

# match metadata rows with gene_mat cols
sample_metadata <- sample_metadata[colnames(gene_mat.c), ]

# check
all(rownames(sample_metadata) == colnames(gene_mat.c))

# Run PCA
pca.genes <- pca(gene_mat.c, metadata = sample_metadata, center = TRUE, scale = FALSE)
pca.res <- pca(re_mat.c, metadata = sample_metadata, center = TRUE, scale = FALSE)
```

### PCA plots

```{r}
screeplot(pca.genes, 
          components = getComponents(pca.genes)[1:20],
          title = "Protein Coding Genes w/Outliers") +
  ggsave(here("results", "combined", "figures", "screeplot-gene.png"),
         width = 8,
         height = 5)

biplot(pca.genes, 
       showLoadings = TRUE,
       lab = NULL,
       colby = "bio_project", 
       title = "Protein Coding Genes w/Outliers",
       subtitle = "Colored by BioProject",
       vline = 0,
       vlineType = 2,
       hline = 0,
       hlineType = 2) +
  ggsave(here("results", "combined", "figures", "pca-gene.png"),
         width = 8,
         height = 5)

# what are the outlier samples?
outlier_samples <- rownames(subset(pca.genes$rotated, PC1 > 100))

# run again without outliers
sample_metadata2 <- subset(sample_metadata, !rownames(sample_metadata) %in% outlier_samples)
gene_mat.c2 <- subset(gene_mat.c, select = -(which(colnames(gene_mat.c) %in% outlier_samples)))

pca.genes2 <- pca(gene_mat.c2, metadata = sample_metadata2, center = TRUE, scale = FALSE)

screeplot(pca.genes2, 
          components = getComponents(pca.genes2)[1:20],
          title = "Protein Coding Genes w/o Outliers") +
  ggsave(here("results", "combined", "figures", "screeplot-gene-noOutliers.png"),
         width = 8,
         height = 5)

biplot(pca.genes2, 
       showLoadings = TRUE,
       lab = NULL,
       colby = "drug", 
       title = "Protein Coding Genes w/o Outliers",
       subtitle = "Colored by Drug",
       vline = 0,
       vlineType = 2,
       hline = 0,
       hlineType = 2) +
  ggsave(here("results", "combined", "figures", "pca-gene-noOutliers.png"),
         width = 8,
         height = 5)

# REs --------------------------------------------------------------------------
screeplot(pca.res, 
          components = getComponents(pca.res)[1:20],
          title = "REs") +
  ggsave(here("results", "combined", "figures", "screeplot-re.png"),
         width = 8,
         height = 5)

biplot(pca.res, 
       showLoadings = TRUE,
       lab = NULL,
       colby = "bio_project", 
       title = "RE",
       subtitle = "Colored by BioProject",
       vline = 0,
       vlineType = 2,
       hline = 0,
       hlineType = 2) +
  ggsave(here("results", "combined", "figures", "pca-re.png"),
         width = 8,
         height = 5)
```

## Interactive plots with plotly

```{r}
library(plotly)


pca_gene_dt <- as.data.table(pca.genes2$rotated, keep.rownames = "sample_name")
sample_metadata2 <- as.data.table(sample_metadata2, keep.rownames = "sample_name")

gene_plot_dt <- merge(x = pca_gene_dt[, sample_name:PC10],
                      y = sample_metadata2,
                      by.x = "sample_name",
                      by.y = "sample_name",
                      all.x = TRUE)

pca_genebySample <- plot_ly(gene_plot_dt,
        x = ~PC1,
        y = ~PC2, 
        z = ~PC3, 
        color = ~drug,
        colors = "Set1",
        hovertext = paste("BioProject:", gene_plot_dt$bio_project,
                         "<br> Cell Type:", gene_plot_dt$cell_type,
                         "<br> Drug:", gene_plot_dt$drug,
                         "<br> Dose", gene_plot_dt$dose,
                         "<br> Time", gene_plot_dt$time_hr)) %>%
  add_markers() %>% 
  layout(title = list(text = "Protein Coding Gene Expression"),
         scene = list(xaxis = list(title = "PC1"), yaxis = list(title = "PC2"), zaxis = list(title = "PC3")))

htmlwidgets::saveWidget(pca_genebySample, here("results", "combined", "figures", "pca-gene-noOutlier.html"))

# REs --------------------------------------------------------------------------
pca_re_dt <- as.data.table(pca.res$rotated, keep.rownames = "sample_name")
sample_metadata <- as.data.table(sample_metadata, keep.rownames = "sample_name")

re_plot_dt <- merge(x = pca_re_dt[, sample_name:PC10],
                    y = sample_metadata,
                    by.x = "sample_name",
                    by.y = "sample_name",
                    all.x = TRUE)

pca_rebySample <- plot_ly(re_plot_dt,
        x = ~PC1,
        y = ~PC2, 
        z = ~PC3, 
        color = ~drug,
        colors = "Dark2",
        hovertext = paste("BioProject:", re_plot_dt$bio_project,
                         "<br> Cell Type:", re_plot_dt$cell_type,
                         "<br> Drug:", re_plot_dt$drug,
                         "<br> Dose", re_plot_dt$dose,
                         "<br> Time", re_plot_dt$time_hr)) %>%
  add_markers() %>% 
  layout(title = list(text = "RE Expression"),
         scene = list(xaxis = list(title = "PC1"), yaxis = list(title = "PC2"), zaxis = list(title = "PC3")))

htmlwidgets::saveWidget(pca_rebySample, here("results", "combined", "figures", "pca-re.html"))
```

---

# Meta-analysis on differential expression data

## Read in DE data

```{r}
dge_files <- list.files(
  path = here("results"),
  pattern = "dge.tsv",
  recursive = TRUE,
  full.names = TRUE
)

dre_files <- list.files(
  path = here("results"),
  pattern = "dre.tsv",
  recursive = TRUE,
  full.names = TRUE
)
```

## Read into single data.tables

```{r}
dge <- as.data.table(vroom::vroom(dge_files,
  id = "fpath",
  delim = "\t",
  col_types = list(
    contrast = "c",
    feature_id = "c",
    logFC = "d",
    unshrunk.logFC = "d",
    logCPM = "d",
    PValue = "d",
    FDR = "d"
  )
))

dre <- as.data.table(vroom::vroom(dre_files,
  id = "fpath",
  delim = "\t",
  col_types = list(
    contrast = "c",
    feature_id = "c",
    logFC = "d",
    unshrunk.logFC = "d",
    logCPM = "d",
    PValue = "d",
    FDR = "d"
  )
))
```

## Clean and add variables

```{r}
# create an experiment variable
dge[, experiment := clean_fpath(fpath)]
dre[, experiment := clean_fpath(fpath)]

# combine the contrast and the experiment into a new column
# (used to differentiate same conditions in different experiments)
dge[, condition := paste(experiment, contrast, sep = ".")]
dre[, condition := paste(experiment, contrast, sep = ".")]

# write the results out (could be used in Shiny app as input)
fwrite(dge, here("results", "combined", "data-files", "dge-all-experiments.csv"))
fwrite(dre, here("results", "combined", "data-files", "dre-all-experiments.csv"))

# create an additional column based on the feature's ranking
dge[, rank := sign(logFC) * -log10(FDR)]
dre[, rank := sign(logFC) * -log10(FDR)]
dge[, rank2 := logFC * -log10(FDR)]
dre[, rank2 := logFC * -log10(FDR)]
```

## Write the unique conditions out for manual annotation

I need to annotate the drugs by class/type/ something so they can be 
differentiated better on plots.

**I will have to manually edit this file since its information is not readily
available when downloading the datasets from GEO**

```{r eval=FALSE, include=FALSE}
fwrite(unique(dge[, .(experiment, contrast)]), file = here("doc", "experiment-annotation.csv"))
```

## Cast DE logFC values into matrices

```{r}
dge_lfc_mat <- as.matrix(dcast(dge[, .(feature_id, condition, logFC)],
  feature_id ~ condition,
  value.var = "logFC"
), rownames = "feature_id")

dre_lfc_mat <- as.matrix(dcast(dre[, .(feature_id, condition, logFC)],
  feature_id ~ condition,
  value.var = "logFC"
), rownames = "feature_id")

# repeat for FDR values
dge_fdr_mat <- as.matrix(dcast(dge[, .(feature_id, condition, FDR)],
  feature_id ~ condition,
  value.var = "FDR"
), rownames = "feature_id")

dre_fdr_mat <- as.matrix(dcast(dre[, .(feature_id, condition, FDR)],
  feature_id ~ condition,
  value.var = "FDR"
), rownames = "feature_id")

# repeat for ranks
dge_rank_mat <- as.matrix(dcast(dge[, .(feature_id, condition, rank)],
  feature_id ~ condition,
  value.var = "rank"
), rownames = "feature_id")

dre_rank_mat <- as.matrix(dcast(dre[, .(feature_id, condition, rank)],
  feature_id ~ condition,
  value.var = "rank"
), rownames = "feature_id")
```

## Replace NA values after cast

LFC values that were NA get 0 for their logFC and FDR values that were NA get 1.
Ranks are trickier. -1/1 rank should put features near the middle. 

```{r}
dge_lfc_mat[is.na(dge_lfc_mat)] <- 0
dre_lfc_mat[is.na(dre_lfc_mat)] <- 0
dge_fdr_mat[is.na(dge_fdr_mat)] <- 1
dre_fdr_mat[is.na(dre_fdr_mat)] <- 1
dge_rank_mat[is.na(dge_rank_mat)] <- 1
dre_rank_mat[is.na(dre_rank_mat)] <- 1
```

## PCA on all conditions

logFC, p-value, and rank

```{r}
pca_dge_lfc <- prcomp(t(dge_lfc_mat), scale. = FALSE, center = TRUE)
pca_dre_lfc <- prcomp(t(dre_lfc_mat), scale. = FALSE, center = TRUE)
pca_dge_fdr <- prcomp(t(dge_fdr_mat), scale. = FALSE, center = TRUE)
pca_dre_fdr <- prcomp(t(dre_fdr_mat), scale. = FALSE, center = TRUE)
pca_dge_rank <- prcomp(t(dge_rank_mat), scale. = FALSE, center = TRUE)
pca_dre_rank <- prcomp(t(dre_rank_mat), scale. = FALSE, center = TRUE)

# extract the components
dge_lfc_pca_df <- as.data.table(pca_dge_lfc$x, keep.rownames = "condition")
dre_lfc_pca_df <- as.data.table(pca_dre_lfc$x, keep.rownames = "condition")
dge_fdr_pca_df <- as.data.table(pca_dge_fdr$x, keep.rownames = "condition")
dre_fdr_pca_df <- as.data.table(pca_dre_fdr$x, keep.rownames = "condition")
dge_rank_pca_df <- as.data.table(pca_dge_rank$x, keep.rownames = "condition")
dre_rank_pca_df <- as.data.table(pca_dre_rank$x, keep.rownames = "condition")
```

### Get the percent variance explained

```{r}
factoextra::fviz_eig(pca_dge_lfc, addlabels = TRUE)
factoextra::fviz_eig(pca_dge_rank, addlabels = TRUE)
factoextra::fviz_eig(pca_dre_lfc, addlabels = TRUE)
factoextra::fviz_eig(pca_dre_rank, addlabels = TRUE)
```


## Join in manual annotation information

```{r}
exp_annot <- fread(here("doc", "2021_03_17-experiment-annotation.csv"))
exp_annot[, condition := paste(experiment, contrast, sep = ".")]
```

## Create PCA plots

```{r}
library(plotly)


df <- merge(dge_lfc_pca_df, exp_annot, by = "condition")
fig1 <- plot_ly(df, 
               x = ~PC1, 
               y = ~PC2, 
               z = ~PC3, 
               color = ~drug,
               colors = "Set1",
               hovertext = paste("Cell Type:", df$cell_type, 
                                 "<br> Drug:", df$drug,
                                 "<br> Dose", df$dose,
                                 "<br> Time", df$time_hr)) %>%
  add_markers() %>% 
  layout(
    title = list(text = "Protein Coding Gene Expression - logFC"),
    scene = list(
      xaxis = list(title = "PC1 (12.1%)"),
      yaxis = list(title = "PC2 (10.1%)"),
      zaxis = list(title = "PC3 (4.5%)")))

# do the same for REs ----------------------------------------------------------
df2 <- merge(dre_lfc_pca_df, exp_annot, by = "condition")
fig2 <- plot_ly(df2, 
               x = ~PC1, 
               y = ~PC2, 
               z = ~PC3, 
               color = ~drug,
               colors = "Set1",
               hovertext = paste("Cell Type:", df2$cell_type, 
                                 "<br> Drug:", df2$drug,
                                 "<br> Dose", df2$dose,
                                 "<br> Time", df2$time_hr)) %>%
  add_markers() %>% 
  layout(
    title = list(text = "RE Expression - logFC"),
    scene = list(
      xaxis = list(title = "PC1 (34.8%)"),
      yaxis = list(title = "PC2 (5.3%)"),
      zaxis = list(title = "PC3 (3.1%)")))
```

Same deal but use rankings

```{r}
df3 <- merge(dge_rank_pca_df, exp_annot, by = "condition")
fig3 <- plot_ly(df3, 
               x = ~PC1, 
               y = ~PC2, 
               z = ~PC3, 
               color = ~drug,
               colors = "Set1",
               hovertext = paste("Cell Type:", df3$cell_type, 
                                 "<br> Drug:", df3$drug,
                                 "<br> Dose", df3$dose,
                                 "<br> Time", df3$time_hr)) %>%
  add_markers() %>% 
    layout(
    title = list(text = "Protein Coding Gene Expression - Gene Ranks"),
    scene = list(
      xaxis = list(title = "PC1 (23.3%)"),
      yaxis = list(title = "PC2 (18.1%)"),
      zaxis = list(title = "PC3 (8.8%%)")))

# same with REs ----------------------------------------------------------------
df4 <- merge(dre_rank_pca_df, exp_annot, by = "condition")
fig4 <- plot_ly(df4, 
               x = ~PC1, 
               y = ~PC2, 
               z = ~PC3, 
               color = ~drug,
               colors = "Set1",
               hovertext = paste("Cell Type:", df4$cell_type, 
                                 "<br> Drug:", df4$drug,
                                 "<br> Dose", df4$dose,
                                 "<br> Time", df4$time_hr)) %>%
  add_markers() %>% 
    layout(
    title = list(text = "RE Expression - Gene Ranks"),
    scene = list(
      xaxis = list(title = "PC1 (66.4%)"),
      yaxis = list(title = "PC2 (5.6%)"),
      zaxis = list(title = "PC3 (4.8%%)")))
```

## Save plots

```{r message=FALSE, warning=FALSE}
htmlwidgets::saveWidget(fig1, here("results", "combined", "figures", "dge-lfc-pca.html"))
htmlwidgets::saveWidget(fig2, here("results", "combined", "figures", "dre-lfc-pca.html"))
htmlwidgets::saveWidget(fig3, here("results", "combined", "figures", "dge-rank-pca.html"))
htmlwidgets::saveWidget(fig4, here("results", "combined", "figures", "dre-rank-pca.html"))
```

## Dendrogram / Heatmap

### Heatmap of most variable logFCs

```{r}
library(matrixStats)


# calculate unsupervised variance across logFC values
dge_lfc_vars <- rowVars(dge_lfc_mat, na.rm = TRUE)
dge_lfc_o <- order(dge_lfc_vars, decreasing = TRUE)
dge_lfc_top <- dge_lfc_mat[head(dge_lfc_o, n = 500), ]

# cluster rows and columns
dd_rows <- as.dendrogram(hclust(dist(dge_lfc_top, method = "euclidean"), method = "ward.D2"))
dd_cols <- as.dendrogram(hclust(dist(t(dge_lfc_top), method = "euclidean"), method = "ward.D2"))

# reorder the matrix by calculated clustering
dge_lfc_top <- dge_lfc_top[order.dendrogram(dd_rows), order.dendrogram(dd_cols)]

# match colnames from matrix with annotations
hm_annot <- exp_annot[match(colnames(dge_lfc_top), condition), ]

fig5 <- plot_ly(x = hm_annot$condition,
                y = rownames(dge_lfc_top),
                z = dge_lfc_top, 
                type = "heatmap") %>% 
  layout(title = list(text = "500 Most Variable Genes by logFC"))
```

Repeat for repeats

```{r}
# calculate unsupervised variance across logFC values
dre_lfc_vars <- rowVars(dre_lfc_mat, na.rm = TRUE)
dre_lfc_o <- order(dre_lfc_vars, decreasing = TRUE)
dre_lfc_top <- dre_lfc_mat[head(dre_lfc_o, n = 100), ]

# cluster rows and columns
dd_rows2 <- as.dendrogram(hclust(dist(dre_lfc_top, method = "euclidean"), method = "ward.D2"))
dd_cols2 <- as.dendrogram(hclust(dist(t(dre_lfc_top), method = "euclidean"), method = "ward.D2"))

# reorder the matrix by calculated clustering
dre_lfc_top <- dre_lfc_top[order.dendrogram(dd_rows2), order.dendrogram(dd_cols2)]

# match colnames from matrix with annotations
hm_annot2 <- exp_annot[match(colnames(dre_lfc_top), condition), ]

fig6 <- plot_ly(x = hm_annot2$condition,
                y = rownames(dre_lfc_top),
                z = dre_lfc_top, 
                type = "heatmap") %>% 
  layout(title = list(text = "100 Most Variable REs by logFC"))
```

and for calculated ranks

```{r}
# calculate unsupervised variance across logFC values
dge_rank_vars <- rowVars(dge_rank_mat, na.rm = TRUE)
dge_rank_o <- order(dge_rank_vars, decreasing = TRUE)
dge_rank_top <- dge_rank_mat[head(dge_rank_o, n = 500), ]

# cluster rows and columns
dd_rows3 <- as.dendrogram(hclust(dist(dge_rank_top, method = "euclidean"), method = "ward.D2"))
dd_cols3 <- as.dendrogram(hclust(dist(t(dge_rank_top), method = "euclidean"), method = "ward.D2"))

# reorder the matrix by calculated clustering
dge_rank_top <- dge_rank_top[order.dendrogram(dd_rows3), order.dendrogram(dd_cols3)]

# match colnames from matrix with annotations
hm_annot3 <- exp_annot[match(colnames(dge_rank_top), condition), ]

fig7 <- plot_ly(x = hm_annot3$condition,
                y = rownames(dge_rank_top),
                z = dge_rank_top, 
                type = "heatmap") %>% 
  layout(title = list(text = "500 Most Variable Genes by Rank"))
```

repeats by rank
 
```{r}
# calculate unsupervised variance across logFC values
dre_rank_vars <- rowVars(dre_rank_mat, na.rm = TRUE)
dre_rank_o <- order(dre_rank_vars, decreasing = TRUE)
dre_rank_top <- dre_rank_mat[head(dre_rank_o, n = 100), ]

# cluster rows and columns
dd_rows4 <- as.dendrogram(hclust(dist(dre_rank_top, method = "euclidean"), method = "ward.D2"))
dd_cols4 <- as.dendrogram(hclust(dist(t(dre_rank_top), method = "euclidean"), method = "ward.D2"))

# reorder the matrix by calculated clustering
dre_rank_top <- dre_rank_top[order.dendrogram(dd_rows4), order.dendrogram(dd_cols4)]

# match colnames from matrix with annotations
hm_annot4 <- exp_annot[match(colnames(dre_rank_top), condition), ]

fig8 <- plot_ly(x = hm_annot4$condition,
                y = rownames(dre_rank_top),
                z = dre_rank_top, 
                type = "heatmap") %>% 
  layout(title = list(text = "100 Most Variable REs by Rank"))
```

## Save widgets

```{r message=FALSE, warning=FALSE}
htmlwidgets::saveWidget(fig5, here("results", "combined", "figures", "dge-lfc-hm.html"))
htmlwidgets::saveWidget(fig6, here("results", "combined", "figures", "dge-rank-hm.html"))
htmlwidgets::saveWidget(fig7, here("results", "combined", "figures", "dre-lfc-hm.html"))
htmlwidgets::saveWidget(fig8, here("results", "combined", "figures", "dre-rank-hm.html"))
```

## Correlation analysis

```{r message=FALSE, warning=FALSE}
dge_rank_cor_mat <- cor(dge_rank_mat)
dd_rows5 <- as.dendrogram(hclust(dist(dge_rank_cor_mat, method = "euclidean"), method = "ward.D2"))
dd_cols5 <- as.dendrogram(hclust(dist(t(dge_rank_cor_mat), method = "euclidean"), method = "ward.D2"))
dge_rank_cor_mat <- dge_rank_cor_mat[order.dendrogram(dd_rows5), order.dendrogram(dd_cols5)]

fig9 <- plot_ly(z = dge_rank_cor_mat, 
        x = rownames(dge_rank_cor_mat),
        y = colnames(dge_rank_cor_mat),
        type = "heatmap") %>% 
  layout(title = list(text = "Correlation by Gene Rank"))

htmlwidgets::saveWidget(fig9, here("results", "combined", "figures", "dge-rank-corr.html"))
```

