---
title: "Combine Datasets"
output:
  html_document:
    df_print: paged
    highlight: tango
    theme: united
    toc: yes
    toc_depth: 4
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Meta-analysis on differential expression data

## Read in DE data

```{r}
library(here)


dge_files <- list.files(
  path = here("results"),
  pattern = "dge.tsv",
  recursive = TRUE,
  full.names = TRUE
)

dre_files <- list.files(
  path = here("results"),
  pattern = "dre.tsv",
  recursive = TRUE,
  full.names = TRUE
)
```

## Read into single data.tables

```{r}
library(data.table)


dge <- setDT(vroom::vroom(dge_files,
  id = "fpath",
  delim = "\t",
  col_types = list(
    contrast = "c",
    feature_id = "c",
    logFC = "d",
    unshrunk.logFC = "d",
    logCPM = "d",
    PValue = "d",
    FDR = "d"
  )
))

dre <- setDT(vroom::vroom(dre_files,
  id = "fpath",
  delim = "\t",
  col_types = list(
    contrast = "c",
    feature_id = "c",
    logFC = "d",
    unshrunk.logFC = "d",
    logCPM = "d",
    PValue = "d",
    FDR = "d"
  )
))
```

## Clean and add variables

```{r}
# function for extracting the experiment name from the filepath
clean_fpath = function(s) { gsub("/data-files/", "", 
                            gsub("/results/", "", 
                            regmatches(s, 
                            regexpr("\\/results\\/.*\\/", s)))) 
}

# create an experiment variable
dge[, experiment := clean_fpath(fpath)]
dre[, experiment := clean_fpath(fpath)]

# combine the contrast and the experiment into a new column
# (used to differentiate same conditions in different experiments)
dge[, condition := paste(experiment, contrast, sep = ".")]
dre[, condition := paste(experiment, contrast, sep = ".")]

# Filter REs to be same sets (Satellites and Simple repeats were removed from most analyses)
dre <- dre[!feature_id %like% "Satellite|Simple_repeat|Unknown|Low_complexity"]

# write the results out after removing fpath column
dge[, fpath := NULL]
dre[, fpath := NULL]
fwrite(dge, here("results", "combined", "data-files", "dge-all-experiments.csv"))
fwrite(dre, here("results", "combined", "data-files", "dre-all-experiments.csv"))

# create an additional column based on the feature's ranking within each condition
dge[, rank := frank(logFC, ties.method = "random"), by = condition]
dre[, rank := frank(logFC, ties.method = "random"), by = condition]
dge[, max_rank := max(rank), by = condition]
dre[, max_rank := max(rank), by = condition]
dge[, norm_rank := (rank - 1) / (max_rank - 1), by = condition]
dre[, norm_rank := (rank - 1) / (max_rank - 1), by = condition]
```

## Write the unique conditions out for manual annotation

I need to annotate the drugs by class/type/ something so they can be 
differentiated better on plots.

**I will have to manually edit this file since its information is not readily
available when downloading the datasets from GEO**

```{r}
fwrite(unique(dge[, .(experiment, contrast)]), file = here("doc", "experiment-annotation.csv"))
```

## After manual editing, read in the contrast metadata

```{r}
metadata <- fread(here("doc", "2021-06-18-experiment-annotation.csv"))
metadata[, id := paste(experiment, contrast, sep = ".")]

head(metadata)
```

## Cast DE values into matrices

```{r}
dge_lfc_mat <- as.matrix(dcast(dge[, .(feature_id, condition, logFC)],
  feature_id ~ condition,
  value.var = "logFC"
), rownames = "feature_id")

dre_lfc_mat <- as.matrix(dcast(dre[, .(feature_id, condition, logFC)],
  feature_id ~ condition,
  value.var = "logFC"
), rownames = "feature_id")

# repeat for FDR values
dge_fdr_mat <- as.matrix(dcast(dge[, .(feature_id, condition, FDR)],
  feature_id ~ condition,
  value.var = "FDR"
), rownames = "feature_id")

dre_fdr_mat <- as.matrix(dcast(dre[, .(feature_id, condition, FDR)],
  feature_id ~ condition,
  value.var = "FDR"
), rownames = "feature_id")

# repeat for normalized ranks
dge_rank_mat <- as.matrix(dcast(dge[, .(feature_id, condition, norm_rank)],
  feature_id ~ condition,
  value.var = "norm_rank"
), rownames = "feature_id")

dre_rank_mat <- as.matrix(dcast(dre[, .(feature_id, condition, norm_rank)],
  feature_id ~ condition,
  value.var = "norm_rank"
), rownames = "feature_id")
```

## Replace NA values after cast

LFC values that were NA get 0 for their logFC and FDR values that were NA get 1.
Ranks have been imputed with the median rank per condition. For normalized ranks
place NAs directly in the middle (0.5)

```{r}
dge_lfc_mat[is.na(dge_lfc_mat)] <- 0
dre_lfc_mat[is.na(dre_lfc_mat)] <- 0
dge_fdr_mat[is.na(dge_fdr_mat)] <- 1
dre_fdr_mat[is.na(dre_fdr_mat)] <- 1
dge_rank_mat[is.na(dge_rank_mat)] <- 0.5
dre_rank_mat[is.na(dre_rank_mat)] <- 0.5

# impute the median rank per column for NA values
# for (i in 1:ncol(dge_rank_mat)) {
#   imputed_val <- median(dge_rank_mat[, i], na.rm = TRUE)
#   dge_rank_mat[, i][is.na(dge_rank_mat[, i])] <- imputed_val
# }
# 
# for (i in 1:ncol(dre_rank_mat)) {
#   imputed_val <- median(dre_rank_mat[, i], na.rm = TRUE)
#   dre_rank_mat[, i][is.na(dre_rank_mat[, i])] <- imputed_val
# }
```

## Reorder metadata by matrix colnames

```{r}
metadata <- metadata[id %chin% colnames(dge_lfc_mat)][order(match(id, colnames(dge_lfc_mat)))]

# check if all rownames and colnames align
all(metadata$id == colnames(dre_lfc_mat))

# convert to data.frame with rownames for PCAtools
metadata <- data.frame(as.matrix(metadata, rownames = "id"))
```

## Combine into a SummarizedExperiment object

This will help with managing filtering the datasets

```{r}
library(SummarizedExperiment)


dge_se <- SummarizedExperiment(
  assays = list(lfc = dge_lfc_mat, fdr = dge_fdr_mat, rank = dge_rank_mat),
  colData = metadata)

dre_se <- SummarizedExperiment(
  assays = list(lfc = dre_lfc_mat, fdr = dre_lfc_mat, rank = dre_rank_mat),
  colData = metadata)

# save objects out to be read into interactive visualization
saveRDS(dge_se, here("results", "combined", "rds-files", "dge_se.rds"), compress = FALSE)
saveRDS(dre_se, here("results", "combined", "rds-files", "dre_se.rds"), compress = FALSE)
```

----

# Analysis Below is Replaced By GEO RNA-seq app

## Create a function for assigning random colors

```{r}
random_rgb <- function(n, transparency = 0.8) {
  rgb(red = runif(n), green = runif(n), blue = runif(n), alpha = transparency)
}
```

## Plot all experiments together

```{r}
library(PCAtools)


pca.dge_lfc <- pca(dge_lfc_mat, metadata, center = TRUE, scale = FALSE, removeVar = 0.1)
pca.dre_lfc <- pca(dre_lfc_mat, metadata, center = TRUE, scale = FALSE, removeVar = 0.1)
pca.dge_rnk <- pca(dge_rank_mat, metadata, center = TRUE, scale = FALSE, removeVar = 0.1)
pca.dre_rnk <- pca(dre_rank_mat, metadata, center = TRUE, scale = FALSE, removeVar = 0.1)

drug2_cols <- c("#9f4a6b", "#5eb847", "#6b61c9", "#a8b635", "#b565c8", "#56802f",
                "#d14395", "#808080", "#d24659", "#4cbcd1", "#d6542d", "#6986ca",
                "#d69935", "#d885b0", "#388864", "#aa5a34", "#acb064", "#e19670",
                "#846f2c")

biplot(pca.dge_lfc, 
       colby = "drug2",
       lab = gsub("NA", "", paste(
         pca.dge_lfc$metadata$drug, 
         pca.dge_lfc$metadata$dose, 
         pca.dge_lfc$metadata$time_hr)
         ),
       colkey = drug2_cols,
       labSize = 2.5,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 15,
       title = "All Experiments",
       subtitle = "Protein Coding Gene Log-Fold Change Relative to Control"
       ) +
  ggsave(here("results", "combined", "figures", "biplot-all-dge-lfc.png"),
         width = 12,
         height = 7)

biplot(pca.dre_lfc, 
       colby = "drug2",
       lab = gsub("NA", "", paste(
         pca.dge_lfc$metadata$drug, 
         pca.dge_lfc$metadata$dose, 
         pca.dge_lfc$metadata$time_hr)
         ),
       colkey = drug2_cols,
       labSize = 2.5,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 15,
       title = "All Experiments",
       subtitle = "Repetitive Element Log-Fold Change Relative to Control"
       ) +
  ggsave(here("results", "combined", "figures", "biplot-all-dre-lfc.png"),
         width = 12,
         height = 7)

biplot(pca.dge_rnk, 
       colby = "drug2",
       lab = gsub("NA", "", paste(
         pca.dge_lfc$metadata$drug, 
         pca.dge_lfc$metadata$dose, 
         pca.dge_lfc$metadata$time_hr)
         ),
       colkey = drug2_cols,
       labSize = 2.5,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 15,
       title = "All Experiments",
       subtitle = "Protein Coding Gene Differential Expression Rank"
       ) +
  ggsave(here("results", "combined", "figures", "biplot-all-dge-rank.png"),
         width = 12,
         height = 7)

biplot(pca.dre_rnk, 
       colby = "drug2",
       lab = gsub("NA", "", paste(
         pca.dge_lfc$metadata$drug, 
         pca.dge_lfc$metadata$dose, 
         pca.dge_lfc$metadata$time_hr)
         ),
       colkey = drug2_cols,
       labSize = 2.5,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 15,
       title = "All Experiments",
       subtitle = "Repetitive Element Differential Expression Rank"
       ) +
  ggsave(here("results", "combined", "figures", "biplot-all-dre-rank.png"),
         width = 12,
         height = 7)
```


## Colon Cell lines

Remove Decitabine + combinations from downstream PCAs

```{r}
colon_dge <- dge_se[, dge_se$tissue %like% "[Cc]olon" & !dge_se$drug %like% "Decitabine \\+"]
colon_dre <- dre_se[, dre_se$tissue %like% "[Cc]olon" & !dre_se$drug %like% "Decitabine \\+"]

# general function for performing PCA on SummarizedExperiment 
do_pca <- function(se, assay_name, ...) {
  pca(mat = assay(se, assay_name),
      metadata = colData(se),
      ...)
}

pca.colon_dge_lfc <- do_pca(colon_dge, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.colon_dge_rnk <- do_pca(colon_dge, "rank", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.colon_dre_lfc <- do_pca(colon_dre, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.colon_dre_rnk <- do_pca(colon_dre, "rank", center = TRUE, scale = FALSE, removeVar = 0.05)
```

### Plot PCAs

```{r}
# random color palette for drugs in colon samples
colon_cols <- random_rgb(n = length(unique(pca.colon_dge_lfc$metadata$drug)))

# Protein Coding Genes - LFC
biplot(pca.colon_dge_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.colon_dge_lfc$metadata$drug, 
         pca.colon_dge_lfc$metadata$dose, 
         pca.colon_dge_lfc$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = colon_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 7,
       title = "Colon Cancer Cell Lines",
       subtitle = "Protein Coding Gene Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-colon-dge-lfc.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.colon_dre_lfc, 
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.colon_dge_lfc$metadata$drug, 
         pca.colon_dge_lfc$metadata$dose, 
         pca.colon_dge_lfc$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = colon_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 15,
       title = "Colon Cancer Cell Lines",
       subtitle = "Repetitive Element Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-colon-dre-lfc.png"),
         width = 12,
         height = 7)

# Protein Coding Genes - Rank
biplot(pca.colon_dge_rnk, 
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.colon_dge_rnk$metadata$drug, 
         pca.colon_dge_rnk$metadata$dose, 
         pca.colon_dge_rnk$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = colon_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
     legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 7,
       title = "Colon Cancer Cell Lines",
       subtitle = "Protein Coding Gene Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-colon-dge-rank.png"),
         width = 12,
         height = 7)

# RE - Rank
biplot(pca.colon_dre_rnk, 
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.colon_dge_rnk$metadata$drug, 
         pca.colon_dge_rnk$metadata$dose, 
         pca.colon_dge_rnk$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = colon_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 15,
       title = "Colon Cancer Cell Lines",
       subtitle = "Repetitive Element Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-colon-dre-rank.png"),
         width = 12,
         height = 7)
```

### Loadings Plots

```{r}
plotloadings(
  pca.colon_dge_lfc,
  rangeRetain = 0.01,
  labSize = 2,
  title = 'Colon Cancer Cell Lines',
  subtitle = 'Protein Coding Gene Log-Fold Change Relative to Control',
  caption = 'Top 1% variables',
  shape = 21,
  shapeSizeRange = c(1, 4),
  drawConnectors = TRUE) +
  ggsave(here("results", "combined", "figures", "loadings-colon-dge-lfc.png"),
         width = 8,
         height = 5)

plotloadings(
  pca.colon_dre_lfc,
  rangeRetain = 0.01,
  labSize = 2,
  title = 'Colon Cancer Cell Lines',
  subtitle = 'Repetitive Element Log-Fold Change Relative to Control',
  caption = 'Top 1% variables',
  shape = 21,
  shapeSizeRange = c(1, 4),
  drawConnectors = TRUE) +
  ggsave(here("results", "combined", "figures", "loadings-colon-dre-lfc.png"),
         width = 8,
         height = 5)

plotloadings(
  pca.colon_dge_rnk,
  rangeRetain = 0.01,
  labSize = 2,
  title = 'Colon Cancer Cell Lines',
  subtitle = 'Protein Coding Gene Differential Expression Rank',
  caption = 'Top 1% variables',
  shape = 21,
  shapeSizeRange = c(1, 4),
  drawConnectors = TRUE) +
  ggsave(here("results", "combined", "figures", "loadings-colon-dge-rank.png"),
         width = 8,
         height = 5)

plotloadings(
  pca.colon_dre_rnk,
  rangeRetain = 0.01,
  labSize = 2,
  title = 'Colon Cancer Cell Lines',
  subtitle = 'Repetitive Element Differential Expression Rank',
  caption = 'Top 1% variables',
  shape = 21,
  shapeSizeRange = c(1, 4),
  drawConnectors = TRUE) +
  ggsave(here("results", "combined", "figures", "loadings-colon-dre-rank.png"),
         width = 8,
         height = 5)
```

### 3D plots

```{r eval=FALSE, include=FALSE}
library(magrittr)
library(plotly)


colon_metadata_dt <- as.data.table(pca.colon_dge_lfc$metadata, keep.rownames = "condition")
pca.colon_dge_lfc_dt <- as.data.table(pca.colon_dge_lfc$rotated, keep.rownames = "condition")
pca.colon_dge_rnk_dt <- as.data.table(pca.colon_dge_rnk$rotated, keep.rownames = "condition")
pca.colon_dre_lfc_dt <- as.data.table(pca.colon_dre_lfc$rotated, keep.rownames = "condition")
pca.colon_dre_rnk_dt <- as.data.table(pca.colon_dre_rnk$rotated, keep.rownames = "condition")
pca.colon_dge_lfc_dt <- merge(pca.colon_dge_lfc_dt, colon_metadata_dt, by = "condition")
pca.colon_dge_rnk_dt <- merge(pca.colon_dge_rnk_dt, colon_metadata_dt, by = "condition")
pca.colon_dre_lfc_dt <- merge(pca.colon_dre_lfc_dt, colon_metadata_dt, by = "condition")
pca.colon_dre_rnk_dt <- merge(pca.colon_dre_rnk_dt, colon_metadata_dt, by = "condition")

# Protein Coding Genes - LFC
colon_3d.dge_lfc <- plot_ly(pca.colon_dge_lfc_dt, 
                            x = ~PC1,
                            y = ~PC2,
                            z = ~PC3,
                            color = ~drug,
                            colors = colon_cols) %>% 
  add_markers() %>% 
  layout(scene = list(xaxis = list(title = 'PC1 (31.6%)'),
                      yaxis = list(title = 'PC2 (10.05)'),
                      zaxis = list(title = 'PC3 (7.83%)')),
         title = "Protein Coding Gene Log-Fold Change Relative to Control")
```

## Bone

```{r}
bone_dge <- dge_se[, dge_se$tissue == "bone"]
bone_dre <- dre_se[, dre_se$tissue == "bone"]

pca.bone_dge_lfc <- do_pca(bone_dge, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.bone_dge_rnk <- do_pca(bone_dge, "rank", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.bone_dre_lfc <- do_pca(bone_dre, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.bone_dre_rnk <- do_pca(bone_dre, "rank", center = TRUE, scale = FALSE, removeVar = 0.05)
```

### Plot Biplots

```{r}
bone_cols <- random_rgb(length(unique(pca.bone_dge_lfc$metadata$drug)))

# Protein Coding Genes - LFC
biplot(pca.bone_dge_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.bone_dge_lfc$metadata$drug, 
         pca.bone_dge_lfc$metadata$dose, 
         pca.bone_dge_lfc$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = bone_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Bone Cell Lines",
       subtitle = "Protein Coding Gene Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-bone-dge-lfc.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.bone_dre_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.bone_dre_lfc$metadata$drug, 
         pca.bone_dre_lfc$metadata$dose, 
         pca.bone_dre_lfc$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = bone_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Bone Cell Lines",
       subtitle = "Repetitive Element Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-bone-dre-lfc.png"),
         width = 12,
         height = 7)

# Protein Coding Genes - Rank
biplot(pca.bone_dge_rnk,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.bone_dge_rnk$metadata$drug, 
         pca.bone_dge_rnk$metadata$dose, 
         pca.bone_dge_rnk$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = bone_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Bone Cell Lines",
       subtitle = "Protein Coding Gene Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-bone-dge-rank.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.bone_dre_rnk,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.bone_dre_rnk$metadata$drug, 
         pca.bone_dre_rnk$metadata$dose, 
         pca.bone_dre_rnk$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = bone_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Bone Cell Lines",
       subtitle = "Repetitive Element Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-bone-dre-rank.png"),
         width = 12,
         height = 7)
```

## Blood

```{r}
blood_dge <- dge_se[, dge_se$tissue == "blood"]
blood_dre <- dre_se[, dre_se$tissue == "blood"]

pca.blood_dge_lfc <- do_pca(blood_dge, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.blood_dge_rnk <- do_pca(blood_dge, "rank", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.blood_dre_lfc <- do_pca(blood_dre, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.blood_dre_rnk <- do_pca(blood_dre, "rank", center = TRUE, scale = FALSE, removeVar = 0.05)
```

### Plot Biplots

```{r}
blood_cols <- random_rgb(length(unique(pca.blood_dge_lfc$metadata$drug)))

# Protein Coding Genes - LFC
biplot(pca.blood_dge_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.blood_dge_lfc$metadata$drug, 
         pca.blood_dge_lfc$metadata$dose, 
         pca.blood_dge_lfc$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = blood_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 15,
       title = "Blood Cell Lines",
       subtitle = "Protein Coding Gene Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-blood-dge-lfc.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.blood_dre_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.blood_dre_lfc$metadata$drug, 
         pca.blood_dre_lfc$metadata$dose, 
         pca.blood_dre_lfc$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = blood_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Blood Cell Lines",
       subtitle = "Repetitive Element Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-blood-dre-lfc.png"),
         width = 12,
         height = 7)

# Protein Coding Genes - Rank
biplot(pca.blood_dge_rnk,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.blood_dge_rnk$metadata$drug, 
         pca.blood_dge_rnk$metadata$dose, 
         pca.blood_dge_rnk$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = blood_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Blood Cell Lines",
       subtitle = "Protein Coding Gene Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-blood-dge-rank.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.blood_dre_rnk,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.blood_dre_rnk$metadata$drug, 
         pca.blood_dre_rnk$metadata$dose, 
         pca.blood_dre_rnk$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = blood_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Blood Cell Lines",
       subtitle = "Repetitive Element Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-blood-dre-rank.png"),
         width = 12,
         height = 7)
```

## Breast

```{r}
breast_dge <- dge_se[, dge_se$tissue == "breast"]
breast_dre <- dre_se[, dre_se$tissue == "breast"]

pca.breast_dge_lfc <- do_pca(breast_dge, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.breast_dge_rnk <- do_pca(breast_dge, "rank", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.breast_dre_lfc <- do_pca(breast_dre, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.breast_dre_rnk <- do_pca(breast_dre, "rank", center = TRUE, scale = FALSE, removeVar = 0.05)
```

### Plot Biplots

```{r}
breast_cols <- random_rgb(length(unique(pca.breast_dge_lfc$metadata$drug)))

# Protein Coding Genes - LFC
biplot(pca.breast_dge_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.breast_dge_lfc$metadata$drug, 
         pca.breast_dge_lfc$metadata$dose, 
         pca.breast_dge_lfc$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = breast_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 15,
       title = "Breast Cell Lines",
       subtitle = "Protein Coding Gene Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-breast-dge-lfc.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.breast_dre_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.breast_dre_lfc$metadata$drug, 
         pca.breast_dre_lfc$metadata$dose, 
         pca.breast_dre_lfc$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = breast_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Breast Cell Lines",
       subtitle = "Repetitive Element Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-breast-dre-lfc.png"),
         width = 12,
         height = 7)

# Protein Coding Genes - Rank
biplot(pca.breast_dge_rnk,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.breast_dge_rnk$metadata$drug, 
         pca.breast_dge_rnk$metadata$dose, 
         pca.breast_dge_rnk$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = breast_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Breast Cell Lines",
       subtitle = "Protein Coding Gene Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-breast-dge-rank.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.breast_dre_rnk,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.breast_dre_rnk$metadata$drug, 
         pca.breast_dre_rnk$metadata$dose, 
         pca.breast_dre_rnk$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = breast_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Breast Cell Lines",
       subtitle = "Repetitive Element Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-breast-dre-rank.png"),
         width = 12,
         height = 7)
```

## Pancreas; small bowel; rectum

```{r}
bowel_dge <- dge_se[, dge_se$tissue %like% "pancreas"]
bowel_dre <- dre_se[, dre_se$tissue %like% "pancreas"]

pca.bowel_dge_lfc <- do_pca(bowel_dge, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.bowel_dge_rnk <- do_pca(bowel_dge, "rank", center = TRUE, scale = FALSE, removeVar = 0.25)
pca.bowel_dre_lfc <- do_pca(bowel_dre, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.bowel_dre_rnk <- do_pca(bowel_dre, "rank", center = TRUE, scale = FALSE, removeVar = 0.25)
```

### Plot Biplots

```{r}
bowel_cols <- random_rgb(length(unique(pca.bowel_dge_lfc$metadata$drug)))

# Protein Coding Genes - LFC
biplot(pca.bowel_dge_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.bowel_dge_lfc$metadata$drug, 
         pca.bowel_dge_lfc$metadata$time_hr)
         ),
       colkey = bowel_cols,
       labSize = 2.5,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 15,
       title = "Pancreas; Bowel; Rectum Cell Lines",
       subtitle = "Protein Coding Gene Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-bowel-dge-lfc.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.bowel_dre_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.bowel_dre_lfc$metadata$drug, 
         pca.bowel_dre_lfc$metadata$time_hr)
         ),
       colkey = bowel_cols,
       labSize = 2.5,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Pancreas; Bowel; Rectum Cell Lines",
       subtitle = "Repetitive Element Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-bowel-dre-lfc.png"),
         width = 12,
         height = 7)

# Protein Coding Genes - Rank
biplot(pca.bowel_dge_rnk,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.bowel_dge_rnk$metadata$drug,
         pca.bowel_dge_rnk$metadata$time_hr)
         ),
       colkey = bowel_cols,
       labSize = 2.5,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Pancreas; Bowel; Rectum Cell Lines",
       subtitle = "Protein Coding Gene Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-bowel-dge-rank.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.bowel_dre_rnk,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.bowel_dre_rnk$metadata$drug, 
         pca.bowel_dre_rnk$metadata$time_hr)
         ),
       colkey = bowel_cols,
       labSize = 2.5,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Pancreas; Bowel; Rectum Cell Lines",
       subtitle = "Repetitive Element Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-bowel-dre-rank.png"),
         width = 12,
         height = 7)
```

## Skin

```{r}
skin_dge <- dge_se[, dge_se$tissue %like% "skin" & !is.na(dge_se$drug)]
skin_dre <- dre_se[, dre_se$tissue %like% "skin" & !is.na(dre_se$drug)]

pca.skin_dge_lfc <- do_pca(skin_dge, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.skin_dge_rnk <- do_pca(skin_dge, "rank", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.skin_dre_lfc <- do_pca(skin_dre, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.skin_dre_rnk <- do_pca(skin_dre, "rank", center = TRUE, scale = FALSE, removeVar = 0.05)
```

### Plot Biplots

```{r}
skin_cols <- random_rgb(length(unique(pca.skin_dge_lfc$metadata$drug)))

# Protein Coding Genes - LFC
biplot(pca.skin_dge_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.skin_dge_lfc$metadata$drug, 
         pca.skin_dge_lfc$metadata$dose, 
         pca.skin_dge_lfc$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = skin_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 15,
       title = "Skin Cell Lines",
       subtitle = "Protein Coding Gene Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-skin-dge-lfc.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.skin_dre_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.skin_dre_lfc$metadata$drug, 
         pca.skin_dre_lfc$metadata$dose, 
         pca.skin_dre_lfc$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = skin_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Skin Cell Lines",
       subtitle = "Repetitive Element Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-skin-dre-lfc.png"),
         width = 12,
         height = 7)

# Protein Coding Genes - Rank
biplot(pca.skin_dge_rnk,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.skin_dge_rnk$metadata$drug, 
         pca.skin_dge_rnk$metadata$dose, 
         pca.skin_dge_rnk$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = skin_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Skin Cell Lines",
       subtitle = "Protein Coding Gene Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-skin-dge-rank.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.skin_dre_rnk,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.skin_dre_rnk$metadata$drug, 
         pca.skin_dre_rnk$metadata$dose, 
         pca.skin_dre_rnk$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = skin_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Skin Cell Lines",
       subtitle = "Repetitive Element Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-skin-dre-rank.png"),
         width = 12,
         height = 7)
```

## Lung

```{r}
lung_dge <- dge_se[, dge_se$tissue %like% "lung"]
lung_dre <- dre_se[, dre_se$tissue %like% "lung"]

pca.lung_dge_lfc <- do_pca(lung_dge, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.lung_dge_rnk <- do_pca(lung_dge, "rank", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.lung_dre_lfc <- do_pca(lung_dre, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.lung_dre_rnk <- do_pca(lung_dre, "rank", center = TRUE, scale = FALSE, removeVar = 0.05)
```

### Plot Biplots

```{r}
lung_cols <- random_rgb(length(unique(pca.lung_dge_lfc$metadata$drug)))

# Protein Coding Genes - LFC
biplot(pca.lung_dge_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.lung_dge_lfc$metadata$drug, 
         pca.lung_dge_lfc$metadata$dose, 
         pca.lung_dge_lfc$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = lung_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 15,
       title = "Lung Cell Lines",
       subtitle = "Protein Coding Gene Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-lung-dge-lfc.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.lung_dre_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.lung_dre_lfc$metadata$drug, 
         pca.lung_dre_lfc$metadata$dose, 
         pca.lung_dre_lfc$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = lung_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Lung Cell Lines",
       subtitle = "Repetitive Element Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-lung-dre-lfc.png"),
         width = 12,
         height = 7)

# Protein Coding Genes - Rank
biplot(pca.lung_dge_rnk,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.lung_dge_rnk$metadata$drug, 
         pca.lung_dge_rnk$metadata$dose, 
         pca.lung_dge_rnk$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = lung_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Lung Cell Lines",
       subtitle = "Protein Coding Gene Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-lung-dge-rank.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.lung_dre_rnk,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.lung_dre_rnk$metadata$drug, 
         pca.lung_dre_rnk$metadata$dose, 
         pca.lung_dre_rnk$metadata$time_hr)
         ),
       labSize = 2.5,
       colkey = lung_cols,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Lung Cell Lines",
       subtitle = "Repetitive Element Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-lung-dre-rank.png"),
         width = 12,
         height = 7)
```

## Kidney

```{r}
kidney_dge <- dge_se[, dge_se$tissue %like% "kidney"]
kidney_dre <- dre_se[, dre_se$tissue %like% "kidney"]

pca.kidney_dge_lfc <- do_pca(kidney_dge, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.kidney_dge_rnk <- do_pca(kidney_dge, "rank", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.kidney_dre_lfc <- do_pca(kidney_dre, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.kidney_dre_rnk <- do_pca(kidney_dre, "rank", center = TRUE, scale = FALSE, removeVar = 0.05)
```

### Plot Biplots

```{r}
kidney_cols <- random_rgb(length(unique(pca.kidney_dge_lfc$metadata$drug)))

# Protein Coding Genes - LFC
biplot(pca.kidney_dge_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.kidney_dge_lfc$metadata$drug, 
         pca.kidney_dge_lfc$metadata$dose, 
         pca.kidney_dge_lfc$metadata$time_hr)
         ),
       colkey = kidney_cols,
       labSize = 2.5,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 15,
       title = "Kidney Cell Lines",
       subtitle = "Protein Coding Gene Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-kidney-dge-lfc.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.kidney_dre_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.kidney_dre_lfc$metadata$drug, 
         pca.kidney_dre_lfc$metadata$dose, 
         pca.kidney_dre_lfc$metadata$time_hr)
         ),
       colkey = kidney_cols,
       labSize = 2.5,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Kidney Cell Lines",
       subtitle = "Repetitive Element Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-kidney-dre-lfc.png"),
         width = 12,
         height = 7)

# Protein Coding Genes - Rank
biplot(pca.kidney_dge_rnk,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.kidney_dge_rnk$metadata$drug, 
         pca.kidney_dge_rnk$metadata$dose, 
         pca.kidney_dge_rnk$metadata$time_hr)
         ),
       colkey = kidney_cols,
       labSize = 2.5,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Kidney Cell Lines",
       subtitle = "Protein Coding Gene Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-kidney-dge-rank.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.kidney_dre_rnk,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.kidney_dre_rnk$metadata$drug, 
         pca.kidney_dre_rnk$metadata$dose, 
         pca.kidney_dre_rnk$metadata$time_hr)
         ),
       colkey = kidney_cols,
       labSize = 2.5,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Kidney Cell Lines",
       subtitle = "Repetitive Element Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-kidney-dre-rank.png"),
         width = 12,
         height = 7)
```

## Leukemia specific

```{r}
leukemia_dge <- dge_se[, dge_se$disease %like% "leukemia"]
leukemia_dre <- dre_se[, dre_se$disease %like% "leukemia"]

pca.leukemia_dge_lfc <- do_pca(leukemia_dge, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.leukemia_dge_rnk <- do_pca(leukemia_dge, "rank", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.leukemia_dre_lfc <- do_pca(leukemia_dre, "lfc", center = TRUE, scale = FALSE, removeVar = 0.05)
pca.leukemia_dre_rnk <- do_pca(leukemia_dre, "rank", center = TRUE, scale = FALSE, removeVar = 0.05)
```

### Plot Biplots

```{r}
leukemia_cols <- random_rgb(length(unique(pca.leukemia_dge_lfc$metadata$drug)))

# Protein Coding Genes - LFC
biplot(pca.leukemia_dge_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.leukemia_dge_lfc$metadata$drug, 
         pca.leukemia_dge_lfc$metadata$dose, 
         pca.leukemia_dge_lfc$metadata$time_hr)
         ),
       colkey = leukemia_cols,
       labSize = 2.5,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 15,
       title = "Leukemia Cell Lines",
       subtitle = "Protein Coding Gene Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-leukemia-dge-lfc.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.leukemia_dre_lfc,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.leukemia_dre_lfc$metadata$drug, 
         pca.leukemia_dre_lfc$metadata$dose, 
         pca.leukemia_dre_lfc$metadata$time_hr)
         ),
       colkey = leukemia_cols,
       labSize = 2.5,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Leukemia Cell Lines",
       subtitle = "Repetitive Element Log-Fold Change Relative to Control") +
  ggsave(here("results", "combined", "figures", "biplot-leukemia-dre-lfc.png"),
         width = 12,
         height = 7)

# Protein Coding Genes - Rank
biplot(pca.leukemia_dge_rnk,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.leukemia_dge_rnk$metadata$drug, 
         pca.leukemia_dge_rnk$metadata$dose, 
         pca.leukemia_dge_rnk$metadata$time_hr)
         ),
       colkey = leukemia_cols,
       labSize = 2.5,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Leukemia Cell Lines",
       subtitle = "Protein Coding Gene Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-leukemia-dge-rank.png"),
         width = 12,
         height = 7)

# RE - LFC
biplot(pca.leukemia_dre_rnk,
       colby = "drug",
       lab = gsub("NA", "", paste(
         pca.leukemia_dre_rnk$metadata$drug, 
         pca.leukemia_dre_rnk$metadata$dose, 
         pca.leukemia_dre_rnk$metadata$time_hr)
         ),
       colkey = leukemia_cols,
       labSize = 2.5,
       hline = 0, 
       hlineType = 2, 
       vline = 0, 
       vlineType = 2,
       legendPosition = "bottom",
       legendLabSize = 8,
       legendIconSize = 2,
       legendTitleSize = 10,
       maxoverlapsConnectors = 10,
       title = "Leukemia Cell Lines",
       subtitle = "Repetitive Element Differential Expression Rank") +
  ggsave(here("results", "combined", "figures", "biplot-leukemia-dre-rank.png"),
         width = 12,
         height = 7)
```