---
title: "Combine Datasets"
output:
  html_document:
    df_print: paged
    highlight: tango
    theme: united
    toc: yes
    toc_depth: 4
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(here)
library(magrittr)
library(data.table)
```

# Meta-analysis on differential expression data

## Read in DE data

```{r}
dge_files <- list.files(
  path = here("results"),
  pattern = "dge.tsv",
  recursive = TRUE,
  full.names = TRUE
)

dre_files <- list.files(
  path = here("results"),
  pattern = "dre.tsv",
  recursive = TRUE,
  full.names = TRUE
)
```

## Read into single data.tables

```{r}
dge <- as.data.table(vroom::vroom(dge_files,
  id = "fpath",
  delim = "\t",
  col_types = list(
    contrast = "c",
    feature_id = "c",
    logFC = "d",
    unshrunk.logFC = "d",
    logCPM = "d",
    PValue = "d",
    FDR = "d"
  )
))

dre <- as.data.table(vroom::vroom(dre_files,
  id = "fpath",
  delim = "\t",
  col_types = list(
    contrast = "c",
    feature_id = "c",
    logFC = "d",
    unshrunk.logFC = "d",
    logCPM = "d",
    PValue = "d",
    FDR = "d"
  )
))
```

## Clean and add variables

```{r}
# create an experiment variable
dge[, experiment := clean_fpath(fpath)]
dre[, experiment := clean_fpath(fpath)]

# combine the contrast and the experiment into a new column
# (used to differentiate same conditions in different experiments)
dge[, condition := paste(experiment, contrast, sep = ".")]
dre[, condition := paste(experiment, contrast, sep = ".")]

# write the results out (could be used in Shiny app as input)
fwrite(dge, here("results", "combined", "data-files", "dge-all-experiments.csv"))
fwrite(dre, here("results", "combined", "data-files", "dre-all-experiments.csv"))

# create an additional column based on the feature's ranking
dge[, rank := sign(logFC) * -log10(FDR)]
dre[, rank := sign(logFC) * -log10(FDR)]
dge[, rank2 := logFC * -log10(FDR)]
dre[, rank2 := logFC * -log10(FDR)]

# remove fpath column
dge[, fpath := NULL]
dre[, fpath := NULL]
```

## Write the unique conditions out for manual annotation

I need to annotate the drugs by class/type/ something so they can be 
differentiated better on plots.

**I will have to manually edit this file since its information is not readily
available when downloading the datasets from GEO**

```{r eval=FALSE, include=FALSE}
fwrite(unique(dge[, .(experiment, contrast)]), file = here("doc", "experiment-annotation.csv"))
```

## Cast DE logFC values into matrices

```{r}
dge_lfc_mat <- as.matrix(dcast(dge[, .(feature_id, condition, logFC)],
  feature_id ~ condition,
  value.var = "logFC"
), rownames = "feature_id")

dre_lfc_mat <- as.matrix(dcast(dre[, .(feature_id, condition, logFC)],
  feature_id ~ condition,
  value.var = "logFC"
), rownames = "feature_id")

# repeat for FDR values
dge_fdr_mat <- as.matrix(dcast(dge[, .(feature_id, condition, FDR)],
  feature_id ~ condition,
  value.var = "FDR"
), rownames = "feature_id")

dre_fdr_mat <- as.matrix(dcast(dre[, .(feature_id, condition, FDR)],
  feature_id ~ condition,
  value.var = "FDR"
), rownames = "feature_id")

# repeat for ranks
dge_rank_mat <- as.matrix(dcast(dge[, .(feature_id, condition, rank)],
  feature_id ~ condition,
  value.var = "rank"
), rownames = "feature_id")

dre_rank_mat <- as.matrix(dcast(dre[, .(feature_id, condition, rank)],
  feature_id ~ condition,
  value.var = "rank"
), rownames = "feature_id")
```

## Replace NA values after cast

LFC values that were NA get 0 for their logFC and FDR values that were NA get 1.
Ranks are trickier. -1/1 rank should put features near the middle. 

```{r}
dge_lfc_mat[is.na(dge_lfc_mat)] <- 0
dre_lfc_mat[is.na(dre_lfc_mat)] <- 0
dge_fdr_mat[is.na(dge_fdr_mat)] <- 1
dre_fdr_mat[is.na(dre_fdr_mat)] <- 1
dge_rank_mat[is.na(dge_rank_mat)] <- 1
dre_rank_mat[is.na(dre_rank_mat)] <- 1
```

## PCA on all conditions

logFC, p-value, and rank

```{r}
pca_dge_lfc <- prcomp(t(dge_lfc_mat), scale. = FALSE, center = TRUE)
pca_dre_lfc <- prcomp(t(dre_lfc_mat), scale. = FALSE, center = TRUE)
pca_dge_fdr <- prcomp(t(dge_fdr_mat), scale. = FALSE, center = TRUE)
pca_dre_fdr <- prcomp(t(dre_fdr_mat), scale. = FALSE, center = TRUE)
pca_dge_rank <- prcomp(t(dge_rank_mat), scale. = FALSE, center = TRUE)
pca_dre_rank <- prcomp(t(dre_rank_mat), scale. = FALSE, center = TRUE)

# extract the components
dge_lfc_pca_df <- as.data.table(pca_dge_lfc$x, keep.rownames = "condition")
dre_lfc_pca_df <- as.data.table(pca_dre_lfc$x, keep.rownames = "condition")
dge_fdr_pca_df <- as.data.table(pca_dge_fdr$x, keep.rownames = "condition")
dre_fdr_pca_df <- as.data.table(pca_dre_fdr$x, keep.rownames = "condition")
dge_rank_pca_df <- as.data.table(pca_dge_rank$x, keep.rownames = "condition")
dre_rank_pca_df <- as.data.table(pca_dre_rank$x, keep.rownames = "condition")
```

### Get the percent variance explained

```{r}
factoextra::fviz_eig(pca_dge_lfc, addlabels = TRUE)
factoextra::fviz_eig(pca_dge_rank, addlabels = TRUE)
factoextra::fviz_eig(pca_dre_lfc, addlabels = TRUE)
factoextra::fviz_eig(pca_dre_rank, addlabels = TRUE)
```


## Join in manual annotation information

```{r}
exp_annot <- fread(here("doc", "2021_03_17-experiment-annotation.csv"))
exp_annot[, condition := paste(experiment, contrast, sep = ".")]
```

## Create PCA plots

```{r}
library(plotly)


df <- merge(dge_lfc_pca_df, exp_annot, by = "condition")
fig1 <- plot_ly(df, 
               x = ~PC1, 
               y = ~PC2, 
               z = ~PC3, 
               color = ~drug,
               colors = "Set1",
               hovertext = paste("Cell Type:", df$cell_type, 
                                 "<br> Drug:", df$drug,
                                 "<br> Dose", df$dose,
                                 "<br> Time", df$time_hr)) %>%
  add_markers() %>% 
  layout(
    title = list(text = "Protein Coding Gene Expression - logFC"),
    scene = list(
      xaxis = list(title = "PC1 (12.1%)"),
      yaxis = list(title = "PC2 (10.1%)"),
      zaxis = list(title = "PC3 (4.5%)")))

# do the same for REs ----------------------------------------------------------
df2 <- merge(dre_lfc_pca_df, exp_annot, by = "condition")
fig2 <- plot_ly(df2, 
               x = ~PC1, 
               y = ~PC2, 
               z = ~PC3, 
               color = ~drug,
               colors = "Set1",
               hovertext = paste("Cell Type:", df2$cell_type, 
                                 "<br> Drug:", df2$drug,
                                 "<br> Dose", df2$dose,
                                 "<br> Time", df2$time_hr)) %>%
  add_markers() %>% 
  layout(
    title = list(text = "RE Expression - logFC"),
    scene = list(
      xaxis = list(title = "PC1 (34.8%)"),
      yaxis = list(title = "PC2 (5.3%)"),
      zaxis = list(title = "PC3 (3.1%)")))
```

Same deal but use rankings

```{r}
df3 <- merge(dge_rank_pca_df, exp_annot, by = "condition")
fig3 <- plot_ly(df3, 
               x = ~PC1, 
               y = ~PC2, 
               z = ~PC3, 
               color = ~drug,
               colors = "Set1",
               hovertext = paste("Cell Type:", df3$cell_type, 
                                 "<br> Drug:", df3$drug,
                                 "<br> Dose", df3$dose,
                                 "<br> Time", df3$time_hr)) %>%
  add_markers() %>% 
    layout(
    title = list(text = "Protein Coding Gene Expression - Gene Ranks"),
    scene = list(
      xaxis = list(title = "PC1 (23.3%)"),
      yaxis = list(title = "PC2 (18.1%)"),
      zaxis = list(title = "PC3 (8.8%%)")))

# same with REs ----------------------------------------------------------------
df4 <- merge(dre_rank_pca_df, exp_annot, by = "condition")
fig4 <- plot_ly(df4, 
               x = ~PC1, 
               y = ~PC2, 
               z = ~PC3, 
               color = ~drug,
               colors = "Set1",
               hovertext = paste("Cell Type:", df4$cell_type, 
                                 "<br> Drug:", df4$drug,
                                 "<br> Dose", df4$dose,
                                 "<br> Time", df4$time_hr)) %>%
  add_markers() %>% 
    layout(
    title = list(text = "RE Expression - Gene Ranks"),
    scene = list(
      xaxis = list(title = "PC1 (66.4%)"),
      yaxis = list(title = "PC2 (5.6%)"),
      zaxis = list(title = "PC3 (4.8%%)")))
```

## Save plots

```{r message=FALSE, warning=FALSE}
htmlwidgets::saveWidget(fig1, here("results", "combined", "figures", "dge-lfc-pca.html"))
htmlwidgets::saveWidget(fig2, here("results", "combined", "figures", "dre-lfc-pca.html"))
htmlwidgets::saveWidget(fig3, here("results", "combined", "figures", "dge-rank-pca.html"))
htmlwidgets::saveWidget(fig4, here("results", "combined", "figures", "dre-rank-pca.html"))
```

## Dendrogram / Heatmap

### Heatmap of most variable logFCs

```{r}
library(matrixStats)


# calculate unsupervised variance across logFC values
dge_lfc_vars <- rowVars(dge_lfc_mat, na.rm = TRUE)
dge_lfc_o <- order(dge_lfc_vars, decreasing = TRUE)
dge_lfc_top <- dge_lfc_mat[head(dge_lfc_o, n = 500), ]

# cluster rows and columns
dd_rows <- as.dendrogram(hclust(dist(dge_lfc_top, method = "euclidean"), method = "ward.D2"))
dd_cols <- as.dendrogram(hclust(dist(t(dge_lfc_top), method = "euclidean"), method = "ward.D2"))

# reorder the matrix by calculated clustering
dge_lfc_top <- dge_lfc_top[order.dendrogram(dd_rows), order.dendrogram(dd_cols)]

# match colnames from matrix with annotations
hm_annot <- exp_annot[match(colnames(dge_lfc_top), condition), ]

fig5 <- plot_ly(x = hm_annot$condition,
                y = rownames(dge_lfc_top),
                z = dge_lfc_top, 
                type = "heatmap") %>% 
  layout(title = list(text = "500 Most Variable Genes by logFC"))
```

Repeat for repeats

```{r}
# calculate unsupervised variance across logFC values
dre_lfc_vars <- rowVars(dre_lfc_mat, na.rm = TRUE)
dre_lfc_o <- order(dre_lfc_vars, decreasing = TRUE)
dre_lfc_top <- dre_lfc_mat[head(dre_lfc_o, n = 100), ]

# cluster rows and columns
dd_rows2 <- as.dendrogram(hclust(dist(dre_lfc_top, method = "euclidean"), method = "ward.D2"))
dd_cols2 <- as.dendrogram(hclust(dist(t(dre_lfc_top), method = "euclidean"), method = "ward.D2"))

# reorder the matrix by calculated clustering
dre_lfc_top <- dre_lfc_top[order.dendrogram(dd_rows2), order.dendrogram(dd_cols2)]

# match colnames from matrix with annotations
hm_annot2 <- exp_annot[match(colnames(dre_lfc_top), condition), ]

fig6 <- plot_ly(x = hm_annot2$condition,
                y = rownames(dre_lfc_top),
                z = dre_lfc_top, 
                type = "heatmap") %>% 
  layout(title = list(text = "100 Most Variable REs by logFC"))
```

and for calculated ranks

```{r}
# calculate unsupervised variance across logFC values
dge_rank_vars <- rowVars(dge_rank_mat, na.rm = TRUE)
dge_rank_o <- order(dge_rank_vars, decreasing = TRUE)
dge_rank_top <- dge_rank_mat[head(dge_rank_o, n = 500), ]

# cluster rows and columns
dd_rows3 <- as.dendrogram(hclust(dist(dge_rank_top, method = "euclidean"), method = "ward.D2"))
dd_cols3 <- as.dendrogram(hclust(dist(t(dge_rank_top), method = "euclidean"), method = "ward.D2"))

# reorder the matrix by calculated clustering
dge_rank_top <- dge_rank_top[order.dendrogram(dd_rows3), order.dendrogram(dd_cols3)]

# match colnames from matrix with annotations
hm_annot3 <- exp_annot[match(colnames(dge_rank_top), condition), ]

fig7 <- plot_ly(x = hm_annot3$condition,
                y = rownames(dge_rank_top),
                z = dge_rank_top, 
                type = "heatmap") %>% 
  layout(title = list(text = "500 Most Variable Genes by Rank"))
```

repeats by rank
 
```{r}
# calculate unsupervised variance across logFC values
dre_rank_vars <- rowVars(dre_rank_mat, na.rm = TRUE)
dre_rank_o <- order(dre_rank_vars, decreasing = TRUE)
dre_rank_top <- dre_rank_mat[head(dre_rank_o, n = 100), ]

# cluster rows and columns
dd_rows4 <- as.dendrogram(hclust(dist(dre_rank_top, method = "euclidean"), method = "ward.D2"))
dd_cols4 <- as.dendrogram(hclust(dist(t(dre_rank_top), method = "euclidean"), method = "ward.D2"))

# reorder the matrix by calculated clustering
dre_rank_top <- dre_rank_top[order.dendrogram(dd_rows4), order.dendrogram(dd_cols4)]

# match colnames from matrix with annotations
hm_annot4 <- exp_annot[match(colnames(dre_rank_top), condition), ]

fig8 <- plot_ly(x = hm_annot4$condition,
                y = rownames(dre_rank_top),
                z = dre_rank_top, 
                type = "heatmap") %>% 
  layout(title = list(text = "100 Most Variable REs by Rank"))
```

## Save widgets

```{r message=FALSE, warning=FALSE}
htmlwidgets::saveWidget(fig5, here("results", "combined", "figures", "dge-lfc-hm.html"))
htmlwidgets::saveWidget(fig6, here("results", "combined", "figures", "dge-rank-hm.html"))
htmlwidgets::saveWidget(fig7, here("results", "combined", "figures", "dre-lfc-hm.html"))
htmlwidgets::saveWidget(fig8, here("results", "combined", "figures", "dre-rank-hm.html"))
```

## Correlation analysis

```{r message=FALSE, warning=FALSE}
dge_rank_cor_mat <- cor(dge_rank_mat)
dd_rows5 <- as.dendrogram(hclust(dist(dge_rank_cor_mat, method = "euclidean"), method = "ward.D2"))
dd_cols5 <- as.dendrogram(hclust(dist(t(dge_rank_cor_mat), method = "euclidean"), method = "ward.D2"))
dge_rank_cor_mat <- dge_rank_cor_mat[order.dendrogram(dd_rows5), order.dendrogram(dd_cols5)]

fig9 <- plot_ly(z = dge_rank_cor_mat, 
        x = rownames(dge_rank_cor_mat),
        y = colnames(dge_rank_cor_mat),
        type = "heatmap") %>% 
  layout(title = list(text = "Correlation by Gene Rank"))

htmlwidgets::saveWidget(fig9, here("results", "combined", "figures", "dge-rank-corr.html"))
```

